<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2025-03-31 Mon 09:58 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>BD4H</title>
<meta name="author" content="W" />
<meta name="generator" content="Org Mode" />
<link rel="stylesheet" type="text/css" href="../src/readtheorg_theme/css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="../src/readtheorg_theme/css/readtheorg.css"/>
<script type="text/javascript" src="../src/lib/js/jquery.min.js"></script>
<script type="text/javascript" src="../src/lib/js/bootstrap.min.js"></script>
<script type="text/javascript" src="../src/lib/js/jquery.stickytableheaders.min.js"></script>
<script type="text/javascript" src="../src/readtheorg_theme/js/readtheorg.js"></script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content" class="content">
<h1 class="title">BD4H</h1>
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#org3b6b15f">1. Introduction to Big Data</a>
<ul>
<li><a href="#orgdcdefdb">1.1. Introduction</a></li>
<li><a href="#org6506ac4">1.2. About BDAH</a></li>
<li><a href="#orgcdf0e65">1.3. Learning Goals</a></li>
<li><a href="#org4113738">1.4. Current Problems in Healthcare</a></li>
<li><a href="#org98f1159">1.5. The Four Vs</a></li>
<li><a href="#org3d84679">1.6. More About The Four Vs</a></li>
<li><a href="#orgd34006b">1.7. Data Science is Sexy</a></li>
<li><a href="#orgecce6da">1.8. BDAH Quiz 1 Question</a></li>
<li><a href="#orgb357372">1.9. BDAH Quiz 1 Solution</a></li>
<li><a href="#orgeb4e790">1.10. BDAH Quiz 2 Question</a></li>
<li><a href="#org1daefa7">1.11. BDAH Quiz 2 Solution</a></li>
<li><a href="#org75d476c">1.12. BDAH Quiz 3 Question</a></li>
<li><a href="#org8c9734a">1.13. BDAH Quiz 3 Solution</a></li>
</ul>
</li>
<li><a href="#org058b484">2. Big Data Course Overview</a>
<ul>
<li><a href="#orgd796fdf">2.1. Introduction to Course Overview</a></li>
<li><a href="#orga639cdc">2.2. Big Data Big Picture</a></li>
<li><a href="#orgc8da741">2.3. Healthcare Applications</a></li>
<li><a href="#orgbd3a843">2.4. Predictive Modeling Quiz Question</a></li>
<li><a href="#org0381397">2.5. Predictive Modeling Quiz Solution</a></li>
<li><a href="#org8f7b8cb">2.6. Predictive Modeling Challenges</a></li>
<li><a href="#orgfc0d90c">2.7. Computational Phenotyping</a></li>
<li><a href="#orgefd08ea">2.8. Computational Phenotyping Quiz Question</a></li>
<li><a href="#org5f83925">2.9. Computational Phenotyping Quiz Solution</a></li>
<li><a href="#org721f8fe">2.10. Phenotyping Algorithm</a></li>
<li><a href="#org3f9a621">2.11. Patient Similarity Quiz Question</a></li>
<li><a href="#org5722b15">2.12. Patient Similarity Quiz Solution</a></li>
<li><a href="#orgba9df92">2.13. Patient Similarity</a></li>
<li><a href="#orgf7b6a7c">2.14. Algorithms</a></li>
<li><a href="#org869c998">2.15. Systems</a></li>
<li><a href="#org4ba4999">2.16. Summary</a></li>
</ul>
</li>
<li><a href="#org2fcb999">3. Predictive Modeling</a>
<ul>
<li><a href="#org36311c8">3.1. Introduction to Predictive Modeling</a></li>
<li><a href="#org98afadd">3.2. Predictive Modeling vs EHR</a></li>
<li><a href="#orgb1b6c19">3.3. Predictive Modeling Pipeline</a></li>
<li><a href="#org752f64b">3.4. Prediction Target</a></li>
<li><a href="#org360bb55">3.5. Heart Failure Quiz</a></li>
<li><a href="#orgc0018bc">3.6. Motivations for Early Detection</a></li>
<li><a href="#org469c439">3.7. Cohort Construction</a></li>
<li><a href="#org86ad9ce">3.8. Prospective vs. Retrospective Studies</a></li>
<li><a href="#org6054c6b">3.9. Prospective vs. Retrospective Quiz</a></li>
<li><a href="#org40b122b">3.10. Cohort Study</a></li>
<li><a href="#org8303860">3.11. Case-Control Study</a></li>
<li><a href="#orgaec5dfa">3.12. Feature Construction</a></li>
<li><a href="#org62d5a44">3.13. Feature Construction Quizzes</a></li>
<li><a href="#org8151c3d">3.14. Prediction Performance on Windows</a></li>
<li><a href="#org63be694">3.15. Feature Selection</a></li>
<li><a href="#org91112b1">3.16. Predictive Model</a></li>
<li><a href="#org4499835">3.17. Performance Evaluation</a></li>
<li><a href="#org1ae6cc0">3.18. Cross-Validation</a></li>
<li><a href="#orga3e12cb">3.19. Conclusion</a></li>
</ul>
</li>
<li><a href="#org2f96732">4. MapReduce</a>
<ul>
<li><a href="#orgec633d6">4.1. Introduction to MapReduce</a></li>
<li><a href="#orgc802c8a">4.2. What is MapReduce</a></li>
<li><a href="#org5ac41bc">4.3. Computational Process</a></li>
<li><a href="#orgad47f56">4.4. Learning Via Aggregation Statistics</a></li>
<li><a href="#org5afb42c">4.5. MapReduce Abstraction</a></li>
<li><a href="#orga6bc0ac">4.6. MapReduce System</a></li>
<li><a href="#org94f5bbc">4.7. MapReduce Fault Recovery</a></li>
<li><a href="#orga41d770">4.8. Distributed File Systems</a></li>
<li><a href="#orgc48e2e9">4.9. MapReduce Design Choice</a></li>
<li><a href="#org186f293">4.10. Analytics with MapReduce</a></li>
<li><a href="#org65040c6">4.11. MapReduce KNN</a></li>
<li><a href="#org381f569">4.12. Linear Regression</a></li>
<li><a href="#org82dbd04">4.13. MapReduce for Linear Regression Quiz Question</a></li>
<li><a href="#org8d7bc82">4.14. MapReduce for Linear Regression Quiz Solution</a></li>
<li><a href="#org72f809c">4.15. Limitations of MapReduce</a></li>
<li><a href="#org2afd447">4.16. MapReduce Summary Quiz Question</a></li>
<li><a href="#org82c1c18">4.17. MapReduce Summary Quiz Solution</a></li>
</ul>
</li>
<li><a href="#orgd9e0d21">5. Classification Model Metrics</a>
<ul>
<li><a href="#org110e43c">5.1. Predictive Model review</a></li>
<li><a href="#org3b07bc6">5.2. Confusion matrix</a></li>
<li><a href="#org0f5b4fc">5.3. Accuracy metrics</a>
<ul>
<li><a href="#orgd5d5725">5.3.1. Other notes</a></li>
</ul>
</li>
<li><a href="#orgc06e628">5.4. Predictive metrics</a></li>
<li><a href="#org0aebabe">5.5. F1 score</a></li>
<li><a href="#org463125f">5.6. Classifier quiz</a></li>
<li><a href="#orgea083e3">5.7. Reversing predictions</a></li>
<li><a href="#org9971d22">5.8. Receiver operating characteristic</a></li>
<li><a href="#org50d2663">5.9. "Best classifier threshold" quiz</a></li>
<li><a href="#org68a7577">5.10. Regression metrics</a>
<ul>
<li><a href="#org74475fe">5.10.1. \(R^2\)</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#orgec4539d">6. Ensemble methods</a>
<ul>
<li><a href="#org6f98068">6.1. Gradient Descent Method (GDM) for Linear Regression</a></li>
<li><a href="#orgc55183d">6.2. Stochastic Gradient Descent (SGD) Method</a></li>
<li><a href="#org3a65c15">6.3. SGD for Linear Regression</a></li>
<li><a href="#org59c0e82">6.4. Ensemble Method Pt 1</a></li>
<li><a href="#orga843412">6.5. Ensemble Method Pt 2</a></li>
<li><a href="#org676a854">6.6. Bias Variance Tradeoff</a></li>
<li><a href="#org4e04bd6">6.7. Bias Variance Tradeoff Quiz Question</a></li>
<li><a href="#org2e7163d">6.8. Bias Variance Tradeoff Quiz Solution</a></li>
<li><a href="#org94349da">6.9. Bias Variance Tradeoff Quiz 2 Question</a></li>
<li><a href="#orga7ba58b">6.10. Bias Variance Tradeoff Quiz 2 Solution</a></li>
<li><a href="#orgc2ca3f8">6.11. Bagging</a></li>
<li><a href="#org0456c0f">6.12. Random Forest</a></li>
<li><a href="#org82fbbc2">6.13. Why Bagging Works</a></li>
<li><a href="#org4c65cab">6.14. Boosting</a></li>
<li><a href="#org4aa4a97">6.15. Bagging vs Boosting Quiz Question</a></li>
<li><a href="#org75a8ba8">6.16. Bagging vs Boosting Quiz Solution</a></li>
<li><a href="#org53023b8">6.17. Summary for Ensemble Methods</a></li>
</ul>
</li>
<li><a href="#org2107e1f">7. Computational Phenotyping</a>
<ul>
<li><a href="#org3ddbe4b">7.1. Introduction to Phenotyping</a></li>
<li><a href="#org91fcee2">7.2. Computational Phenotyping</a></li>
<li><a href="#org9894fe5">7.3. Phenotyping Algorithm</a></li>
<li><a href="#orgafd5819">7.4. Applications of Phenotyping</a></li>
<li><a href="#orgdf4d0ad">7.5. Genomic Wide Association Study</a></li>
<li><a href="#org4bccaac">7.6. Why Do We Care About Phenotyping</a></li>
<li><a href="#orgfd29f72">7.7. Clinical Predictive Modeling</a></li>
<li><a href="#org4644e3f">7.8. Pragmatic Clinical Trials</a></li>
<li><a href="#org7533f0c">7.9. Healthcare Quality Measurement</a></li>
<li><a href="#orgcb20273">7.10. Phenotyping Methods Part 1</a></li>
<li><a href="#org110cb00">7.11. Phenotyping Methods Part 2</a></li>
<li><a href="#org09c04e8">7.12. Phenotyping Quiz Question</a></li>
<li><a href="#org3b90b46">7.13. Phenotyping Quiz Solution</a></li>
</ul>
</li>
<li><a href="#org492c6a3">8. Clustering</a>
<ul>
<li><a href="#org97fee4c">8.1. Introduction to Clustering</a></li>
<li><a href="#org2d75ab7">8.2. Healthcare Applications</a></li>
<li><a href="#orgf7d6310">8.3. What is Clustering</a></li>
<li><a href="#orgc8b2616">8.4. Algorithm Overview</a></li>
<li><a href="#orga113f24">8.5. K Means</a></li>
<li><a href="#org8dc849b">8.6. K Means Quiz Question</a></li>
<li><a href="#orgbaff805">8.7. K Means Quiz Solution</a></li>
<li><a href="#org8befffa">8.8. Hierarchical Clustering</a></li>
<li><a href="#orgd28ac28">8.9. Agglomerative Clustering</a></li>
<li><a href="#org81a71a7">8.10. Gaussian Mixture Model</a></li>
<li><a href="#org07d0f11">8.11. GMM Expectation Maximization</a></li>
<li><a href="#org9316cfe">8.12. GMM Steps</a></li>
<li><a href="#orgcf433fc">8.13. GMM Visual Illustration</a></li>
<li><a href="#org2354da2">8.14. K Means Vs GMM</a></li>
<li><a href="#orgc83d226">8.15. Mini Batch K Means</a></li>
<li><a href="#orge52ce88">8.16. Mini Batch K Means Quiz Question</a></li>
<li><a href="#org04f9a7a">8.17. Mini Batch K Means Quiz Solution</a></li>
<li><a href="#org9a1c82c">8.18. DBScan</a></li>
<li><a href="#org87c5647">8.19. DBScan Key Concepts</a></li>
<li><a href="#org67e8219">8.20. DBScan Algorithm</a></li>
<li><a href="#orgb0b3c1a">8.21. DBScan Example</a></li>
<li><a href="#org901317f">8.22. DBScan Quiz Question</a></li>
<li><a href="#org9ffc0a2">8.23. DBScan Quiz Solution</a></li>
<li><a href="#orgb259f02">8.24. Clustering Evaluation Metrics</a></li>
<li><a href="#orgcab3b37">8.25. Rand Index</a></li>
<li><a href="#orgbad96b2">8.26. Mutual Information</a></li>
<li><a href="#org4c3a617">8.27. Summary of RI and MI</a></li>
<li><a href="#org5dcd43f">8.28. Silhouette Coefficient</a></li>
<li><a href="#orgb6f4182">8.29. Silhouette Coefficient Pros and Cons</a></li>
</ul>
</li>
<li><a href="#orge3f9a30">9. Spark</a>
<ul>
<li><a href="#orgae58bc4">9.1. Introduction to Spark</a></li>
<li><a href="#org4670b80">9.2. Environment</a></li>
<li><a href="#org241e324">9.3. Motivation</a></li>
<li><a href="#orgf1bad78">9.4. Iteration in Map Reduce</a></li>
<li><a href="#org22f95a2">9.5. Workload Illustration</a></li>
<li><a href="#org4795d7b">9.6. Challenge</a></li>
<li><a href="#org4e62c95">9.7. Solution  RDDs</a></li>
<li><a href="#org696454a">9.8. RDD Recovery</a></li>
<li><a href="#org367ba48">9.9. Spark Stack</a></li>
<li><a href="#org3de0073">9.10. Spark Programming Interface</a></li>
<li><a href="#org4a94afc">9.11. RDD Transformations</a></li>
<li><a href="#orga4dfb01">9.12. RDD Transformations Quiz Question</a></li>
<li><a href="#org9ce33ca">9.13. RDD Transformations Quiz Solution</a></li>
<li><a href="#orge32532e">9.14. Spark Operations</a></li>
<li><a href="#org391c68a">9.15. Shared Variable</a></li>
<li><a href="#org799c26e">9.16. Fault Tolerance</a></li>
<li><a href="#org0ca2b32">9.17. Example  Logistic Regression</a></li>
<li><a href="#org3f4fbd8">9.18. Example  Disease Risk Prediction</a></li>
<li><a href="#org126714f">9.19. Serial ALS</a></li>
<li><a href="#org49f92a0">9.20. Naive Spark ALS</a></li>
</ul>
</li>
<li><a href="#org335123a">10. Medical Ontology</a>
<ul>
<li><a href="#orga0a5f88">10.1. Introduction to Medical Ontology</a></li>
<li><a href="#org4297432">10.2. Health Data Standards</a></li>
<li><a href="#org18de379">10.3. ICD</a></li>
<li><a href="#org83e331e">10.4. ICD 9 to ICD 10 Mapping</a></li>
<li><a href="#orgf0d5795">10.5. ICD 9 Quiz Question</a></li>
<li><a href="#orgd9c18c7">10.6. ICD 9 Quiz Solution</a></li>
<li><a href="#orge861efc">10.7. ICD Code Quiz Question</a></li>
<li><a href="#org54a19a8">10.8. ICD Code Quiz Solution</a></li>
<li><a href="#org2f13aca">10.9. CPT</a></li>
<li><a href="#orga0f3e3e">10.10. CPT Code Quiz Question</a></li>
<li><a href="#org78504db">10.11. CPT Code Quiz Solution</a></li>
<li><a href="#org60d970e">10.12. LOINC</a></li>
<li><a href="#org9ced1af">10.13. LOINC Code Quiz Question</a></li>
<li><a href="#org9413dd1">10.14. LOINC Code Quiz Solution</a></li>
<li><a href="#orge522008">10.15. NDC</a></li>
<li><a href="#org7b51962">10.16. NDC Code Quiz Question</a></li>
<li><a href="#org7123e47">10.17. NDC Code Quiz Solution</a></li>
<li><a href="#org3d0ba6e">10.18. SNOMED</a></li>
<li><a href="#orgafab91f">10.19. Logical Model of SNOMED CT</a></li>
<li><a href="#orge6725c7">10.20. SNOMED Example</a></li>
<li><a href="#orgd040faa">10.21. SNOMED Relationships</a></li>
<li><a href="#org32d9b73">10.22. SNOMED Design</a></li>
<li><a href="#org7deea18">10.23. SNOMED Code Quiz Question</a></li>
<li><a href="#orge0ad860">10.24. SNOMED Code Quiz Solution</a></li>
<li><a href="#org95b914e">10.25. SNOMED Quiz Question</a></li>
<li><a href="#org504b8e5">10.26. SNOMED Quiz Solution</a></li>
<li><a href="#org9f68b32">10.27. UMLS</a></li>
<li><a href="#org252770f">10.28. Metathesaurus Concepts</a></li>
<li><a href="#orgaf7e657">10.29. Semantic Network</a></li>
<li><a href="#orgef4b9af">10.30. Specialist Lexicon</a></li>
</ul>
</li>
<li><a href="#org188b11d">11. Graph Analysis</a>
<ul>
<li><a href="#orgc95d7fe">11.1. Introduction to Graph Analysis</a></li>
<li><a href="#org3358431">11.2. Agenda</a></li>
<li><a href="#org3544f4c">11.3. PageRank</a></li>
<li><a href="#orge21abfe">11.4. MapReduce  PageRank</a></li>
<li><a href="#org1214d90">11.5. PageRank Quiz Question</a></li>
<li><a href="#org11fb5a9">11.6. PageRank Quiz Solution</a></li>
<li><a href="#orgdb6804e">11.7. Spectral Clustering</a></li>
<li><a href="#org6263fd3">11.8. Similarity Graph Construction</a></li>
<li><a href="#org4304566">11.9. E Neighborhood Graph</a></li>
<li><a href="#org1ad4d87">11.10. K Nearest Neighbor Graph</a></li>
<li><a href="#orgbf0eb70">11.11. Fully Connected Graph</a></li>
<li><a href="#org4e4cf2a">11.12. E Neighborhood Graph Quiz Question</a></li>
<li><a href="#orgfd58f5f">11.13. E Neighborhood Graph Quiz Solution</a></li>
<li><a href="#org35437d3">11.14. Spectral Clustering Algorithm</a></li>
<li><a href="#orge9bff4e">11.15. Big Data Conclusion</a></li>
</ul>
</li>
<li><a href="#org4d21bbb">12. Dimensionality Reduction</a>
<ul>
<li><a href="#org8e048ca">12.1. Introduction to Dimensionality Reduction</a></li>
<li><a href="#org4ff8afe">12.2. Dimensionality Reduction</a></li>
<li><a href="#org82e6e06">12.3. Singular Value Decomposition</a></li>
<li><a href="#org9b4fa61">12.4. SVD Example</a></li>
<li><a href="#org737d351">12.5. SVD Properties</a></li>
<li><a href="#org53fc049">12.6. Quiz  SVD Interpretation Question</a></li>
<li><a href="#org8d27be7">12.7. Quiz  SVD Interpretation Solution</a></li>
<li><a href="#orgaa0abbc">12.8. Principal Component Analysis</a></li>
<li><a href="#orge2af891">12.9. PCA Interpretation</a></li>
<li><a href="#org31677cc">12.10. Sparsity Problem with SVD</a></li>
<li><a href="#org0c98494">12.11. CUR Decomposition</a></li>
<li><a href="#orgb68196b">12.12. CUR Algorithm</a></li>
<li><a href="#org3e85fcf">12.13. CUR Quiz Question</a></li>
<li><a href="#org2793f85">12.14. CUR Quiz Solution</a></li>
<li><a href="#org5730ef6">12.15. Tensor for EHR</a></li>
<li><a href="#orga4b312e">12.16. Tensor Slicing</a></li>
<li><a href="#org239aae5">12.17. Rank 1 Tensor</a></li>
<li><a href="#org50be58b">12.18. Example Phenotype</a></li>
<li><a href="#org080cf2c">12.19. Phenotyping Through Tensor Factorization</a></li>
<li><a href="#org17eb805">12.20. CP Decomposition</a></li>
<li><a href="#org5ea4ea0">12.21. Phenotyping Process Using Tensor Factorization</a></li>
<li><a href="#org1d7accb">12.22. Phenotyping for Predictive Modeling</a></li>
<li><a href="#orgfb62dca">12.23. Predictive Performance</a></li>
<li><a href="#org7f3d309">12.24. Major Disease Phenotypes</a></li>
<li><a href="#orge64f63d">12.25. Tensor vs NMF</a></li>
<li><a href="#org1d92418">12.26. Summary  Phenotyping Via Tensor Factorization</a></li>
</ul>
</li>
<li><a href="#orgf3d6086">13. Patient Similarity</a>
<ul>
<li><a href="#orgbcbc22f">13.1. Introduction to Patient Similarity</a></li>
<li><a href="#orgb5d3f2d">13.2. Motivation</a></li>
<li><a href="#org39c18c2">13.3. Traditional Paradigm</a></li>
<li><a href="#orgd304aff">13.4. New Paradigm</a></li>
<li><a href="#org4bd0f0a">13.5. Randomized Clinical Trials</a></li>
<li><a href="#org83773af">13.6. RCT Quiz Question</a></li>
<li><a href="#orgb1b32f8">13.7. RCT Quiz Solution</a></li>
<li><a href="#org2eee673">13.8. Pragmatic Trials</a></li>
<li><a href="#orge02ff4d">13.9. Pragmatic Trial Quiz Question</a></li>
<li><a href="#org4d6a1fa">13.10. Pragmatic Trial Quiz Solution</a></li>
<li><a href="#org3cdb4d5">13.11. Utilize Patient Similarity</a></li>
<li><a href="#orgb03a606">13.12. Patient Similarity Approaches</a></li>
<li><a href="#org710b6fd">13.13. Patient Similarity Through LSML</a></li>
</ul>
</li>
<li><a href="#org1a80890">14. Deep Neural Network</a>
<ul>
<li><a href="#org703cee1">14.1. Backward Computation for a Neuron</a></li>
<li><a href="#orgf59b2c5">14.2. Sigmoid Function</a></li>
<li><a href="#org26f5bfc">14.3. TANH Function</a></li>
<li><a href="#orgd344a12">14.4. Rectified Linear Function</a></li>
<li><a href="#orge2a1e4a">14.5. Activation Functions_ Summary</a></li>
<li><a href="#org8f9e8f5">14.6. Train a Single Neuron</a></li>
<li><a href="#org5a4df1d">14.7. Stochastic Gradient Descent</a></li>
<li><a href="#orga65194e">14.8. Forward Computation for a Neuron</a></li>
<li><a href="#org80f90d6">14.9. Backward Computation for a Neuron</a></li>
<li><a href="#orge581478">14.10. Train a Multilayer Neural Network</a></li>
<li><a href="#org0589a95">14.11. Forward Computation Part 1</a></li>
<li><a href="#org5c9a037">14.12. Forward Computation Part 2</a></li>
<li><a href="#orgfd5aba4">14.13. Forward Computation_ Vector Form</a></li>
<li><a href="#org53f7221">14.14. Forward Computation_ More General Form</a></li>
<li><a href="#orgc1d9a51">14.15. Forward Computation_ Summary</a></li>
<li><a href="#orge96fd64">14.16. Gradient Descent for Neural Networks</a></li>
<li><a href="#org41d7fe6">14.17. Backward Propagation Part 1</a></li>
<li><a href="#org770436b">14.18. Backward Propagation Part 2</a></li>
<li><a href="#org47e187e">14.19. Backward Propagation Part 3</a></li>
<li><a href="#org01d630f">14.20. Backward Propagation Part 4</a></li>
<li><a href="#orgf68ad05">14.21. Backward Propagation Part 5</a></li>
<li><a href="#org469385f">14.22. Backward Propagation Part 6</a></li>
<li><a href="#orgf0dc753">14.23. Backward Propagation Summary</a></li>
</ul>
</li>
<li><a href="#org36895ef">15. Convolutional Neural Network</a>
<ul>
<li><a href="#orgb3b133c">15.1. Convolution Local Networks</a></li>
<li><a href="#orgb0af3aa">15.2. Pooling Handling Distortion</a></li>
<li><a href="#org3c6c7fa">15.3. Convolutional Neural Networks</a></li>
<li><a href="#orgb829397">15.4. Key Stages of CNN</a></li>
<li><a href="#org4960d75">15.5. 1-D Convolution</a></li>
<li><a href="#org8db390e">15.6. Neural Network for 1-D Convolution</a></li>
<li><a href="#orgc40f458">15.7. Convolution Layers</a></li>
<li><a href="#orgae1ec35">15.8. Pooling Layers</a></li>
<li><a href="#orga3bf11f">15.9. Dimension Calculation of Convolution Layers</a></li>
<li><a href="#orgd94a853">15.10. Dimension Calculation of Pooling Layer</a></li>
<li><a href="#org49cb9be">15.11. Parameter Count</a></li>
<li><a href="#org3c0dee4">15.12. Forward Calculation of Convolution</a></li>
<li><a href="#org6236e8f">15.13. Forward Calculation of Pooling Layers</a></li>
<li><a href="#org387bb8e">15.14. Forward Calculation of Fully-Connected Layers</a></li>
<li><a href="#org1ad6655">15.15. CNN Architectures and Healthcare Applications</a></li>
<li><a href="#org837abcc">15.16. Diabetic Retinopathy Diagnosis</a></li>
<li><a href="#org11a908e">15.17. Dermatologist Classification of Skin Cancer with DNN</a></li>
</ul>
</li>
<li><a href="#org331eb2b">16. Recurrent Neural Network</a>
<ul>
<li><a href="#orgeee34ff">16.1. Basic Concepts of RNN</a></li>
<li><a href="#org97ebede">16.2. Basic RNN Structure</a></li>
<li><a href="#org0c8403b">16.3. Forward Computation</a></li>
<li><a href="#org15ad563">16.4. Backpropagation Through Time</a></li>
<li><a href="#org9de46bd">16.5. Standard RNN</a></li>
<li><a href="#org4ce52cb">16.6. Gated Recurred Unit</a></li>
<li><a href="#org0fef65a">16.7. Bidirectional RNN</a></li>
<li><a href="#org325bc7e">16.8. Healthcare Applications</a></li>
</ul>
</li>
<li><a href="#orgd19363f">17. Focus</a></li>
</ul>
</div>
</div>
<div id="outline-container-org3b6b15f" class="outline-2">
<h2 id="org3b6b15f"><span class="section-number-2">1.</span> Introduction to Big Data</h2>
<div class="outline-text-2" id="text-1">
</div>
<div id="outline-container-orgdcdefdb" class="outline-3">
<h3 id="orgdcdefdb"><span class="section-number-3">1.1.</span> Introduction</h3>
<div class="outline-text-3" id="text-1-1">
<ul class="org-ul">
<li>Course: Big Data Analytics for Healthcare (BDAH)</li>
<li>Instructor: Jimeng Sun, associate professor at Georgia Tech</li>
<li>Background: Expertise in healthcare analytics and data mining, previous work at IBM TJ Watson Research Center</li>
<li>Course focus: What students will learn and why it matters</li>
</ul>
</div>
</div>

<div id="outline-container-org6506ac4" class="outline-3">
<h3 id="org6506ac4"><span class="section-number-3">1.2.</span> About BDAH</h3>
<div class="outline-text-3" id="text-1-2">
<ul class="org-ul">
<li>Intersection of healthcare and big data</li>
<li>Covers healthcare applications, data, analytics, and big data processing</li>
<li>Focus on how data science is applied to healthcare</li>
</ul>
</div>
</div>

<div id="outline-container-orgcdf0e65" class="outline-3">
<h3 id="orgcdf0e65"><span class="section-number-3">1.3.</span> Learning Goals</h3>
<div class="outline-text-3" id="text-1-3">
<ul class="org-ul">
<li>Understand healthcare data</li>
<li>Learn analytics algorithms</li>
<li>Work with big data systems</li>
<li>Application: Build models for disease risk prediction, treatment recommendation, patient clustering, and similarity analysis</li>
<li>Assessments: Homework using big data tools, project with system building, report writing, and presentations</li>
</ul>
</div>
</div>

<div id="outline-container-org4113738" class="outline-3">
<h3 id="org4113738"><span class="section-number-3">1.4.</span> Current Problems in Healthcare</h3>
<div class="outline-text-3" id="text-1-4">
<ul class="org-ul">
<li>High costs: $3.8 trillion/year in the U.S.</li>
<li>Massive waste: $765 billion/year</li>
<li>Poor quality: 200,000-400,000 preventable deaths annually</li>
<li>Preventable deaths ranked as the third leading cause of death</li>
<li>Hope: Big data can improve care and reduce costs</li>
</ul>
</div>
</div>

<div id="outline-container-org98f1159" class="outline-3">
<h3 id="org98f1159"><span class="section-number-3">1.5.</span> The Four Vs</h3>
<div class="outline-text-3" id="text-1-5">
<ul class="org-ul">
<li>Volume: Large healthcare data amounts</li>
<li>Variety: Different types of healthcare data sources</li>
<li>Velocity: Real-time data processing needs</li>
<li>Veracity: Issues with data quality (noise, missing data, errors)</li>
</ul>
</div>
</div>

<div id="outline-container-org3d84679" class="outline-3">
<h3 id="org3d84679"><span class="section-number-3">1.6.</span> More About The Four Vs</h3>
<div class="outline-text-3" id="text-1-6">
<ul class="org-ul">
<li>Examples of data volume:
<ul class="org-ul">
<li>Genomic data: 200 GB per genome</li>
<li>fMRI scan: ~300 GB per scan</li>
<li>US medical imaging data per year: ~100 petabytes</li>
</ul></li>
<li>Data variety:
<ul class="org-ul">
<li>Clinical (demographics, diagnosis, procedures, etc.)</li>
<li>Patient-generated (wearables, sensors)</li>
<li>Real-time (ICU monitoring data)</li>
</ul></li>
<li>Focus on managing diverse data types in the course</li>
</ul>
</div>
</div>

<div id="outline-container-orgd34006b" class="outline-3">
<h3 id="orgd34006b"><span class="section-number-3">1.7.</span> Data Science is Sexy</h3>
<div class="outline-text-3" id="text-1-7">
<ul class="org-ul">
<li>Data science is a critical field</li>
<li>Harvard Business Review: "Data Scientist, The Sexiest Job in the 21st Century"</li>
<li>Data scientists:
<ul class="org-ul">
<li>Capitalize on big data</li>
<li>Overcome technical limitations</li>
<li>Develop key tools like Hadoop and Spark</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgecce6da" class="outline-3">
<h3 id="orgecce6da"><span class="section-number-3">1.8.</span> BDAH Quiz 1 Question</h3>
<div class="outline-text-3" id="text-1-8">
<ul class="org-ul">
<li>True/False: A graduate-level degree is necessary to become a data scientist.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb357372" class="outline-3">
<h3 id="orgb357372"><span class="section-number-3">1.9.</span> BDAH Quiz 1 Solution</h3>
<div class="outline-text-3" id="text-1-9">
<ul class="org-ul">
<li>Answer: True</li>
</ul>
</div>
</div>

<div id="outline-container-orgeb4e790" class="outline-3">
<h3 id="orgeb4e790"><span class="section-number-3">1.10.</span> BDAH Quiz 2 Question</h3>
<div class="outline-text-3" id="text-1-10">
<ul class="org-ul">
<li>Question: What is the average salary of a data scientist?</li>
</ul>
</div>
</div>

<div id="outline-container-org1daefa7" class="outline-3">
<h3 id="org1daefa7"><span class="section-number-3">1.11.</span> BDAH Quiz 2 Solution</h3>
<div class="outline-text-3" id="text-1-11">
<ul class="org-ul">
<li>Answer: $120,000 per year (experienced data scientists earn over $150,000)</li>
</ul>
</div>
</div>

<div id="outline-container-org75d476c" class="outline-3">
<h3 id="org75d476c"><span class="section-number-3">1.12.</span> BDAH Quiz 3 Question</h3>
<div class="outline-text-3" id="text-1-12">
<ul class="org-ul">
<li>Question: What skills do data scientists need?</li>
</ul>
</div>
</div>

<div id="outline-container-org8c9734a" class="outline-3">
<h3 id="org8c9734a"><span class="section-number-3">1.13.</span> BDAH Quiz 3 Solution</h3>
<div class="outline-text-3" id="text-1-13">
<ul class="org-ul">
<li>Answer:
<ul class="org-ul">
<li>Math and statistics</li>
<li>Domain knowledge</li>
<li>Programming and databases</li>
<li>Communication and visualization</li>
</ul></li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org058b484" class="outline-2">
<h2 id="org058b484"><span class="section-number-2">2.</span> Big Data Course Overview</h2>
<div class="outline-text-2" id="text-2">
</div>
<div id="outline-container-orgd796fdf" class="outline-3">
<h3 id="orgd796fdf"><span class="section-number-3">2.1.</span> Introduction to Course Overview</h3>
<div class="outline-text-3" id="text-2-1">
<ul class="org-ul">
<li>Overview of big data analytics for healthcare.</li>
<li>Topics covered:
<ul class="org-ul">
<li>Healthcare applications of big data.</li>
<li>Algorithms used in applications.</li>
<li>Software systems for implementation.</li>
</ul></li>
<li>Course structure alternates among these topics.</li>
</ul>
</div>
</div>

<div id="outline-container-orga639cdc" class="outline-3">
<h3 id="orga639cdc"><span class="section-number-3">2.2.</span> Big Data Big Picture</h3>
<div class="outline-text-3" id="text-2-2">
<ul class="org-ul">
<li>Course focuses on three key areas:
<ul class="org-ul">
<li>Big data systems.</li>
<li>Scalable machine learning algorithms.</li>
<li>Healthcare applications.</li>
</ul></li>
<li>Integration of these elements to solve healthcare problems.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc8da741" class="outline-3">
<h3 id="orgc8da741"><span class="section-number-3">2.3.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-2-3">
<ul class="org-ul">
<li>Three types of healthcare applications:
<ol class="org-ol">
<li><b><b>Predictive Modeling</b></b>: Uses historical data to predict future outcomes.</li>
<li><b><b>Computational Phenotyping</b></b>: Converts electronic health records (EHRs) into meaningful clinical concepts.</li>
<li><b><b>Patient Similarity</b></b>: Identifies groups of patients with similar characteristics.</li>
</ol></li>
<li>Starts with predictive modeling.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbd3a843" class="outline-3">
<h3 id="orgbd3a843"><span class="section-number-3">2.4.</span> Predictive Modeling Quiz Question</h3>
<div class="outline-text-3" id="text-2-4">
<ul class="org-ul">
<li>Predictive modeling example: predicting treatment effectiveness for epilepsy patients.</li>
<li>Quiz: Estimate percentage of patients responding to different treatments.</li>
</ul>
</div>
</div>

<div id="outline-container-org0381397" class="outline-3">
<h3 id="org0381397"><span class="section-number-3">2.5.</span> Predictive Modeling Quiz Solution</h3>
<div class="outline-text-3" id="text-2-5">
<ul class="org-ul">
<li>Solution breakdown:
<ul class="org-ul">
<li><b><b>Group A</b></b> (responded within 2 years): 32%.</li>
<li><b><b>Group B</b></b> (responded between 2-5 years): 24%.</li>
<li><b><b>Group C</b></b> (did not respond after 5 years): 44%.</li>
</ul></li>
<li>Goal of predictive modeling:
<ul class="org-ul">
<li>Improve early response rates.</li>
<li>Identify non-responders for alternative treatments.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8f7b8cb" class="outline-3">
<h3 id="org8f7b8cb"><span class="section-number-3">2.6.</span> Predictive Modeling Challenges</h3>
<div class="outline-text-3" id="text-2-6">
<ul class="org-ul">
<li>Challenges in predictive modeling:
<ul class="org-ul">
<li>Handling large patient datasets with various types of data.</li>
<li>Managing multiple models in a complex computational pipeline.</li>
<li>Evaluating and comparing multiple predictive pipelines.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgfc0d90c" class="outline-3">
<h3 id="orgfc0d90c"><span class="section-number-3">2.7.</span> Computational Phenotyping</h3>
<div class="outline-text-3" id="text-2-7">
<ul class="org-ul">
<li>Converts raw patient data (e.g., demographics, diagnoses, medications, procedures, lab tests, clinical notes) into clinical concepts or phenotypes.</li>
</ul>
</div>
</div>

<div id="outline-container-orgefd08ea" class="outline-3">
<h3 id="orgefd08ea"><span class="section-number-3">2.8.</span> Computational Phenotyping Quiz Question</h3>
<div class="outline-text-3" id="text-2-8">
<ul class="org-ul">
<li>Identifying data issues in phenotyping.</li>
<li>Quiz: List potential "waste products" in raw data.</li>
</ul>
</div>
</div>

<div id="outline-container-org5f83925" class="outline-3">
<h3 id="org5f83925"><span class="section-number-3">2.9.</span> Computational Phenotyping Quiz Solution</h3>
<div class="outline-text-3" id="text-2-9">
<ul class="org-ul">
<li>Common data issues:
<ul class="org-ul">
<li><b><b>Missing values</b></b>: Some important data may be absent.</li>
<li><b><b>Duplicates</b></b>: Patient records may appear multiple times.</li>
<li><b><b>Irrelevant data</b></b>: Not all raw information is useful.</li>
<li><b><b>Redundant information</b></b>: Different records may indicate the same condition (e.g., diagnosis and medication for diabetes).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org721f8fe" class="outline-3">
<h3 id="org721f8fe"><span class="section-number-3">2.10.</span> Phenotyping Algorithm</h3>
<div class="outline-text-3" id="text-2-10">
<ul class="org-ul">
<li>Example: Identifying Type 2 Diabetes from EHR data.</li>
<li>Decision process:
<ul class="org-ul">
<li>Check for Type 1 Diabetes diagnosis.</li>
<li>Check for Type 2 Diabetes diagnosis.</li>
<li>Verify medication records and abnormal lab results.</li>
</ul></li>
<li>Importance:
<ul class="org-ul">
<li>EHR data is often unreliable.</li>
<li>Multiple data sources improve diagnostic accuracy.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org3f9a621" class="outline-3">
<h3 id="org3f9a621"><span class="section-number-3">2.11.</span> Patient Similarity Quiz Question</h3>
<div class="outline-text-3" id="text-2-11">
<ul class="org-ul">
<li>Different types of reasoning doctors use:
<ul class="org-ul">
<li>Flowchart-based reasoning (like phenotyping algorithms).</li>
<li>Instinct and intuition.</li>
<li>Case-based reasoning (comparing patients to past cases).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org5722b15" class="outline-3">
<h3 id="org5722b15"><span class="section-number-3">2.12.</span> Patient Similarity Quiz Solution</h3>
<div class="outline-text-3" id="text-2-12">
<ul class="org-ul">
<li>Correct answer: <b><b>Case-based reasoning</b></b>.</li>
<li>Doctors often compare current patients to previous similar cases.</li>
</ul>
</div>
</div>

<div id="outline-container-orgba9df92" class="outline-3">
<h3 id="orgba9df92"><span class="section-number-3">2.13.</span> Patient Similarity</h3>
<div class="outline-text-3" id="text-2-13">
<ul class="org-ul">
<li>Simulating case-based reasoning using algorithms.</li>
<li>Process:
<ul class="org-ul">
<li>Search for similar patients in a database.</li>
<li>Identify treatment outcomes for similar cases.</li>
<li>Recommend best treatment based on historical data.</li>
</ul></li>
<li>Objective: Utilize full database knowledge instead of relying on a single doctor's experience.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf7b6a7c" class="outline-3">
<h3 id="orgf7b6a7c"><span class="section-number-3">2.14.</span> Algorithms</h3>
<div class="outline-text-3" id="text-2-14">
<ul class="org-ul">
<li>Introduction to machine learning algorithms in healthcare.</li>
<li>Covered topics:
<ol class="org-ol">
<li><b><b>Classification</b></b>: Mapping patient data to target variables (e.g., predicting heart attack risk).</li>
<li><b><b>Clustering</b></b>: Grouping similar patients based on health conditions.</li>
<li><b><b>Dimensionality Reduction</b></b>: Reducing large patient datasets to essential features.</li>
<li><b><b>Graph Analysis</b></b>: Analyzing relationships between patients and diseases.</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org869c998" class="outline-3">
<h3 id="org869c998"><span class="section-number-3">2.15.</span> Systems</h3>
<div class="outline-text-3" id="text-2-15">
<ul class="org-ul">
<li>Introduction to big data systems for healthcare applications.</li>
<li>Covered systems:
<ul class="org-ul">
<li><b><b>Hadoop</b></b>: Disk-based distributed system.</li>
<li><b><b>Spark</b></b>: In-memory distributed system (faster than Hadoop).</li>
</ul></li>
<li>Topics include:
<ul class="org-ul">
<li>Hadoop infrastructure (MapReduce, HDFS, Pig, Hive, HBase).</li>
<li>Spark infrastructure (Spark SQL, Spark Streaming, MLlib, GraphX).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org4ba4999" class="outline-3">
<h3 id="org4ba4999"><span class="section-number-3">2.16.</span> Summary</h3>
<div class="outline-text-3" id="text-2-16">
<ul class="org-ul">
<li>Recap of three key areas:
<ul class="org-ul">
<li>Healthcare applications.</li>
<li>Machine learning algorithms.</li>
<li>Big data systems.</li>
</ul></li>
<li>Course integrates these areas:
<ul class="org-ul">
<li>Example: Using logistic regression on Hadoop to predict heart failure.</li>
</ul></li>
<li>Next step: Begin applying concepts.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org2fcb999" class="outline-2">
<h2 id="org2fcb999"><span class="section-number-2">3.</span> Predictive Modeling</h2>
<div class="outline-text-2" id="text-3">
</div>
<div id="outline-container-org36311c8" class="outline-3">
<h3 id="org36311c8"><span class="section-number-3">3.1.</span> Introduction to Predictive Modeling</h3>
<div class="outline-text-3" id="text-3-1">
<dl class="org-dl">
<dt>Predictive modeling</dt><dd>using historical data to predict future events.

<ul class="org-ul">
<li>Example: Using electronic health records (EHR) to model heart failure</li>
<li>Key Goal: Develop a good predictive model using EHR efficiently.</li>
</ul></dd>
</dl>
</div>
</div>

<div id="outline-container-org98afadd" class="outline-3">
<h3 id="org98afadd"><span class="section-number-3">3.2.</span> Predictive Modeling vs EHR</h3>
<div class="outline-text-3" id="text-3-2">
<ul class="org-ul">
<li>Importance: Increased research interest in using EHR for clinical predictive modeling.</li>
<li>Data Sources: EHR has become a major source for predictive modeling research.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb1b6c19" class="outline-3">
<h3 id="orgb1b6c19"><span class="section-number-3">3.3.</span> Predictive Modeling Pipeline</h3>
<div class="outline-text-3" id="text-3-3">
<ul class="org-ul">
<li>Predictive modeling is a multi-step computational process:
<ol class="org-ol">
<li>Define the prediction target.</li>
<li>Construct the relevant patient cohort.</li>
<li>Identify potentially relevant features.</li>
<li>Select the most relevant features.</li>
<li>Compute the predictive model.</li>
<li>Evaluate the predictive model.</li>
</ol></li>
<li>Iterative process until a satisfactory model is obtained.</li>
</ul>
</div>
</div>

<div id="outline-container-org752f64b" class="outline-3">
<h3 id="org752f64b"><span class="section-number-3">3.4.</span> Prediction Target</h3>
<div class="outline-text-3" id="text-3-4">
<ul class="org-ul">
<li>Investigators may have many targets, but only some are feasible.</li>
<li>Selection criteria: The target should be interesting and possible with the available data.</li>
<li>Example: Predicting the onset of heart failure.</li>
</ul>
</div>
</div>

<div id="outline-container-org360bb55" class="outline-3">
<h3 id="org360bb55"><span class="section-number-3">3.5.</span> Heart Failure Quiz</h3>
<div class="outline-text-3" id="text-3-5">
<ul class="org-ul">
<li>Question: How many new heart failure cases occur annually in the U.S.?</li>
<li>Options: 17,000; 260,000; 550,000; 1,250,000.</li>
<li>Answer: 550,000 cases per year.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc0018bc" class="outline-3">
<h3 id="orgc0018bc"><span class="section-number-3">3.6.</span> Motivations for Early Detection</h3>
<div class="outline-text-3" id="text-3-6">
<ul class="org-ul">
<li>Heart failure is complex with diverse symptoms and subsets.</li>
<li>Early detection can:
<ul class="org-ul">
<li>Reduce hospitalization costs.</li>
<li>Introduce early interventions to slow progression.</li>
<li>Improve clinical guidelines for heart failure prevention.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org469c439" class="outline-3">
<h3 id="org469c439"><span class="section-number-3">3.7.</span> Cohort Construction</h3>
<div class="outline-text-3" id="text-3-7">
<ul class="org-ul">
<li>Defines the study population for predictive modeling.</li>
<li>Study population is a subset of all patients.</li>
<li>Four study designs based on two axes:
<ul class="org-ul">
<li>Prospective vs. Retrospective studies.</li>
<li>Cohort vs. Case-Control studies.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org86ad9ce" class="outline-3">
<h3 id="org86ad9ce"><span class="section-number-3">3.8.</span> Prospective vs. Retrospective Studies</h3>
<div class="outline-text-3" id="text-3-8">
<ul class="org-ul">
<li><b><b>Prospective Study</b></b>: Define cohort first, then collect data.</li>
<li><b><b>Retrospective Study</b></b>: Use existing data from past records.</li>
</ul>
</div>
</div>

<div id="outline-container-org6054c6b" class="outline-3">
<h3 id="org6054c6b"><span class="section-number-3">3.9.</span> Prospective vs. Retrospective Quiz</h3>
<div class="outline-text-3" id="text-3-9">
<ul class="org-ul">
<li>Comparison of study properties:
<ul class="org-ul">
<li>Retrospective studies have more noise.</li>
<li>Prospective studies are more expensive and time-consuming.</li>
<li>Retrospective studies can handle larger datasets.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org40b122b" class="outline-3">
<h3 id="org40b122b"><span class="section-number-3">3.10.</span> Cohort Study</h3>
<div class="outline-text-3" id="text-3-10">
<ul class="org-ul">
<li>Includes patients exposed to a particular risk (e.g., heart failure readmission).</li>
<li>Defines inclusion and exclusion criteria.</li>
</ul>
</div>
</div>

<div id="outline-container-org8303860" class="outline-3">
<h3 id="org8303860"><span class="section-number-3">3.11.</span> Case-Control Study</h3>
<div class="outline-text-3" id="text-3-11">
<ul class="org-ul">
<li>Compares patients with a condition (cases) to similar patients without it (controls).</li>
<li>Matching criteria: Age, gender, clinic visits.</li>
<li>Cases are rarer than controls in real-world data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaec5dfa" class="outline-3">
<h3 id="orgaec5dfa"><span class="section-number-3">3.12.</span> Feature Construction</h3>
<div class="outline-text-3" id="text-3-12">
<ul class="org-ul">
<li>Defines patient features for predicting outcomes.</li>
<li>Data sequences from EHR:
<ul class="org-ul">
<li>Events (diagnoses, medications, lab results).</li>
<li>Observation window (used for feature extraction).</li>
<li>Prediction window (future period to predict).</li>
</ul></li>
<li>Impact of window sizes on model accuracy.</li>
</ul>
</div>
</div>

<div id="outline-container-org62d5a44" class="outline-3">
<h3 id="org62d5a44"><span class="section-number-3">3.13.</span> Feature Construction Quizzes</h3>
<div class="outline-text-3" id="text-3-13">
<ul class="org-ul">
<li>Quiz 1: Best timeline for modeling → Large observation, small prediction window.</li>
<li>Quiz 2: Most useful model → Small observation, large prediction window (ideal but difficult).</li>
</ul>
</div>
</div>

<div id="outline-container-org8151c3d" class="outline-3">
<h3 id="org8151c3d"><span class="section-number-3">3.14.</span> Prediction Performance on Windows</h3>
<div class="outline-text-3" id="text-3-14">
<ul class="org-ul">
<li><b><b>Prediction Window</b></b>: Longer window reduces accuracy.</li>
<li><b><b>Observation Window</b></b>: Longer window improves model performance until plateau.</li>
</ul>
</div>
</div>

<div id="outline-container-org63be694" class="outline-3">
<h3 id="org63be694"><span class="section-number-3">3.15.</span> Feature Selection</h3>
<div class="outline-text-3" id="text-3-15">
<ul class="org-ul">
<li>Identifies relevant features from EHR data.</li>
<li>EHR provides a large number of potential features (e.g., demographics, vitals).</li>
<li>Different targets require different feature subsets.</li>
</ul>
</div>
</div>

<div id="outline-container-org91112b1" class="outline-3">
<h3 id="org91112b1"><span class="section-number-3">3.16.</span> Predictive Model</h3>
<div class="outline-text-3" id="text-3-16">
<ul class="org-ul">
<li>Maps input features to predicted outcomes.</li>
<li>Regression models for continuous targets (e.g., healthcare costs).</li>
<li>Classification models for categorical targets (e.g., heart failure).</li>
<li>Common methods: Logistic regression, decision trees, random forests.</li>
</ul>
</div>
</div>

<div id="outline-container-org4499835" class="outline-3">
<h3 id="org4499835"><span class="section-number-3">3.17.</span> Performance Evaluation</h3>
<div class="outline-text-3" id="text-3-17">
<ul class="org-ul">
<li>Key metric: Testing error (not training error).</li>
<li>Models must generalize to unseen data.</li>
</ul>
</div>
</div>

<div id="outline-container-org1ae6cc0" class="outline-3">
<h3 id="org1ae6cc0"><span class="section-number-3">3.18.</span> Cross-Validation</h3>
<div class="outline-text-3" id="text-3-18">
<ul class="org-ul">
<li>Splits data into training and validation sets.</li>
<li>Common methods:
<ul class="org-ul">
<li>Leave-One-Out Cross-Validation.</li>
<li>K-Fold Cross-Validation.</li>
<li>Randomized Cross-Validation.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orga3e12cb" class="outline-3">
<h3 id="orga3e12cb"><span class="section-number-3">3.19.</span> Conclusion</h3>
<div class="outline-text-3" id="text-3-19">
<ul class="org-ul">
<li>Summary of the predictive modeling pipeline:
<ul class="org-ul">
<li>Define the prediction target.</li>
<li>Construct the patient cohort.</li>
<li>Generate relevant features.</li>
<li>Select important features.</li>
<li>Train the predictive model.</li>
<li>Evaluate model performance.</li>
</ul></li>
<li>Goal: Design high-level predictive modeling studies using EHR data.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org2f96732" class="outline-2">
<h2 id="org2f96732"><span class="section-number-2">4.</span> MapReduce</h2>
<div class="outline-text-2" id="text-4">
</div>
<div id="outline-container-orgec633d6" class="outline-3">
<h3 id="orgec633d6"><span class="section-number-3">4.1.</span> Introduction to MapReduce</h3>
<div class="outline-text-3" id="text-4-1">
<ul class="org-ul">
<li>Overview of MapReduce as a big data processing tool.</li>
<li>Utilizes distributed computation and storage.</li>
<li>Covers:
<ul class="org-ul">
<li>What MapReduce is</li>
<li>Fault tolerance in distributed environments</li>
<li>Analytical use cases and limitations</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc802c8a" class="outline-3">
<h3 id="orgc802c8a"><span class="section-number-3">4.2.</span> What is MapReduce</h3>
<div class="outline-text-3" id="text-4-2">
<ul class="org-ul">
<li>Hadoop/MapReduce is:
<ul class="org-ul">
<li>A programming model for parallel computation</li>
<li>An execution environment (Hadoop with HDFS)</li>
<li>A software package with various tools</li>
</ul></li>
<li>Capabilities:
<ul class="org-ul">
<li>Distributed storage via HDFS</li>
<li>Distributed computation via MapReduce</li>
<li>Built-in fault tolerance</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org5ac41bc" class="outline-3">
<h3 id="org5ac41bc"><span class="section-number-3">4.3.</span> Computational Process</h3>
<div class="outline-text-3" id="text-4-3">
<ul class="org-ul">
<li>Originated at Google (2004), open-sourced via Apache Hadoop.</li>
<li>Java-based platform.</li>
<li>Programming constrained to Map and Reduce functions for scalability.</li>
<li>Designed for parallel, fault-tolerant processing.</li>
<li>Emphasis on mastering computational patterns for analytics.</li>
</ul>
</div>
</div>

<div id="outline-container-orgad47f56" class="outline-3">
<h3 id="orgad47f56"><span class="section-number-3">4.4.</span> Learning Via Aggregation Statistics</h3>
<div class="outline-text-3" id="text-4-4">
<ul class="org-ul">
<li>Core design: express ML algorithms as aggregation tasks.</li>
<li>Example: Heart failure risk factor frequency analysis.</li>
<li>Map: extract risk factors per patient.</li>
<li>Reduce: aggregate frequencies across population.</li>
<li>Leads into deeper abstraction needs and trade-offs.</li>
</ul>
</div>
</div>

<div id="outline-container-org5afb42c" class="outline-3">
<h3 id="org5afb42c"><span class="section-number-3">4.5.</span> MapReduce Abstraction</h3>
<div class="outline-text-3" id="text-4-5">
<ul class="org-ul">
<li>Example task: count disease cases from patient records.</li>
<li>Map:
<ul class="org-ul">
<li>Emit (disease, 1) for each mention</li>
</ul></li>
<li>Shuffle:
<ul class="org-ul">
<li>Group by disease</li>
</ul></li>
<li>Reduce:
<ul class="org-ul">
<li>Sum values for each disease</li>
</ul></li>
<li>Emphasizes two-phase logic for scalability.</li>
</ul>
</div>
</div>

<div id="outline-container-orga6bc0ac" class="outline-3">
<h3 id="orga6bc0ac"><span class="section-number-3">4.6.</span> MapReduce System</h3>
<div class="outline-text-3" id="text-4-6">
<ul class="org-ul">
<li>Real-world data is too large for single machines.</li>
<li>Data is partitioned and processed by multiple mappers.</li>
<li>Intermediate results are shuffled and passed to reducers.</li>
<li>Final results computed from reduce function.</li>
<li>Three stages:
<ul class="org-ul">
<li>Map</li>
<li>Shuffle</li>
<li>Reduce</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org94f5bbc" class="outline-3">
<h3 id="org94f5bbc"><span class="section-number-3">4.7.</span> MapReduce Fault Recovery</h3>
<div class="outline-text-3" id="text-4-7">
<ul class="org-ul">
<li>Fault tolerance is a built-in system feature.</li>
<li>On failure:
<ul class="org-ul">
<li>Mappers/reducers are restarted with minimal recomputation.</li>
</ul></li>
<li>Goal: Only failed components are recomputed.</li>
<li>System handles all failure recovery.</li>
</ul>
</div>
</div>

<div id="outline-container-orga41d770" class="outline-3">
<h3 id="orga41d770"><span class="section-number-3">4.8.</span> Distributed File Systems</h3>
<div class="outline-text-3" id="text-4-8">
<ul class="org-ul">
<li>HDFS: Hadoop's storage system.</li>
<li>Splits large files into partitions.</li>
<li>Partitions stored on multiple machines with redundancy.</li>
<li>Benefits:
<ul class="org-ul">
<li>Faster concurrent access</li>
<li>Fault tolerance (recover from worker failures)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc48e2e9" class="outline-3">
<h3 id="orgc48e2e9"><span class="section-number-3">4.9.</span> MapReduce Design Choice</h3>
<div class="outline-text-3" id="text-4-9">
<ul class="org-ul">
<li>Design principle: minimal functionality for reliability and scalability.</li>
<li>Restricted computation model (e.g., aggregation queries).</li>
<li>Map: operate on individual records</li>
<li>Reduce: aggregate results</li>
<li>Supports straggler mitigation (slow mappers duplicated)</li>
</ul>
</div>
</div>

<div id="outline-container-org186f293" class="outline-3">
<h3 id="org186f293"><span class="section-number-3">4.10.</span> Analytics with MapReduce</h3>
<div class="outline-text-3" id="text-4-10">
<ul class="org-ul">
<li>Application examples:
<ul class="org-ul">
<li>K-Nearest Neighbors (KNN)</li>
<li>Linear Regression</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org65040c6" class="outline-3">
<h3 id="org65040c6"><span class="section-number-3">4.11.</span> MapReduce KNN</h3>
<div class="outline-text-3" id="text-4-11">
<ul class="org-ul">
<li>KNN implementation:
<ul class="org-ul">
<li>Map:
<ul class="org-ul">
<li>Find K nearest neighbors per partition</li>
</ul></li>
<li>Reduce:
<ul class="org-ul">
<li>Combine local results to find global nearest neighbors</li>
</ul></li>
</ul></li>
<li>Patient data partitioned, processed in parallel</li>
</ul>
</div>
</div>

<div id="outline-container-org381f569" class="outline-3">
<h3 id="org381f569"><span class="section-number-3">4.12.</span> Linear Regression</h3>
<div class="outline-text-3" id="text-4-12">
<ul class="org-ul">
<li>Goal: map patient features to heart disease risk</li>
<li>Normal equation:
<ul class="org-ul">
<li>β = (XᵗX)⁻¹Xᵗy</li>
</ul></li>
<li>MapReduce:
<ul class="org-ul">
<li>Map f1: compute xi * xiᵗ</li>
<li>Map f2: compute xi * yi</li>
<li>Reduce: aggregate sums</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org82dbd04" class="outline-3">
<h3 id="org82dbd04"><span class="section-number-3">4.13.</span> MapReduce for Linear Regression Quiz Question</h3>
<div class="outline-text-3" id="text-4-13">
<ul class="org-ul">
<li>Quiz: specify map/reduce pseudo code for computing Xᵗy</li>
</ul>
</div>
</div>

<div id="outline-container-org8d7bc82" class="outline-3">
<h3 id="org8d7bc82"><span class="section-number-3">4.14.</span> MapReduce for Linear Regression Quiz Solution</h3>
<div class="outline-text-3" id="text-4-14">
<ul class="org-ul">
<li>Solution:
<ul class="org-ul">
<li>Map: compute xi * yi</li>
<li>Reduce: aggregate results</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org72f809c" class="outline-3">
<h3 id="org72f809c"><span class="section-number-3">4.15.</span> Limitations of MapReduce</h3>
<div class="outline-text-3" id="text-4-15">
<ul class="org-ul">
<li>Example: Logistic Regression via Gradient Descent</li>
<li>Challenge: Requires iterative computation
<ul class="org-ul">
<li>Each iteration loads data twice</li>
</ul></li>
<li>MapReduce not efficient for:
<ul class="org-ul">
<li>Iterative, multi-stage computation</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org2afd447" class="outline-3">
<h3 id="org2afd447"><span class="section-number-3">4.16.</span> MapReduce Summary Quiz Question</h3>
<div class="outline-text-3" id="text-4-16">
<ul class="org-ul">
<li>Quiz on ideal conditions for MapReduce:
<ul class="org-ul">
<li>Single vs. multiple passes</li>
<li>Skewed vs. uniform key distribution</li>
<li>Synchronization needs</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org82c1c18" class="outline-3">
<h3 id="org82c1c18"><span class="section-number-3">4.17.</span> MapReduce Summary Quiz Solution</h3>
<div class="outline-text-3" id="text-4-17">
<ul class="org-ul">
<li>Best suited for:
<ul class="org-ul">
<li>Single-pass jobs (e.g., histograms)</li>
<li>Both skewed and uniform key distributions</li>
<li>Minimal synchronization (only between Map and Reduce phases)</li>
</ul></li>
</ul>
</div>
</div>
</div>
<div id="outline-container-orgd9e0d21" class="outline-2">
<h2 id="orgd9e0d21"><span class="section-number-2">5.</span> Classification Model Metrics</h2>
<div class="outline-text-2" id="text-5">
</div>
<div id="outline-container-org110e43c" class="outline-3">
<h3 id="org110e43c"><span class="section-number-3">5.1.</span> Predictive Model review</h3>
<div class="outline-text-3" id="text-5-1">
<dl class="org-dl">
<dt>Predictive model</dt><dd>mapping function between model inputs and outputs (pred's)</dd>
<dt>Model metrics</dt><dd>needed to know how well they're performing</dd>
</dl>
</div>
</div>
<div id="outline-container-org3b07bc6" class="outline-3">
<h3 id="org3b07bc6"><span class="section-number-3">5.2.</span> Confusion matrix</h3>
<div class="outline-text-3" id="text-5-2">
<ul class="org-ul">
<li>True positive/negative also known as "Condition" positive/negative</li>
<li>Predicted positive/negative also known as "Prediction Outcome" positive/negative</li>
</ul>
</div>
</div>
<div id="outline-container-org0f5b4fc" class="outline-3">
<h3 id="org0f5b4fc"><span class="section-number-3">5.3.</span> Accuracy metrics</h3>
<div class="outline-text-3" id="text-5-3">
<p>
All divided by ground truth values.
</p>
<dl class="org-dl">
<dt>Accuracy</dt><dd>(TP+TN)/TP. Not most useful for imbalanced class</dd>
<dt>True positive rate</dt><dd>TP/CP (Sensitivity, Recall)</dd>
<dt>False positive rate</dt><dd>FP/CN</dd>
<dt>False negative</dt><dd>FN/CP</dd>
<dt>True negative rate</dt><dd>TN/CN (Specificity)</dd>
</dl>
</div>
<div id="outline-container-orgd5d5725" class="outline-4">
<h4 id="orgd5d5725"><span class="section-number-4">5.3.1.</span> Other notes</h4>
<div class="outline-text-4" id="text-5-3-1">
<ul class="org-ul">
<li>FP is a type I error.</li>
<li>FN is a type II error.</li>
<li>Hard to perform well on all metrics</li>
<li>Important to choose the right metrics</li>
</ul>
</div>
</div>
</div>
<div id="outline-container-orgc06e628" class="outline-3">
<h3 id="orgc06e628"><span class="section-number-3">5.4.</span> Predictive metrics</h3>
<div class="outline-text-3" id="text-5-4">
<p>
All divided by prediction outcomes.
</p>
<dl class="org-dl">
<dt>Prevalence</dt><dd>CP/Total population. How likely disease occurs in population</dd>
<dt>Positive predictive value (Precision)</dt><dd>TP/Pred outcome positive.</dd>
<dt>False discovery rate</dt><dd>FP/Pred outcome positive</dd>
<dt>Negative predictive value</dt><dd>TN/Pred outcome negative</dd>
<dt>False omission rate</dt><dd>FN/Pred outcome negative</dd>
</dl>
</div>
</div>
<div id="outline-container-org0aebabe" class="outline-3">
<h3 id="org0aebabe"><span class="section-number-3">5.5.</span> F1 score</h3>
<div class="outline-text-3" id="text-5-5">
<p>
Harmonic mean of PPV and TPR
\[
F_1 = 2 \times \frac{PPV \times TPR}{PPV + TPR}
\]
</p>
</div>
</div>
<div id="outline-container-org463125f" class="outline-3">
<h3 id="org463125f"><span class="section-number-3">5.6.</span> Classifier quiz</h3>
<div class="outline-text-3" id="text-5-6">
<p>
Which is the best classifier?
</p>
<ul class="org-ul">
<li>Highest F1, PPV and Accuracy is the best classifier</li>
</ul>
</div>
</div>
<div id="outline-container-orgea083e3" class="outline-3">
<h3 id="orgea083e3"><span class="section-number-3">5.7.</span> Reversing predictions</h3>
<div class="outline-text-3" id="text-5-7">
<p>
It's always possible to reverse the predictions so 0.21 might perform better than 0.69/0.50.
</p>
</div>
</div>
<div id="outline-container-org9971d22" class="outline-3">
<h3 id="org9971d22"><span class="section-number-3">5.8.</span> Receiver operating characteristic</h3>
<div class="outline-text-3" id="text-5-8">
<ul class="org-ul">
<li>Predictive models generally output continuous score.</li>
<li>Threshold is needed as precision bound to force to a certain category</li>
<li>ROC provides a way to compare performance of classifiers as the decision boundary is varied</li>
<li>ROC curve is the plot of TP rate vs FP rate at various threshold values
<ul class="org-ul">
<li>We sort by prediction score (highest first)</li>
<li>Use prediction score as threshold values</li>
<li>Plot on the chart and see how many are misclassified (needs True value to be known)</li>
</ul></li>
<li>AUROC does not depend on the choice of threshold.</li>
<li>AUROC is the most popular metric for classification</li>
</ul>
</div>
</div>
<div id="outline-container-org50d2663" class="outline-3">
<h3 id="org50d2663"><span class="section-number-3">5.9.</span> "Best classifier threshold" quiz</h3>
<div class="outline-text-3" id="text-5-9">
<ul class="org-ul">
<li>There isn't a standard answer, it depends on whether TP, TN or other metric is prioritized</li>
</ul>
</div>
</div>
<div id="outline-container-org68a7577" class="outline-3">
<h3 id="org68a7577"><span class="section-number-3">5.10.</span> Regression metrics</h3>
<div class="outline-text-3" id="text-5-10">
<dl class="org-dl">
<dt>MAE</dt><dd>average of absolute errors, harder to work with since the absolute value is not differentiable</dd>
<dt>MSE</dt><dd>average of squared errors, easier to work with as the derivative of squared term is linear. Increases a lot faster than MAE</dd>
<dt>\(R^2\)</dt><dd>bounded by \((-\infty,1)\). AKA coefficient of determination.</dd>
</dl>
</div>
<div id="outline-container-org74475fe" class="outline-4">
<h4 id="org74475fe"><span class="section-number-4">5.10.1.</span> \(R^2\)</h4>
<div class="outline-text-4" id="text-5-10-1">
<p>
\[
R^2 = 1-\frac{\sum_i (y_i - \hat{y}_i)^2}{\sum_i(y_i - \bar{y})^2}
\]
i.e. 1-MSE/Variance
</p>

<p>
Negative \(R^2\) values means they perform worse than a simple average of raw data.
As noise increases, \(R^2\) decreases.
</p>
</div>
</div>
</div>
</div>
<div id="outline-container-orgec4539d" class="outline-2">
<h2 id="orgec4539d"><span class="section-number-2">6.</span> Ensemble methods</h2>
<div class="outline-text-2" id="text-6">
</div>
<div id="outline-container-org6f98068" class="outline-3">
<h3 id="org6f98068"><span class="section-number-3">6.1.</span> Gradient Descent Method (GDM) for Linear Regression</h3>
<div class="outline-text-3" id="text-6-1">
<ul class="org-ul">
<li>Illustrates gradient descent using linear regression.</li>
<li>Dataset: each row = patient; features in X, outcome Y (e.g., hospital cost).</li>
<li>Objective: learn linear mapping from features X to outcome Y.</li>
<li>Assumes Gaussian distribution → log-likelihood leads to squared error minimization.</li>
<li>Gradient: derived by taking derivative of log-likelihood w.r.t. coefficients β.</li>
<li>Update rule: move β in direction of gradient with step size η.</li>
<li><span style='background-color: #FFFF00;'>Requires multiple iterations over full dataset → computationally expensive.</span></li>
</ul>
</div>
</div>

<div id="outline-container-orgc55183d" class="outline-3">
<h3 id="orgc55183d"><span class="section-number-3">6.2.</span> Stochastic Gradient Descent (SGD) Method</h3>
<div class="outline-text-3" id="text-6-2">
<ul class="org-ul">
<li>SGD is a scalable variant of gradient descent for large datasets.</li>
<li>Instead of using full dataset, <span style='background-color: #FFFF00;'>computes gradients on random subsets (mini-batches)</span>.</li>
<li>One-point update = SGD; larger batches = mini-batch gradient descent.</li>
<li>Faster per-iteration updates compared to full batch gradient descent.</li>
</ul>
</div>
</div>

<div id="outline-container-org3a65c15" class="outline-3">
<h3 id="org3a65c15"><span class="section-number-3">6.3.</span> SGD for Linear Regression</h3>
<div class="outline-text-3" id="text-6-3">
<ul class="org-ul">
<li>Applies SGD to same regression problem.</li>
<li>Each update uses one patient:
<ul class="org-ul">
<li>Compute individual log-likelihood and gradient.</li>
<li>Update β using gradient.</li>
</ul></li>
<li>More efficient than full gradient computation.</li>
<li>Key idea: updates based on single or small group of samples.</li>
</ul>
</div>
</div>

<div id="outline-container-org59c0e82" class="outline-3">
<h3 id="org59c0e82"><span class="section-number-3">6.4.</span> Ensemble Method Pt 1</h3>
<div class="outline-text-3" id="text-6-4">
<ul class="org-ul">
<li>Ensemble methods combine multiple models for improved performance.</li>
<li>Can use same (e.g., Random Forest) or different base models.</li>
<li>Real-world example: Netflix Prize → ensemble models won.</li>
<li>Many teams merged, forming more powerful ensembles.</li>
</ul>
</div>
</div>

<div id="outline-container-orga843412" class="outline-3">
<h3 id="orga843412"><span class="section-number-3">6.5.</span> Ensemble Method Pt 2</h3>
<div class="outline-text-3" id="text-6-5">
<ul class="org-ul">
<li>General ensemble steps:
<ol class="org-ol">
<li>Generate multiple datasets (via bagging or boosting).</li>
<li>Train separate models on each.</li>
<li>Aggregate outputs using functions like averaging or weighted sum.</li>
</ol></li>
<li>Outcome: different ensemble strategies.</li>
</ul>
</div>
</div>

<div id="outline-container-org676a854" class="outline-3">
<h3 id="org676a854"><span class="section-number-3">6.6.</span> Bias Variance Tradeoff</h3>
<div class="outline-text-3" id="text-6-6">
<ul class="org-ul">
<li>Key insight: <span style='background-color: #FFFF00;'>model error = bias² + variance</span>.</li>
<li>Bias: error from wrong assumptions (e.g., assuming linear when it's not).</li>
<li>Variance: error from sensitivity to training data.</li>
<li>Ideal model: low bias and low variance.</li>
<li>Graphical example: dartboard showing different combinations.
<img src="./bias-variance-tradeoff.png" alt="bias-variance-tradeoff.png" /></li>
<li>As model complexity increases:
<ul class="org-ul">
<li>Bias ↓</li>
<li>Variance ↑</li>
<li>Total error = minimum at an optimal complexity.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org4e04bd6" class="outline-3">
<h3 id="org4e04bd6"><span class="section-number-3">6.7.</span> Bias Variance Tradeoff Quiz Question</h3>
<div class="outline-text-3" id="text-6-7">

<div id="orge2dfd4c" class="figure">
<p><img src="./bias-var-quiz.png" alt="bias-var-quiz.png" />
</p>
</div>
<ul class="org-ul">
<li>Task: rank four models (flat, linear, smooth curve, wiggly) by complexity.</li>
<li>A → D: D (flat) &lt; A (linear) &lt; B (curve) &lt; C (wiggly)</li>
</ul>
</div>
</div>

<div id="outline-container-org2e7163d" class="outline-3">
<h3 id="org2e7163d"><span class="section-number-3">6.8.</span> Bias Variance Tradeoff Quiz Solution</h3>
<div class="outline-text-3" id="text-6-8">
<ul class="org-ul">
<li>Explanation:
<ul class="org-ul">
<li>D: lowest complexity (flat line).</li>
<li>A: next, simple linear model.</li>
<li>B: more complex curve.</li>
<li>C: most complex (wiggly).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org94349da" class="outline-3">
<h3 id="org94349da"><span class="section-number-3">6.9.</span> Bias Variance Tradeoff Quiz 2 Question</h3>
<div class="outline-text-3" id="text-6-9">
<ul class="org-ul">
<li>Given same models, asked which best fits data.</li>
</ul>
</div>
</div>

<div id="outline-container-orga7ba58b" class="outline-3">
<h3 id="orga7ba58b"><span class="section-number-3">6.10.</span> Bias Variance Tradeoff Quiz 2 Solution</h3>
<div class="outline-text-3" id="text-6-10">
<ul class="org-ul">
<li>Answer: B (smooth curve)</li>
<li>Balances bias and variance well.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc2ca3f8" class="outline-3">
<h3 id="orgc2ca3f8"><span class="section-number-3">6.11.</span> Bagging</h3>
<div class="outline-text-3" id="text-6-11">
<ul class="org-ul">
<li>Bagging = <span style='background-color: #FFFF00;'>Bootstrap Aggregation</span>.</li>
<li>Method:
<ul class="org-ul">
<li>Sample data <span style='background-color: #FFFF00;'>with replacement</span>.</li>
<li>Train models on each sample.</li>
<li>Combine predictions via majority vote or averaging.</li>
</ul></li>
<li>Reduces variance of the final model.</li>
</ul>
</div>
</div>

<div id="outline-container-org0456c0f" class="outline-3">
<h3 id="org0456c0f"><span class="section-number-3">6.12.</span> Random Forest</h3>
<div class="outline-text-3" id="text-6-12">
<ul class="org-ul">
<li><span style='background-color: #FFFF00;'>Special case of bagging</span> with decision trees.
<img src="./rf1.png" alt="rf1.png" />
<img src="./rf2.png" alt="rf2.png" /></li>
<li>Steps:
<ul class="org-ul">
<li>Randomly sample both rows (patients) and columns (features).</li>
<li>Train simple decision trees on these subsets.</li>
</ul></li>
<li>Final prediction = average of all trees.</li>
<li>Simple trees help maintain diversity and lower computation.</li>
</ul>
</div>
</div>

<div id="outline-container-org82fbbc2" class="outline-3">
<h3 id="org82fbbc2"><span class="section-number-3">6.13.</span> Why Bagging Works</h3>
<div class="outline-text-3" id="text-6-13">
<ul class="org-ul">
<li><span style='background-color: #FFFF00;'>Reduces variance without increasing bias.</span></li>
<li>Variance reduction from averaging independent model predictions.</li>
<li>Based on statistical principle: variance of mean = variance / \(T\) for \(T\) models, assuming
the resulting bootstrap samples are close to i.i.d.</li>
</ul>
</div>
</div>

<div id="outline-container-org4c65cab" class="outline-3">
<h3 id="org4c65cab"><span class="section-number-3">6.14.</span> Boosting</h3>
<div class="outline-text-3" id="text-6-14">
<ul class="org-ul">
<li>Builds models sequentially.</li>
<li>Each new model focuses on mistakes of previous ones.</li>
<li>Final model = weighted combination of all models.</li>
<li>Pros: better accuracy.</li>
<li>Cons: <span style='background-color: #FFFF00;'>prone to overfitting</span> vs bagging.</li>
<li>Example: AdaBoost is most popular.</li>
</ul>
</div>
</div>

<div id="outline-container-org4aa4a97" class="outline-3">
<h3 id="org4aa4a97"><span class="section-number-3">6.15.</span> Bagging vs Boosting Quiz Question</h3>
<div class="outline-text-3" id="text-6-15">
<ul class="org-ul">
<li>Quiz to compare characteristics:
<ul class="org-ul">
<li>Combination method (simple vs weighted average)</li>
<li>Parallelism</li>
<li>Noise sensitivity</li>
<li>Accuracy</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org75a8ba8" class="outline-3">
<h3 id="org75a8ba8"><span class="section-number-3">6.16.</span> Bagging vs Boosting Quiz Solution</h3>
<div class="outline-text-3" id="text-6-16">
<ul class="org-ul">
<li>Bagging:
<ul class="org-ul">
<li>Simple average, parallel-friendly, less sensitive to noise, reliable accuracy (good in all cases).</li>
</ul></li>
<li>Boosting:
<ul class="org-ul">
<li>Weighted average, sequential (hard to parallelize), sensitive to noise, potentially higher accuracy but less reliable.l</li>
</ul></li>
<li>Analogy:
<ul class="org-ul">
<li><span style='background-color: #FFFF00;'>Bagging = reliable Japanese car.</span></li>
<li><span style='background-color: #FFFF00;'>Boosting = high-performance but risky sports car.</span></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org53023b8" class="outline-3">
<h3 id="org53023b8"><span class="section-number-3">6.17.</span> Summary for Ensemble Methods</h3>
<div class="outline-text-3" id="text-6-17">
<ul class="org-ul">
<li>Pros:
<ul class="org-ul">
<li>Simple, flexible, few parameters, theoretical backing.</li>
</ul></li>
<li>Cons:
<ul class="org-ul">
<li>Computational cost (training + inference).</li>
<li>Harder interpretation (due to model complexity).</li>
</ul></li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org2107e1f" class="outline-2">
<h2 id="org2107e1f"><span class="section-number-2">7.</span> Computational Phenotyping</h2>
<div class="outline-text-2" id="text-7">
</div>
<div id="outline-container-org3ddbe4b" class="outline-3">
<h3 id="org3ddbe4b"><span class="section-number-3">7.1.</span> Introduction to Phenotyping</h3>
<div class="outline-text-3" id="text-7-1">
<ul class="org-ul">
<li>Introduction to clustering in healthcare, specifically "phenotyping".</li>
<li>Phenotypes refer to diseases or conditions; known phenotypes exist, but many more are undiscovered.</li>
<li>Computational phenotyping helps discover novel phenotypes using available data.</li>
<li>Applications include disease diagnosis, cost prediction, readmission risk, and genomic studies.</li>
</ul>
</div>
</div>

<div id="outline-container-org91fcee2" class="outline-3">
<h3 id="org91fcee2"><span class="section-number-3">7.2.</span> Computational Phenotyping</h3>
<div class="outline-text-3" id="text-7-2">
<ul class="org-ul">
<li>Converts raw EHR data into meaningful phenotypes using algorithms.</li>
<li>EHR data includes demographics, diagnosis codes, medications, procedures, labs, notes.</li>
<li>Challenges:
<ul class="org-ul">
<li>Noisy and incomplete data esp. raw data.</li>
<li>Primarily designed for clinical use, not directly for supporting research.</li>
<li>Redundancy across data sources.</li>
</ul></li>
<li>Goal: derive research-grade phenotypes from raw clinical data.</li>
</ul>
</div>
</div>

<div id="outline-container-org9894fe5" class="outline-3">
<h3 id="org9894fe5"><span class="section-number-3">7.3.</span> Phenotyping Algorithm</h3>
<div class="outline-text-3" id="text-7-3">
<ul class="org-ul">
<li>Example: Type 2 diabetes phenotyping algorithm.</li>
<li>Sequential logic:
<ol class="org-ol">
<li>Exclude Type 1 diagnosis.</li>
<li>Check for Type 2 diagnosis.</li>
<li>Check medication and lab results.</li>
</ol></li>
<li>Multiple confirmation paths possible.</li>
<li>Developed manually by clinical experts.</li>
</ul>
</div>
</div>

<div id="outline-container-orgafd5819" class="outline-3">
<h3 id="orgafd5819"><span class="section-number-3">7.4.</span> Applications of Phenotyping</h3>
<div class="outline-text-3" id="text-7-4">
<ul class="org-ul">
<li>Genomic studies: link genomic and phenotypic data.</li>
<li>Clinical predictive modeling: forecast disease onset, hospitalizations.</li>
<li>Pragmatic clinical trials: assess treatment effectiveness in real-world settings.</li>
<li>Healthcare quality measurement: compare care quality across institutions.</li>
<li>All rely on robust phenotyping algorithms.</li>
</ul>
</div>
</div>

<div id="outline-container-orgdf4d0ad" class="outline-3">
<h3 id="orgdf4d0ad"><span class="section-number-3">7.5.</span> Genomic Wide Association Study</h3>
<div class="outline-text-3" id="text-7-5">
<ul class="org-ul">
<li>GWAS scans SNPs to associate genetic variations with diseases.</li>
<li>Steps:
<ol class="org-ol">
<li>Identify phenotypes.</li>
<li>Divide subjects into cases/controls.</li>
<li>Collect DNA and scan for SNPs.</li>
<li>Calculate odds ratios and p-values.</li>
</ol></li>
<li>Statistically significant SNPs may indicate disease relevance.</li>
<li>High-quality phenotyping critical for accurate GWAS.</li>
</ul>
</div>
</div>

<div id="outline-container-org4bccaac" class="outline-3">
<h3 id="org4bccaac"><span class="section-number-3">7.6.</span> Why Do We Care About Phenotyping</h3>
<div class="outline-text-3" id="text-7-6">
<p>
<span style='background-color: #FFFF00;'>We need rich and deep phentotyping to analyze genomic data.</span>
</p>
<ul class="org-ul">
<li>Genomic data is rapidly becoming cheaper and more abundant.</li>
<li>Phenotypic data remains complex and costly to acquire.</li>
<li>Demand for better phenotyping algorithms to support scalable, high-quality genomic research.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfd29f72" class="outline-3">
<h3 id="orgfd29f72"><span class="section-number-3">7.7.</span> Clinical Predictive Modeling</h3>
<div class="outline-text-3" id="text-7-7">
<ul class="org-ul">
<li>Predictive modeling using raw EHR data is problematic (noise, complexity, lack of portability).</li>
<li>Phenotyping transforms raw data into simpler, meaningful concepts.</li>
<li>Benefits:
<ul class="org-ul">
<li>Better model accuracy.</li>
<li>Easier interpretation.</li>
<li>Generalizability across hospitals/settings (input data schemas may be different).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org4644e3f" class="outline-3">
<h3 id="org4644e3f"><span class="section-number-3">7.8.</span> Pragmatic Clinical Trials</h3>
<div class="outline-text-3" id="text-7-8">
<ul class="org-ul">
<li>Traditional vs. pragmatic trials:
<ul class="org-ul">
<li>Traditional: controlled, single condition/drug, strict selection, randomized.</li>
<li>Pragmatic: <span style='background-color: #FFFF00;'>real-world</span>, <span style='background-color: #FFFF00;'>multiple</span> conditions/drugs, minimal selection, often <span style='background-color: #FFFF00;'>not randomized</span>.</li>
</ul></li>
<li>High-quality phenotyping essential for identifying patient conditions and treatments in pragmatic settings.</li>
</ul>
</div>
</div>

<div id="outline-container-org7533f0c" class="outline-3">
<h3 id="org7533f0c"><span class="section-number-3">7.9.</span> Healthcare Quality Measurement</h3>
<div class="outline-text-3" id="text-7-9">
<ul class="org-ul">
<li>Challenge: hospitals use <span style='background-color: #FFFF00;'>diverse EHR formats</span>.</li>
<li>Centralized processing of raw data is complex.</li>
<li>Solution: local phenotyping at hospitals → send standardized phenotype data to central site.</li>
<li>Enables <span style='background-color: #FFFF00;'>scalable and fair</span> quality comparison across institutions.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcb20273" class="outline-3">
<h3 id="orgcb20273"><span class="section-number-3">7.10.</span> Phenotyping Methods Part 1</h3>
<div class="outline-text-3" id="text-7-10">
<ul class="org-ul">
<li>Two main approaches:
<ul class="org-ul">
<li>Supervised learning: uses labeled data (function <span style='background-color: #FFFF00;'>approximation</span>).</li>
<li>Unsupervised learning: identifies patterns/structures without labels (<span style='background-color: #FFFF00;'>description</span>/summarization).</li>
</ul></li>
<li>Illustrative examples and playful dialogue explore these ideas in depth.</li>
</ul>
</div>
</div>

<div id="outline-container-org110cb00" class="outline-3">
<h3 id="org110cb00"><span class="section-number-3">7.11.</span> Phenotyping Methods Part 2</h3>
<div class="outline-text-3" id="text-7-11">
<ul class="org-ul">
<li>Supervised: expert-defined rules (Boolean logic, decision trees), or machine learning classifiers.
<ul class="org-ul">
<li>Pros: interpretable, low validation effort.</li>
<li>Cons: high development <span style='background-color: #FFFF00;'>cost</span>, <span style='background-color: #FFFF00;'>not good for unknown phenotypes</span>, potential <span style='background-color: #FFFF00;'>poor transferability</span>.</li>
</ul></li>
<li>Unsupervised: clustering, dimensionality reduction.
<ul class="org-ul">
<li>Pros: less manual effort, no need for labels.</li>
<li>Cons: <span style='background-color: #FFFF00;'>hard to validate</span>, large data requirements.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org09c04e8" class="outline-3">
<h3 id="org09c04e8"><span class="section-number-3">7.12.</span> Phenotyping Quiz Question</h3>
<div class="outline-text-3" id="text-7-12">
<ul class="org-ul">
<li>Quiz:
<ul class="org-ul">
<li>Which approach requires more human effort during evaluation?</li>
<li>Which is easier to interpret?</li>
<li>Choices: expert-defined rules vs. classification models.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org3b90b46" class="outline-3">
<h3 id="org3b90b46"><span class="section-number-3">7.13.</span> Phenotyping Quiz Solution</h3>
<div class="outline-text-3" id="text-7-13">
<ul class="org-ul">
<li>Answer:
<ul class="org-ul">
<li>Classification models require more evaluation effort (need more labeled data).</li>
<li>Expert-defined rules are easier to interpret (clinician-derived, intuitive).</li>
</ul></li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org492c6a3" class="outline-2">
<h2 id="org492c6a3"><span class="section-number-2">8.</span> Clustering</h2>
<div class="outline-text-2" id="text-8">
</div>
<div id="outline-container-org97fee4c" class="outline-3">
<h3 id="org97fee4c"><span class="section-number-3">8.1.</span> Introduction to Clustering</h3>
<div class="outline-text-3" id="text-8-1">
<ul class="org-ul">
<li>Introduction of clustering as a method distinct from classification.</li>
<li>Classification: Grouping by known labels.</li>
<li>Clustering: Discovering groups from raw data.</li>
<li>Topics covered:
<ul class="org-ul">
<li>Definition of clustering.</li>
<li>Algorithms: K-means, Gaussian Mixture Models (GMM).</li>
<li>Scalable algorithms for large datasets.</li>
<li>Applications in healthcare.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org2d75ab7" class="outline-3">
<h3 id="org2d75ab7"><span class="section-number-3">8.2.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-8-2">
<ul class="org-ul">
<li>Clustering in healthcare:
<ul class="org-ul">
<li>Patient stratification: group patients with similar characteristics.</li>
<li>Disease hierarchy discovery: learn structure/relationships among diseases.</li>
<li>Phenotyping: convert raw data into meaningful phenotypes (clinical concepts).</li>
<li>Phenotypes = clusters of similar patients.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf7d6310" class="outline-3">
<h3 id="orgf7d6310"><span class="section-number-3">8.3.</span> What is Clustering</h3>
<div class="outline-text-3" id="text-8-3">
<ul class="org-ul">
<li>Clustering explained via patient-disease matrix:
<ul class="org-ul">
<li>Rows = patients, Columns = diseases.</li>
<li>Apply clustering to rows → patient clusters (e.g., P1, P2, P3).</li>
<li>Apply clustering to columns → disease clusters (e.g., D1, D2, D3).</li>
<li>Supports applications like phenotyping and disease discovery.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc8b2616" class="outline-3">
<h3 id="orgc8b2616"><span class="section-number-3">8.4.</span> Algorithm Overview</h3>
<div class="outline-text-3" id="text-8-4">
<ul class="org-ul">
<li>Two categories of clustering algorithms:
<ul class="org-ul">
<li>Classical: K-means, Hierarchical, Gaussian Mixture Model.</li>
<li>Scalable: Mini-batch K-means, DBSCAN.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orga113f24" class="outline-3">
<h3 id="orga113f24"><span class="section-number-3">8.5.</span> K Means</h3>
<div class="outline-text-3" id="text-8-5">
<ul class="org-ul">
<li>Inputs: data points X₁ to Xₙ, number of clusters K.</li>
<li>Outputs: K clusters S₁ to Sₖ.</li>
<li>Objective: Minimize distance of data points to cluster centers (µᵢ).</li>
<li>Algorithm steps:
<ol class="org-ol">
<li>Initialize K centers.</li>
<li>Assign points to nearest center.</li>
<li>Update centers based on mean of assigned points.</li>
<li>Repeat until convergence.</li>
</ol></li>
<li>Example illustrated using K=2.</li>
</ul>
</div>
</div>

<div id="outline-container-org8dc849b" class="outline-3">
<h3 id="org8dc849b"><span class="section-number-3">8.6.</span> K Means Quiz Question</h3>
<div class="outline-text-3" id="text-8-6">
<ul class="org-ul">
<li>Quiz: Given n, k, d (dimensions), i (iterations), what's the complexity?</li>
</ul>
</div>
</div>

<div id="outline-container-orgbaff805" class="outline-3">
<h3 id="orgbaff805"><span class="section-number-3">8.7.</span> K Means Quiz Solution</h3>
<div class="outline-text-3" id="text-8-7">
<ul class="org-ul">
<li>Complexity: O(n × k × d × i)
<ul class="org-ul">
<li>Cost driven by assignment and update operations.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8befffa" class="outline-3">
<h3 id="org8befffa"><span class="section-number-3">8.8.</span> Hierarchical Clustering</h3>
<div class="outline-text-3" id="text-8-8">
<ul class="org-ul">
<li>Builds a hierarchy of data points.</li>
<li>Two approaches:
<ul class="org-ul">
<li>Agglomerative (bottom-up): Merge smallest clusters until one remains.</li>
<li>Divisive (top-down): Split one big cluster into smaller ones.</li>
</ul></li>
<li>Agglomerative more commonly used.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd28ac28" class="outline-3">
<h3 id="orgd28ac28"><span class="section-number-3">8.9.</span> Agglomerative Clustering</h3>
<div class="outline-text-3" id="text-8-9">
<ul class="org-ul">
<li>Steps:
<ol class="org-ol">
<li>Compute n×n distance matrix.</li>
<li>Initialize each point as its own cluster.</li>
<li>Iteratively merge closest clusters and update distance matrix.</li>
<li>Stop when one cluster remains (produces a hierarchy).</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org81a71a7" class="outline-3">
<h3 id="org81a71a7"><span class="section-number-3">8.10.</span> Gaussian Mixture Model</h3>
<div class="outline-text-3" id="text-8-10">
<ul class="org-ul">
<li>Soft clustering: points belong to multiple clusters with probabilities.</li>
<li>Each cluster is a Gaussian (normal distribution).</li>
<li>Parameters: mean (µₖ), variance (σₖ), mixing coefficient (πₖ).</li>
<li>Goal: Estimate GMM parameters from data.</li>
</ul>
</div>
</div>

<div id="outline-container-org07d0f11" class="outline-3">
<h3 id="org07d0f11"><span class="section-number-3">8.11.</span> GMM Expectation Maximization</h3>
<div class="outline-text-3" id="text-8-11">
<ul class="org-ul">
<li>Uses Expectation-Maximization (EM) algorithm:
<ul class="org-ul">
<li>Initialize parameters.</li>
<li>E-step: Compute assignment scores (γₙₖ).</li>
<li>M-step: Update parameters using γₙₖ.</li>
<li>Iterate until convergence.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org9316cfe" class="outline-3">
<h3 id="org9316cfe"><span class="section-number-3">8.12.</span> GMM Steps</h3>
<div class="outline-text-3" id="text-8-12">
<ul class="org-ul">
<li>Details of EM:
<ul class="org-ul">
<li>Initialization: K-means can be used for better results.</li>
<li>E-step: Compute γₙₖ using cluster probabilities.</li>
<li>M-step:
<ul class="org-ul">
<li>Compute cluster size (Nₖ), mean (µₖ), covariance (σₖ), mixing coefficient (πₖ).</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgcf433fc" class="outline-3">
<h3 id="orgcf433fc"><span class="section-number-3">8.13.</span> GMM Visual Illustration</h3>
<div class="outline-text-3" id="text-8-13">
<ul class="org-ul">
<li>Visual demo of GMM clustering.</li>
<li>Iteratively updates Gaussians until convergence.</li>
<li>Final model reflects soft cluster assignments.</li>
</ul>
</div>
</div>

<div id="outline-container-org2354da2" class="outline-3">
<h3 id="org2354da2"><span class="section-number-3">8.14.</span> K Means Vs GMM</h3>
<div class="outline-text-3" id="text-8-14">
<ul class="org-ul">
<li>Comparison:
<ul class="org-ul">
<li>K-means: hard assignment, single center (µₖ).</li>
<li>GMM: soft assignment, parameters include µₖ, σₖ, πₖ.</li>
<li>Both use iterative updates.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc83d226" class="outline-3">
<h3 id="orgc83d226"><span class="section-number-3">8.15.</span> Mini Batch K Means</h3>
<div class="outline-text-3" id="text-8-15">
<ul class="org-ul">
<li>Scalable version of K-means.</li>
<li>Works with small batches (M) instead of full dataset.</li>
<li>Updates centers incrementally.</li>
<li>Reduces computational cost for large data.</li>
</ul>
</div>
</div>

<div id="outline-container-orge52ce88" class="outline-3">
<h3 id="orge52ce88"><span class="section-number-3">8.16.</span> Mini Batch K Means Quiz Question</h3>
<div class="outline-text-3" id="text-8-16">
<ul class="org-ul">
<li>Quiz: Given k, t (iterations), b (batch size), d (dimensionality), compute complexity.</li>
</ul>
</div>
</div>

<div id="outline-container-org04f9a7a" class="outline-3">
<h3 id="org04f9a7a"><span class="section-number-3">8.17.</span> Mini Batch K Means Quiz Solution</h3>
<div class="outline-text-3" id="text-8-17">
<ul class="org-ul">
<li>Complexity: O(t × b × k × d)</li>
</ul>
</div>
</div>

<div id="outline-container-org9a1c82c" class="outline-3">
<h3 id="org9a1c82c"><span class="section-number-3">8.18.</span> DBScan</h3>
<div class="outline-text-3" id="text-8-18">
<ul class="org-ul">
<li>Density-Based Spatial Clustering.</li>
<li>Clusters = dense areas, separated by sparse regions.</li>
<li>Can identify clusters of arbitrary shape.</li>
<li>Effective for noisy data.</li>
</ul>
</div>
</div>

<div id="outline-container-org87c5647" class="outline-3">
<h3 id="org87c5647"><span class="section-number-3">8.19.</span> DBScan Key Concepts</h3>
<div class="outline-text-3" id="text-8-19">
<ul class="org-ul">
<li>Density = number of points within ε-distance.</li>
<li>Key concepts:
<ul class="org-ul">
<li>Core points: high-density.</li>
<li>Border points: near core points.</li>
<li>Noise: not in dense region.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org67e8219" class="outline-3">
<h3 id="org67e8219"><span class="section-number-3">8.20.</span> DBScan Algorithm</h3>
<div class="outline-text-3" id="text-8-20">
<ul class="org-ul">
<li>Steps:
<ol class="org-ol">
<li>Build graph connecting core points to neighbors.</li>
<li>Identify connected components as clusters.</li>
<li>Remove clustered points and repeat.</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-orgb0b3c1a" class="outline-3">
<h3 id="orgb0b3c1a"><span class="section-number-3">8.21.</span> DBScan Example</h3>
<div class="outline-text-3" id="text-8-21">
<ul class="org-ul">
<li>Visual example: finds 6 clusters, some noise points remain.</li>
</ul>
</div>
</div>

<div id="outline-container-org901317f" class="outline-3">
<h3 id="org901317f"><span class="section-number-3">8.22.</span> DBScan Quiz Question</h3>
<div class="outline-text-3" id="text-8-22">
<ul class="org-ul">
<li>Quiz: How many clusters can a point belong to in DBSCAN?</li>
</ul>
</div>
</div>

<div id="outline-container-org9ffc0a2" class="outline-3">
<h3 id="org9ffc0a2"><span class="section-number-3">8.23.</span> DBScan Quiz Solution</h3>
<div class="outline-text-3" id="text-8-23">
<ul class="org-ul">
<li>Answer: 0 or 1 (noise = 0, core/border = 1)</li>
</ul>
</div>
</div>

<div id="outline-container-orgb259f02" class="outline-3">
<h3 id="orgb259f02"><span class="section-number-3">8.24.</span> Clustering Evaluation Metrics</h3>
<div class="outline-text-3" id="text-8-24">
<ul class="org-ul">
<li>Metrics:
<ul class="org-ul">
<li>Rand Index (RI)</li>
<li>Mutual Information (MI)</li>
<li>Silhouette Coefficient</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgcab3b37" class="outline-3">
<h3 id="orgcab3b37"><span class="section-number-3">8.25.</span> Rand Index</h3>
<div class="outline-text-3" id="text-8-25">
<ul class="org-ul">
<li>Measures agreement between predicted and true clusters.</li>
<li>RI = (number of correct pairs) / (total pairs)</li>
<li>Value between 0 (bad) and 1 (perfect).</li>
</ul>
</div>
</div>

<div id="outline-container-orgbad96b2" class="outline-3">
<h3 id="orgbad96b2"><span class="section-number-3">8.26.</span> Mutual Information</h3>
<div class="outline-text-3" id="text-8-26">
<ul class="org-ul">
<li>From information theory: measures shared info between predicted and true clusters.</li>
<li>Normalized Mutual Information (NMI) between 0 and 1.</li>
</ul>
</div>
</div>

<div id="outline-container-org4c3a617" class="outline-3">
<h3 id="org4c3a617"><span class="section-number-3">8.27.</span> Summary of RI and MI</h3>
<div class="outline-text-3" id="text-8-27">
<ul class="org-ul">
<li>Pros:
<ul class="org-ul">
<li>Bounded values, interpretable.</li>
<li>No assumptions about cluster shapes.</li>
</ul></li>
<li>Cons:
<ul class="org-ul">
<li>Require ground truth, which may be unavailable.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org5dcd43f" class="outline-3">
<h3 id="org5dcd43f"><span class="section-number-3">8.28.</span> Silhouette Coefficient</h3>
<div class="outline-text-3" id="text-8-28">
<ul class="org-ul">
<li>Measures how similar a point is to its own cluster vs. others.</li>
<li>Formula: (b - a) / max(a, b)</li>
<li>Does not require ground truth.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb6f4182" class="outline-3">
<h3 id="orgb6f4182"><span class="section-number-3">8.29.</span> Silhouette Coefficient Pros and Cons</h3>
<div class="outline-text-3" id="text-8-29">
<ul class="org-ul">
<li>Range: -1 (bad) to 1 (good), 0 = overlapping clusters.</li>
<li>Assumes spherical clusters → less effective for complex shapes (e.g. DBSCAN).</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orge3f9a30" class="outline-2">
<h2 id="orge3f9a30"><span class="section-number-2">9.</span> Spark</h2>
<div class="outline-text-2" id="text-9">
</div>
<div id="outline-container-orgae58bc4" class="outline-3">
<h3 id="orgae58bc4"><span class="section-number-3">9.1.</span> Introduction to Spark</h3>
<div class="outline-text-3" id="text-9-1">
<ul class="org-ul">
<li>Spark is introduced as an alternative to MapReduce, better suited for iterative workloads.</li>
<li>Leverages distributed memory across machines.</li>
<li>Core concept: Resilient Distributed Dataset (RDD).</li>
<li>Promotes efficiency for machine learning and healthcare applications.</li>
</ul>
</div>
</div>

<div id="outline-container-org4670b80" class="outline-3">
<h3 id="org4670b80"><span class="section-number-3">9.2.</span> Environment</h3>
<div class="outline-text-3" id="text-9-2">
<ul class="org-ul">
<li>Big data analytics typically run in data centers or cloud platforms like AWS, GCP, Azure.</li>
<li>Environment includes interconnected racks of servers.</li>
<li>Context set to understand the need for systems like Spark.</li>
</ul>
</div>
</div>

<div id="outline-container-org241e324" class="outline-3">
<h3 id="org241e324"><span class="section-number-3">9.3.</span> Motivation</h3>
<div class="outline-text-3" id="text-9-3">
<ul class="org-ul">
<li>Hadoop &amp; MapReduce explained with acyclic data flow using stable storage.</li>
<li>Strength: fault-tolerance via disk storage.</li>
<li>Weakness: inefficiency with iterative or interactive workloads (e.g. ML, graph analysis).</li>
<li>Need for in-memory computation system like Spark.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf1bad78" class="outline-3">
<h3 id="orgf1bad78"><span class="section-number-3">9.4.</span> Iteration in Map Reduce</h3>
<div class="outline-text-3" id="text-9-4">
<ul class="org-ul">
<li>Machine learning models need repeated computation.</li>
<li>Hadoop repeatedly reads/writes from disk, causing inefficiency.</li>
<li>Highlights the need for in-memory operations.</li>
</ul>
</div>
</div>

<div id="outline-container-org22f95a2" class="outline-3">
<h3 id="org22f95a2"><span class="section-number-3">9.5.</span> Workload Illustration</h3>
<div class="outline-text-3" id="text-9-5">
<ul class="org-ul">
<li>Iterative workloads: same computation on same data (e.g., model training).</li>
<li>Interactive workloads: varying computations on evolving data (e.g., data exploration).</li>
<li>Key: keep working sets in memory for speed.</li>
</ul>
</div>
</div>

<div id="outline-container-org4795d7b" class="outline-3">
<h3 id="org4795d7b"><span class="section-number-3">9.6.</span> Challenge</h3>
<div class="outline-text-3" id="text-9-6">
<ul class="org-ul">
<li>Challenge: memory is not fault-tolerant.</li>
<li>Existing storage abstractions rely on fine-grained updates, which are expensive to track.</li>
<li>Need for a better balance between efficiency and fault tolerance.</li>
</ul>
</div>
</div>

<div id="outline-container-org4e62c95" class="outline-3">
<h3 id="org4e62c95"><span class="section-number-3">9.7.</span> Solution  RDDs</h3>
<div class="outline-text-3" id="text-9-7">
<ul class="org-ul">
<li>Spark uses RDDs to balance fault-tolerance and efficiency.</li>
<li>RDDs support coarse-grained transformations (e.g., map, group-by, join).</li>
<li>Lineage tracking enables fault recovery without excessive replication.</li>
</ul>
</div>
</div>

<div id="outline-container-org696454a" class="outline-3">
<h3 id="org696454a"><span class="section-number-3">9.8.</span> RDD Recovery</h3>
<div class="outline-text-3" id="text-9-8">
<ul class="org-ul">
<li>Recovery through recomputation based on lineage.</li>
<li>Interactive: rerun only failed iterations.</li>
<li>Iterative: reload or recompute chunks from previous iterations.</li>
</ul>
</div>
</div>

<div id="outline-container-org367ba48" class="outline-3">
<h3 id="org367ba48"><span class="section-number-3">9.9.</span> Spark Stack</h3>
<div class="outline-text-3" id="text-9-9">
<ul class="org-ul">
<li>Core: scheduling, memory management, fault recovery.</li>
<li>APIs: Spark SQL, Streaming, MLib (ML), GraphX.</li>
<li>Compatible with YARN, Mesos, or standalone.</li>
</ul>
</div>
</div>

<div id="outline-container-org3de0073" class="outline-3">
<h3 id="org3de0073"><span class="section-number-3">9.10.</span> Spark Programming Interface</h3>
<div class="outline-text-3" id="text-9-10">
<ul class="org-ul">
<li>Interfaces: Scala, Python, Java.</li>
<li>RDD operations: transformations and actions.</li>
<li>Support for shared variables (broadcast, accumulator).</li>
</ul>
</div>
</div>

<div id="outline-container-org4a94afc" class="outline-3">
<h3 id="org4a94afc"><span class="section-number-3">9.11.</span> RDD Transformations</h3>
<div class="outline-text-3" id="text-9-11">
<ul class="org-ul">
<li>Examples: map vs flatMap, distinct, union, intersection, subtract.</li>
<li>flatMap flattens nested lists from map.</li>
<li>Intersection and subtract are more costly due to set operations.</li>
</ul>
</div>
</div>

<div id="outline-container-orga4dfb01" class="outline-3">
<h3 id="orga4dfb01"><span class="section-number-3">9.12.</span> RDD Transformations Quiz Question</h3>
<div class="outline-text-3" id="text-9-12">
<ul class="org-ul">
<li>Quiz: Apply map and filter on input RDD to understand transformation output.</li>
</ul>
</div>
</div>

<div id="outline-container-org9ce33ca" class="outline-3">
<h3 id="org9ce33ca"><span class="section-number-3">9.13.</span> RDD Transformations Quiz Solution</h3>
<div class="outline-text-3" id="text-9-13">
<ul class="org-ul">
<li>Map squares each number: [1, 4, 9, 16].</li>
<li>Filter excludes ‘1’: resulting in [2, 3, 4].</li>
</ul>
</div>
</div>

<div id="outline-container-orge32532e" class="outline-3">
<h3 id="orge32532e"><span class="section-number-3">9.14.</span> Spark Operations</h3>
<div class="outline-text-3" id="text-9-14">
<ul class="org-ul">
<li>Transformations: map, filter, groupByKey, etc.</li>
<li>Actions: reduce, collect, count, etc.</li>
<li>Broader set than Hadoop MapReduce.</li>
</ul>
</div>
</div>

<div id="outline-container-org391c68a" class="outline-3">
<h3 id="org391c68a"><span class="section-number-3">9.15.</span> Shared Variable</h3>
<div class="outline-text-3" id="text-9-15">
<ul class="org-ul">
<li>Broadcast variables: efficient read-only shared data.</li>
<li>Example Spark job with clinical notes:
<ul class="org-ul">
<li>Load data → filter → split → cache.</li>
<li>Perform actions like counting specific symptoms (e.g., fever, cough).</li>
<li>Benefits from caching and distributed execution.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org799c26e" class="outline-3">
<h3 id="org799c26e"><span class="section-number-3">9.16.</span> Fault Tolerance</h3>
<div class="outline-text-3" id="text-9-16">
<ul class="org-ul">
<li>RDD tracks transformations (lineage).</li>
<li>Lost partitions can be recomputed using lineage.</li>
<li>Visuals: sequence of filters and maps reconstruct lost data.</li>
</ul>
</div>
</div>

<div id="outline-container-org0ca2b32" class="outline-3">
<h3 id="org0ca2b32"><span class="section-number-3">9.17.</span> Example  Logistic Regression</h3>
<div class="outline-text-3" id="text-9-17">
<ul class="org-ul">
<li>Spark efficiently supports logistic regression via iteration.</li>
<li>Data cached in memory → quick updates.</li>
<li>Uses map + reduce to compute gradient.</li>
<li>Comparison: Spark (6s/iter) vs Hadoop (127s/iter).</li>
</ul>
</div>
</div>

<div id="outline-container-org3f4fbd8" class="outline-3">
<h3 id="org3f4fbd8"><span class="section-number-3">9.18.</span> Example  Disease Risk Prediction</h3>
<div class="outline-text-3" id="text-9-18">
<ul class="org-ul">
<li>Predict patient disease risks using collaborative filtering.</li>
<li>Matrix R = patient × disease.</li>
<li>ALS algorithm: alternates updating patient and disease features.</li>
</ul>
</div>
</div>

<div id="outline-container-org126714f" class="outline-3">
<h3 id="org126714f"><span class="section-number-3">9.19.</span> Serial ALS</h3>
<div class="outline-text-3" id="text-9-19">
<ul class="org-ul">
<li>Serial implementation of ALS.</li>
<li>Iterative updates of A (patient) and B (disease) matrices.</li>
<li>Uses map functions over patient/disease ranges.</li>
</ul>
</div>
</div>

<div id="outline-container-org49f92a0" class="outline-3">
<h3 id="org49f92a0"><span class="section-number-3">9.20.</span> Naive Spark ALS</h3>
<div class="outline-text-3" id="text-9-20">
<ul class="org-ul">
<li>Parallel ALS using `spark.parallelize`.</li>
<li>Broadcast used for efficient data sharing.</li>
<li>Broadcast improves performance 3× over naive version.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org335123a" class="outline-2">
<h2 id="org335123a"><span class="section-number-2">10.</span> Medical Ontology</h2>
<div class="outline-text-2" id="text-10">
</div>
<div id="outline-container-orga0a5f88" class="outline-3">
<h3 id="orga0a5f88"><span class="section-number-3">10.1.</span> Introduction to Medical Ontology</h3>
<div class="outline-text-3" id="text-10-1">
<ul class="org-ul">
<li>Healthcare benefits from structured medical knowledge.</li>
<li>Ontologies and knowledge graphs are common in this domain.</li>
<li>Historically, healthcare has developed ontologies for:
<ul class="org-ul">
<li>Diseases</li>
<li>Medical procedures</li>
<li>Medications</li>
<li>Lab tests</li>
</ul></li>
<li>Ontologies aid understanding of healthcare data and validate data models.</li>
<li>The lesson discusses various ontologies and their role in analytics.</li>
</ul>
</div>
</div>

<div id="outline-container-org4297432" class="outline-3">
<h3 id="org4297432"><span class="section-number-3">10.2.</span> Health Data Standards</h3>
<div class="outline-text-3" id="text-10-2">
<ul class="org-ul">
<li>Overview of major health data standards:
<ul class="org-ul">
<li>LOINC: Lab test codes</li>
<li>ICD: Diagnostic codes</li>
<li>CPT: Procedure codes</li>
<li>NDC: Medication codes</li>
</ul></li>
<li>SNOMED: A comprehensive ontology integrating all the above.</li>
<li>UMLS: Software system that accesses integrated medical knowledge.</li>
<li>Use case: Patient encounter example illustrating all standard usages.</li>
<li>Importance:
<ul class="org-ul">
<li>Insurance claim processing</li>
<li>Research and data analysis</li>
</ul></li>
<li>Preview of upcoming detailed discussions on individual standards.</li>
</ul>
</div>
</div>

<div id="outline-container-org18de379" class="outline-3">
<h3 id="org18de379"><span class="section-number-3">10.3.</span> ICD</h3>
<div class="outline-text-3" id="text-10-3">
<ul class="org-ul">
<li>ICD = International Classification of Diseases.</li>
<li>Developed by WHO, categorizes diseases.</li>
<li>ICD-9: 17,000+ codes; 3-5 digits; includes main and supplementary categories (e.g., E, V codes).</li>
<li>ICD-10: 141,000+ codes; 7-character alphanumeric structure.</li>
<li>Detailed coding for category, etiology, body part, severity, and extension.</li>
<li>Examples:
<ul class="org-ul">
<li>ICD-9: 250.01 = Type 1 diabetes, no complications</li>
<li>ICD-10: M1A.31X2 = Chronic gout with renal impairment in left shoulder, no tophus</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org83e331e" class="outline-3">
<h3 id="org83e331e"><span class="section-number-3">10.4.</span> ICD 9 to ICD 10 Mapping</h3>
<div class="outline-text-3" id="text-10-4">
<ul class="org-ul">
<li>Mapping is mostly one-to-many due to ICD-10's higher specificity.</li>
<li>Some mappings are one-to-one (e.g., Tietze's Syndrome).</li>
<li>Complicated mappings: e.g., ICD-9 code maps to 2,530 ICD-10 codes.</li>
<li>Complexity highlights the increased detail of ICD-10.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf0d5795" class="outline-3">
<h3 id="orgf0d5795"><span class="section-number-3">10.5.</span> ICD 9 Quiz Question</h3>
<div class="outline-text-3" id="text-10-5">
<ul class="org-ul">
<li>Identifies valid and invalid ICD-9 codes.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd9c18c7" class="outline-3">
<h3 id="orgd9c18c7"><span class="section-number-3">10.6.</span> ICD 9 Quiz Solution</h3>
<div class="outline-text-3" id="text-10-6">
<ul class="org-ul">
<li>Valid: 501, 802.3, V70, E820.0</li>
<li>Invalid: U80.1, 5A0.01</li>
</ul>
</div>
</div>

<div id="outline-container-orge861efc" class="outline-3">
<h3 id="orge861efc"><span class="section-number-3">10.7.</span> ICD Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-7">
<ul class="org-ul">
<li>Task: Find ICD-9 and ICD-10 codes for Influenza.</li>
</ul>
</div>
</div>

<div id="outline-container-org54a19a8" class="outline-3">
<h3 id="org54a19a8"><span class="section-number-3">10.8.</span> ICD Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-8">
<ul class="org-ul">
<li>ICD-9: 487.1</li>
<li>ICD-10: J11.1</li>
</ul>
</div>
</div>

<div id="outline-container-org2f13aca" class="outline-3">
<h3 id="org2f13aca"><span class="section-number-3">10.9.</span> CPT</h3>
<div class="outline-text-3" id="text-10-9">
<ul class="org-ul">
<li>CPT = Current Procedural Terminology (US standard).</li>
<li>Used for medical, surgical, diagnostic procedures.</li>
<li>Maintained by the AMA.</li>
<li>Key in insurance reimbursement.</li>
<li>Categories:
<ul class="org-ul">
<li>Category 1: Common procedures (e.g., surgery, radiology)</li>
<li>Category 2: Quality metrics</li>
<li>Category 3: Experimental procedures (e.g., codes ending in T)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orga0f3e3e" class="outline-3">
<h3 id="orga0f3e3e"><span class="section-number-3">10.10.</span> CPT Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-10">
<ul class="org-ul">
<li>Task: Find CPT code for detailed office visit.</li>
</ul>
</div>
</div>

<div id="outline-container-org78504db" class="outline-3">
<h3 id="org78504db"><span class="section-number-3">10.11.</span> CPT Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-11">
<ul class="org-ul">
<li>Office visit range: 99201–99205</li>
<li>Time-based codes (e.g., 99201 = 10 mins, 99205 = 1 hour)</li>
</ul>
</div>
</div>

<div id="outline-container-org60d970e" class="outline-3">
<h3 id="org60d970e"><span class="section-number-3">10.12.</span> LOINC</h3>
<div class="outline-text-3" id="text-10-12">
<ul class="org-ul">
<li>LOINC = Logical Observation Identifiers Names and Codes.</li>
<li>Created by Regenstrief Institute.</li>
<li>Standard for lab tests and clinical observations.</li>
<li>Example: 2865-4 = a specific lab test</li>
<li>Attributes:
<ul class="org-ul">
<li>Component</li>
<li>Property</li>
<li>Time aspect (e.g., Pt = point in time)</li>
<li>Sample type</li>
<li>Scale (quantitative, ordinal, etc.)</li>
<li>Method</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org9ced1af" class="outline-3">
<h3 id="org9ced1af"><span class="section-number-3">10.13.</span> LOINC Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-13">
<ul class="org-ul">
<li>Task: Find LOINC code for Creatinine lab test.</li>
</ul>
</div>
</div>

<div id="outline-container-org9413dd1" class="outline-3">
<h3 id="org9413dd1"><span class="section-number-3">10.14.</span> LOINC Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-14">
<ul class="org-ul">
<li>Answer: 2160-0</li>
</ul>
</div>
</div>

<div id="outline-container-orge522008" class="outline-3">
<h3 id="orge522008"><span class="section-number-3">10.15.</span> NDC</h3>
<div class="outline-text-3" id="text-10-15">
<ul class="org-ul">
<li>NDC = National Drug Code (for medications).</li>
<li>Maintained by the FDA.</li>
<li>Structure: 3 segments
<ul class="org-ul">
<li>Labeler code (company)</li>
<li>Product code (drug type)</li>
<li>Package code (package details)</li>
</ul></li>
<li>Different digit combinations exist but total is 10 digits.</li>
</ul>
</div>
</div>

<div id="outline-container-org7b51962" class="outline-3">
<h3 id="org7b51962"><span class="section-number-3">10.16.</span> NDC Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-16">
<ul class="org-ul">
<li>Task: Find NDC code for metformin hydrochloride, 500mg.</li>
</ul>
</div>
</div>

<div id="outline-container-org7123e47" class="outline-3">
<h3 id="org7123e47"><span class="section-number-3">10.17.</span> NDC Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-17">
<ul class="org-ul">
<li>Example code: 0093-7214-01 (labeler-dependent)</li>
</ul>
</div>
</div>

<div id="outline-container-org3d0ba6e" class="outline-3">
<h3 id="org3d0ba6e"><span class="section-number-3">10.18.</span> SNOMED</h3>
<div class="outline-text-3" id="text-10-18">
<ul class="org-ul">
<li>SNOMED = Systematized Nomenclature of Medicine.</li>
<li>Maintained by IHTSDO (Denmark).</li>
<li>Supports clinical documentation and semantic interoperability.</li>
<li>National editions and implementation subsets exist.</li>
<li>Broad user base: clinicians, researchers, analysts.</li>
</ul>
</div>
</div>

<div id="outline-container-orgafab91f" class="outline-3">
<h3 id="orgafab91f"><span class="section-number-3">10.19.</span> Logical Model of SNOMED CT</h3>
<div class="outline-text-3" id="text-10-19">
<ul class="org-ul">
<li>Core components:
<ul class="org-ul">
<li>Concepts (with unique ID)</li>
<li>Descriptions (FSN and synonyms)</li>
<li>Relationships (e.g., “is a”, attributes)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orge6725c7" class="outline-3">
<h3 id="orge6725c7"><span class="section-number-3">10.20.</span> SNOMED Example</h3>
<div class="outline-text-3" id="text-10-20">
<ul class="org-ul">
<li>Example concept: 22298006 = Myocardial infarction disorder</li>
<li>Multiple descriptions including synonyms</li>
<li>Preferred vs. acceptable terms are defined by reference sets.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd040faa" class="outline-3">
<h3 id="orgd040faa"><span class="section-number-3">10.21.</span> SNOMED Relationships</h3>
<div class="outline-text-3" id="text-10-21">
<ul class="org-ul">
<li>“Is a” relationship creates concept hierarchies.</li>
<li>Example: Cellulitis of foot “is a” cellulitis, and “is a” disorder of foot.</li>
<li>Relationships are directional and acyclic.</li>
</ul>
</div>
</div>

<div id="outline-container-org32d9b73" class="outline-3">
<h3 id="org32d9b73"><span class="section-number-3">10.22.</span> SNOMED Design</h3>
<div class="outline-text-3" id="text-10-22">
<ul class="org-ul">
<li>19-level hierarchy of concepts.</li>
<li>Concepts range from general to highly specific.</li>
<li>Relationships include “is a” and others with attributes.</li>
<li>Each concept has a unique ID and descriptions.</li>
</ul>
</div>
</div>

<div id="outline-container-org7deea18" class="outline-3">
<h3 id="org7deea18"><span class="section-number-3">10.23.</span> SNOMED Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-23">
<ul class="org-ul">
<li>Task: Find SNOMED code for chronic gouty arthritis.</li>
</ul>
</div>
</div>

<div id="outline-container-orge0ad860" class="outline-3">
<h3 id="orge0ad860"><span class="section-number-3">10.24.</span> SNOMED Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-24">
<ul class="org-ul">
<li>Answer: 68451005</li>
</ul>
</div>
</div>

<div id="outline-container-org95b914e" class="outline-3">
<h3 id="org95b914e"><span class="section-number-3">10.25.</span> SNOMED Quiz Question</h3>
<div class="outline-text-3" id="text-10-25">
<ul class="org-ul">
<li>Question: What is the structure of "is a" relationships in SNOMED?</li>
</ul>
</div>
</div>

<div id="outline-container-org504b8e5" class="outline-3">
<h3 id="org504b8e5"><span class="section-number-3">10.26.</span> SNOMED Quiz Solution</h3>
<div class="outline-text-3" id="text-10-26">
<ul class="org-ul">
<li>Answer: Directed acyclic graph (not a tree).</li>
</ul>
</div>
</div>

<div id="outline-container-org9f68b32" class="outline-3">
<h3 id="org9f68b32"><span class="section-number-3">10.27.</span> UMLS</h3>
<div class="outline-text-3" id="text-10-27">
<ul class="org-ul">
<li>UMLS = Unified Medical Language System.</li>
<li>Maintained by US National Library of Medicine.</li>
<li>Integrates multiple standards into a unified system.</li>
<li>Components:
<ul class="org-ul">
<li>Metathesaurus</li>
<li>Semantic network</li>
<li>Specialist lexicon and tools</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org252770f" class="outline-3">
<h3 id="org252770f"><span class="section-number-3">10.28.</span> Metathesaurus Concepts</h3>
<div class="outline-text-3" id="text-10-28">
<ul class="org-ul">
<li>Concept hierarchy:
<ul class="org-ul">
<li>Atom (AUI) → String (SUI) → Term (LUI) → Concept (CUI)</li>
</ul></li>
<li>Multiple ontologies/sources integrated.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaf7e657" class="outline-3">
<h3 id="orgaf7e657"><span class="section-number-3">10.29.</span> Semantic Network</h3>
<div class="outline-text-3" id="text-10-29">
<ul class="org-ul">
<li>135+ semantic types (e.g., diseases, drugs)</li>
<li>54+ semantic relationships (e.g., cause, treat)</li>
<li>Organizes concepts and their interrelations.</li>
</ul>
</div>
</div>

<div id="outline-container-orgef4b9af" class="outline-3">
<h3 id="orgef4b9af"><span class="section-number-3">10.30.</span> Specialist Lexicon</h3>
<div class="outline-text-3" id="text-10-30">
<ul class="org-ul">
<li>Lexicon of 300,000+ biomedical terms.</li>
<li>Includes syntax, morphology, orthography.</li>
<li>Used in NLP tools like MetaMap and MMTx.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org188b11d" class="outline-2">
<h2 id="org188b11d"><span class="section-number-2">11.</span> Graph Analysis</h2>
<div class="outline-text-2" id="text-11">
</div>
<div id="outline-container-orgc95d7fe" class="outline-3">
<h3 id="orgc95d7fe"><span class="section-number-3">11.1.</span> Introduction to Graph Analysis</h3>
<div class="outline-text-3" id="text-11-1">
<ul class="org-ul">
<li>Overview of graph analysis and its applications.</li>
<li>Used in search engines, social networks, and healthcare.</li>
<li>Identifies clusters of related entities (webpages, patients, conditions).</li>
<li>Will cover similarity graphs and spectral clustering.</li>
</ul>
</div>
</div>

<div id="outline-container-org3358431" class="outline-3">
<h3 id="org3358431"><span class="section-number-3">11.2.</span> Agenda</h3>
<div class="outline-text-3" id="text-11-2">
<ul class="org-ul">
<li>Two main graph-based algorithms introduced:
<ul class="org-ul">
<li>PageRank: Ranks nodes in a directed graph based on importance.</li>
<li>Spectral Clustering: A graph partitioning-based clustering algorithm.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org3544f4c" class="outline-3">
<h3 id="org3544f4c"><span class="section-number-3">11.3.</span> PageRank</h3>
<div class="outline-text-3" id="text-11-3">
<ul class="org-ul">
<li>Developed by Google co-founders for webpage ranking.</li>
<li>Traditional content-based ranking is prone to spam.</li>
<li>PageRank relies on the graph structure (links between pages).</li>
<li>Intuition: More high-quality pages linking to a page → higher rank.</li>
<li>Represent graph using an adjacency matrix:
<ul class="org-ul">
<li>Normalize matrix rows to represent transition probabilities.</li>
<li>PageRank computed using recursion: browsing + teleporting.</li>
<li>Browsing redistributes rank via adjacency matrix; teleporting randomly jumps.</li>
<li>Equation: `q = cAq + (1 - c)e/N`, where `c` is damping factor (e.g., 0.85).</li>
</ul></li>
<li>Next steps: Implement PageRank using big data systems like Hadoop MapReduce.</li>
</ul>
</div>
</div>

<div id="outline-container-orge21abfe" class="outline-3">
<h3 id="orge21abfe"><span class="section-number-3">11.4.</span> MapReduce  PageRank</h3>
<div class="outline-text-3" id="text-11-4">
<ul class="org-ul">
<li>Implementation of PageRank in MapReduce:
<ul class="org-ul">
<li><b><b>Map Phase</b></b>:
<ul class="org-ul">
<li>Emit current page and its PageRank.</li>
<li>Distribute PageRank to each outgoing link equally.</li>
</ul></li>
<li><b><b>Reduce Phase</b></b>:
<ul class="org-ul">
<li>Aggregate partial PageRanks from map phase.</li>
<li>Add teleporting component.</li>
<li>Emit updated PageRank.</li>
</ul></li>
<li>Hadoop shuffles partial sums to reducers.</li>
<li>Iterative process: repeat map-reduce steps to converge.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org1214d90" class="outline-3">
<h3 id="org1214d90"><span class="section-number-3">11.5.</span> PageRank Quiz Question</h3>
<div class="outline-text-3" id="text-11-5">
<ul class="org-ul">
<li>Quiz: Rank webpages in a directed graph by PageRank (based on structure).</li>
</ul>
</div>
</div>

<div id="outline-container-org11fb5a9" class="outline-3">
<h3 id="org11fb5a9"><span class="section-number-3">11.6.</span> PageRank Quiz Solution</h3>
<div class="outline-text-3" id="text-11-6">
<ul class="org-ul">
<li>Highest: YouTube (3 incoming links).</li>
<li>Lowest: Google (no incoming links).</li>
<li>Recursive PageRank shows differences despite similar local structures.</li>
<li>Amazon &gt; Wikipedia &gt; Facebook in mid-tier due to recursion effects.</li>
</ul>
</div>
</div>

<div id="outline-container-orgdb6804e" class="outline-3">
<h3 id="orgdb6804e"><span class="section-number-3">11.7.</span> Spectral Clustering</h3>
<div class="outline-text-3" id="text-11-7">
<ul class="org-ul">
<li>Spectral Clustering vs Traditional Clustering:
<ul class="org-ul">
<li>Uses similarity graph of patients based on features.</li>
<li>Steps:
<ol class="org-ol">
<li>Construct similarity graph (nodes = patients, edges = similarity).</li>
<li>Convert graph to adjacency matrix.</li>
<li>Compute top-k eigenvectors of the matrix.</li>
<li>Cluster eigenvectors into groups.</li>
</ol></li>
<li>Effective for high-dimensional or noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org6263fd3" class="outline-3">
<h3 id="org6263fd3"><span class="section-number-3">11.8.</span> Similarity Graph Construction</h3>
<div class="outline-text-3" id="text-11-8">
<ul class="org-ul">
<li>Several ways to construct similarity graphs:
<ul class="org-ul">
<li>Epsilon-neighborhood.</li>
<li>K-nearest neighbors.</li>
<li>Fully connected with edge weights (e.g., Gaussian kernel).</li>
</ul></li>
<li>Purpose: capture local similarity between patients.</li>
</ul>
</div>
</div>

<div id="outline-container-org4304566" class="outline-3">
<h3 id="org4304566"><span class="section-number-3">11.9.</span> E Neighborhood Graph</h3>
<div class="outline-text-3" id="text-11-9">
<ul class="org-ul">
<li>Connect nodes within ε distance.</li>
<li>No edge if distance &gt; ε.</li>
<li>Defines sparsity based on ε.</li>
</ul>
</div>
</div>

<div id="outline-container-org1ad4d87" class="outline-3">
<h3 id="org1ad4d87"><span class="section-number-3">11.10.</span> K Nearest Neighbor Graph</h3>
<div class="outline-text-3" id="text-11-10">
<ul class="org-ul">
<li>Each node connects to its k-nearest neighbors (directed).</li>
<li>Results in sparse graphs for small k.</li>
<li>Variants: binary edges vs weighted edges based on distance.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbf0eb70" class="outline-3">
<h3 id="orgbf0eb70"><span class="section-number-3">11.11.</span> Fully Connected Graph</h3>
<div class="outline-text-3" id="text-11-11">
<ul class="org-ul">
<li>Every node connected to every other.</li>
<li>Edge weights determined by similarity (e.g., Gaussian kernel).</li>
<li>Most dense graph representation.</li>
</ul>
</div>
</div>

<div id="outline-container-org4e4cf2a" class="outline-3">
<h3 id="org4e4cf2a"><span class="section-number-3">11.12.</span> E Neighborhood Graph Quiz Question</h3>
<div class="outline-text-3" id="text-11-12">
<ul class="org-ul">
<li>Quiz: Select best ε value to reflect clustering structure in 2D patient data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfd58f5f" class="outline-3">
<h3 id="orgfd58f5f"><span class="section-number-3">11.13.</span> E Neighborhood Graph Quiz Solution</h3>
<div class="outline-text-3" id="text-11-13">
<ul class="org-ul">
<li>Small ε: too many disconnected clusters.</li>
<li>Large ε: single cluster.</li>
<li>Optimal ε (e.g., B or C) reveals natural clusters in data.</li>
</ul>
</div>
</div>

<div id="outline-container-org35437d3" class="outline-3">
<h3 id="org35437d3"><span class="section-number-3">11.14.</span> Spectral Clustering Algorithm</h3>
<div class="outline-text-3" id="text-11-14">
<ul class="org-ul">
<li>Variations exist:
<ul class="org-ul">
<li>Unnormalized vs Normalized Spectral Clustering.</li>
</ul></li>
<li>Example: simplest uses eigenvectors + k-means.</li>
<li>Spectral clustering transforms data to eigenspace where clusters are clearer.</li>
<li>Recommended reading for theoretical underpinnings.</li>
</ul>
</div>
</div>

<div id="outline-container-orge9bff4e" class="outline-3">
<h3 id="orge9bff4e"><span class="section-number-3">11.15.</span> Big Data Conclusion</h3>
<div class="outline-text-3" id="text-11-15">
<ul class="org-ul">
<li>Wrap-up of the video series.</li>
<li>Presenter: Jimmy Son.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org4d21bbb" class="outline-2">
<h2 id="org4d21bbb"><span class="section-number-2">12.</span> Dimensionality Reduction</h2>
<div class="outline-text-2" id="text-12">
</div>
<div id="outline-container-org8e048ca" class="outline-3">
<h3 id="org8e048ca"><span class="section-number-3">12.1.</span> Introduction to Dimensionality Reduction</h3>
<div class="outline-text-3" id="text-12-1">
<ul class="org-ul">
<li>Overview of dimensionality reduction as a method for simplifying noisy, high-dimensional data.</li>
<li>Methods covered:
<ul class="org-ul">
<li>Singular Value Decomposition (SVD)</li>
<li>Principal Component Analysis (PCA)</li>
<li>CUR Decomposition</li>
<li>Tensor Factorization</li>
</ul></li>
<li>Applications in predictive modeling and healthcare phenotyping.</li>
</ul>
</div>
</div>

<div id="outline-container-org4ff8afe" class="outline-3">
<h3 id="org4ff8afe"><span class="section-number-3">12.2.</span> Dimensionality Reduction</h3>
<div class="outline-text-3" id="text-12-2">
<ul class="org-ul">
<li>Recap of clustering algorithms (hard, soft, and scalable).</li>
<li>Motivation for dimensionality reduction: efficiency and robustness.</li>
<li>Classical methods:
<ul class="org-ul">
<li>SVD and PCA: summarize features via linear combinations.</li>
<li>CUR: uses actual data rows/columns, preserves sparsity.</li>
</ul></li>
<li>Introduction to Tensor Factorization for higher-order data.</li>
</ul>
</div>
</div>

<div id="outline-container-org82e6e06" class="outline-3">
<h3 id="org82e6e06"><span class="section-number-3">12.3.</span> Singular Value Decomposition</h3>
<div class="outline-text-3" id="text-12-3">
<ul class="org-ul">
<li>Definition: SVD decomposes a matrix X into U, Σ, Vᵗ.</li>
<li>U and V: orthogonal matrices (singular vectors).</li>
<li>Σ: diagonal matrix with singular values (in descending order).</li>
<li>Visual interpretation with patients (rows) and diseases (columns).</li>
</ul>
</div>
</div>

<div id="outline-container-org9b4fa61" class="outline-3">
<h3 id="org9b4fa61"><span class="section-number-3">12.4.</span> SVD Example</h3>
<div class="outline-text-3" id="text-12-4">
<ul class="org-ul">
<li>Spectral view: rank-1 approximations via σᵢUᵢVᵢᵗ.</li>
<li>Example with documents and terms:
<ul class="org-ul">
<li>CS vs. Medical documents.</li>
<li>Each rank-1 component captures topic-specific vocabulary.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org737d351" class="outline-3">
<h3 id="org737d351"><span class="section-number-3">12.5.</span> SVD Properties</h3>
<div class="outline-text-3" id="text-12-5">
<ul class="org-ul">
<li>V: eigenvectors of XᵗX with eigenvalues Σ².</li>
<li>U: eigenvectors of XXᵗ.</li>
<li>Reinforces relationship between SVD and eigen-decomposition.</li>
</ul>
</div>
</div>

<div id="outline-container-org53fc049" class="outline-3">
<h3 id="org53fc049"><span class="section-number-3">12.6.</span> Quiz  SVD Interpretation Question</h3>
<div class="outline-text-3" id="text-12-6">
<ul class="org-ul">
<li>Question about AᵗA matrix from a document-term matrix.</li>
</ul>
</div>
</div>

<div id="outline-container-org8d27be7" class="outline-3">
<h3 id="org8d27be7"><span class="section-number-3">12.7.</span> Quiz  SVD Interpretation Solution</h3>
<div class="outline-text-3" id="text-12-7">
<ul class="org-ul">
<li>Answer: AᵗA is a term-to-term similarity matrix.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaa0abbc" class="outline-3">
<h3 id="orgaa0abbc"><span class="section-number-3">12.8.</span> Principal Component Analysis</h3>
<div class="outline-text-3" id="text-12-8">
<ul class="org-ul">
<li>PCA is derived from SVD: UΣ = principal components, Vᵗ = loadings.</li>
<li>Visualized with a patient-by-disease matrix.</li>
<li>Principal components capture the main variance direction.</li>
</ul>
</div>
</div>

<div id="outline-container-orge2af891" class="outline-3">
<h3 id="orge2af891"><span class="section-number-3">12.9.</span> PCA Interpretation</h3>
<div class="outline-text-3" id="text-12-9">
<ul class="org-ul">
<li>PCA reduces dimensions by projecting onto principal component directions.</li>
<li>Example with weight and height to one dimension.</li>
</ul>
</div>
</div>

<div id="outline-container-org31677cc" class="outline-3">
<h3 id="org31677cc"><span class="section-number-3">12.10.</span> Sparsity Problem with SVD</h3>
<div class="outline-text-3" id="text-12-10">
<ul class="org-ul">
<li>SVD and PCA generate dense outputs, losing sparsity of original data.</li>
<li>CUR decomposition preserves sparsity and reduces storage/computation cost.</li>
</ul>
</div>
</div>

<div id="outline-container-org0c98494" class="outline-3">
<h3 id="org0c98494"><span class="section-number-3">12.11.</span> CUR Decomposition</h3>
<div class="outline-text-3" id="text-12-11">
<ul class="org-ul">
<li>CUR approximates A ≈ CUR using actual rows and columns.</li>
<li>Comparison with SVD: CUR retains interpretability and sparsity.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb68196b" class="outline-3">
<h3 id="orgb68196b"><span class="section-number-3">12.12.</span> CUR Algorithm</h3>
<div class="outline-text-3" id="text-12-12">
<ul class="org-ul">
<li>Algorithm steps:
<ol class="org-ol">
<li>Perform SVD.</li>
<li>Sample rows/columns based on vector norms.</li>
<li>Compute U as pseudo-inverse operations.</li>
</ol></li>
<li>Sampling based on row norms of U and V.</li>
</ul>
</div>
</div>

<div id="outline-container-org3e85fcf" class="outline-3">
<h3 id="org3e85fcf"><span class="section-number-3">12.13.</span> CUR Quiz Question</h3>
<div class="outline-text-3" id="text-12-13">
<ul class="org-ul">
<li>Question: what do C and R represent in CUR decomposition?</li>
</ul>
</div>
</div>

<div id="outline-container-org2793f85" class="outline-3">
<h3 id="org2793f85"><span class="section-number-3">12.14.</span> CUR Quiz Solution</h3>
<div class="outline-text-3" id="text-12-14">
<ul class="org-ul">
<li>Answer: C = actual diagnoses (columns), R = actual patients (rows).</li>
<li>Emphasizes CUR’s interpretability and sparsity preservation.</li>
</ul>
</div>
</div>

<div id="outline-container-org5730ef6" class="outline-3">
<h3 id="org5730ef6"><span class="section-number-3">12.15.</span> Tensor for EHR</h3>
<div class="outline-text-3" id="text-12-15">
<ul class="org-ul">
<li>Tensor = generalized matrix (e.g., 3D tensor: patient × diagnosis × medication).</li>
<li>Captures multi-dimensional interactions in EHR data.</li>
</ul>
</div>
</div>

<div id="outline-container-orga4b312e" class="outline-3">
<h3 id="orga4b312e"><span class="section-number-3">12.16.</span> Tensor Slicing</h3>
<div class="outline-text-3" id="text-12-16">
<ul class="org-ul">
<li>Extract subtensors (matrices) by fixing all but two modes.</li>
<li>Examples:
<ul class="org-ul">
<li>Diagnosis-medication matrix for a patient.</li>
<li>Patients and diseases by medication.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org239aae5" class="outline-3">
<h3 id="org239aae5"><span class="section-number-3">12.17.</span> Rank 1 Tensor</h3>
<div class="outline-text-3" id="text-12-17">
<ul class="org-ul">
<li>Rank-1 tensor = outer product of vectors from each mode.</li>
<li>Element-wise: aᵢ * bⱼ * cₖ.</li>
</ul>
</div>
</div>

<div id="outline-container-org50be58b" class="outline-3">
<h3 id="org50be58b"><span class="section-number-3">12.18.</span> Example Phenotype</h3>
<div class="outline-text-3" id="text-12-18">
<ul class="org-ul">
<li>Phenotypes modeled as rank-1 tensors.</li>
<li>Example: hypertension phenotype with specific meds and patients.</li>
</ul>
</div>
</div>

<div id="outline-container-org080cf2c" class="outline-3">
<h3 id="org080cf2c"><span class="section-number-3">12.19.</span> Phenotyping Through Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-19">
<ul class="org-ul">
<li>Tensor = sum of rank-1 tensors (each = phenotype).</li>
<li>Each component has patient, diagnosis, and medication factors.</li>
<li>λ = importance of each phenotype.</li>
</ul>
</div>
</div>

<div id="outline-container-org17eb805" class="outline-3">
<h3 id="org17eb805"><span class="section-number-3">12.20.</span> CP Decomposition</h3>
<div class="outline-text-3" id="text-12-20">
<ul class="org-ul">
<li>Canonical Polyadic (CP) Decomposition = sum of R rank-1 tensors.</li>
<li>Each tensor has one λ and one vector per mode.</li>
<li>Output: three factor matrices and λ vector.</li>
<li>Often used with non-negativity constraints.</li>
</ul>
</div>
</div>

<div id="outline-container-org5ea4ea0" class="outline-3">
<h3 id="org5ea4ea0"><span class="section-number-3">12.21.</span> Phenotyping Process Using Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-21">
<ul class="org-ul">
<li>Process:
<ol class="org-ol">
<li>Factorize training tensor to learn phenotype factors.</li>
<li>Project new patients onto phenotype space.</li>
</ol></li>
<li>Result: low-dimensional patient representation.</li>
</ul>
</div>
</div>

<div id="outline-container-org1d7accb" class="outline-3">
<h3 id="org1d7accb"><span class="section-number-3">12.22.</span> Phenotyping for Predictive Modeling</h3>
<div class="outline-text-3" id="text-12-22">
<ul class="org-ul">
<li>Build tensor from EHR events within an observation window.</li>
<li>Target = predict heart failure.</li>
<li>Tensor = patient × diagnosis × medication.</li>
<li>Phenotypes used as features for prediction.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfb62dca" class="outline-3">
<h3 id="orgfb62dca"><span class="section-number-3">12.23.</span> Predictive Performance</h3>
<div class="outline-text-3" id="text-12-23">
<ul class="org-ul">
<li>Compared PCA, NMF, and Tensor Factorization.</li>
<li>All outperformed baseline with fewer features (phenotypes).</li>
<li>Tensor and NMF had better performance than PCA.</li>
</ul>
</div>
</div>

<div id="outline-container-org7f3d309" class="outline-3">
<h3 id="org7f3d309"><span class="section-number-3">12.24.</span> Major Disease Phenotypes</h3>
<div class="outline-text-3" id="text-12-24">
<ul class="org-ul">
<li>Tensor reveals intuitive disease phenotypes:
<ul class="org-ul">
<li>Uncomplicated diabetes (17.6%)</li>
<li>Mild hypertension (31.1%)</li>
</ul></li>
<li>Also identifies subtypes (e.g., mild/moderate/severe hypertension).</li>
</ul>
</div>
</div>

<div id="outline-container-orge64f63d" class="outline-3">
<h3 id="orge64f63d"><span class="section-number-3">12.25.</span> Tensor vs NMF</h3>
<div class="outline-text-3" id="text-12-25">
<ul class="org-ul">
<li>Tensor factorization offers more concise, interpretable phenotypes.</li>
<li>NMF generates complex feature sets with many interactions.</li>
</ul>
</div>
</div>

<div id="outline-container-org1d92418" class="outline-3">
<h3 id="org1d92418"><span class="section-number-3">12.26.</span> Summary  Phenotyping Via Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-26">
<ul class="org-ul">
<li>Tensor factorization = unsupervised, interpretable phenotyping.</li>
<li>Phenotypes help in prediction (e.g., heart failure).</li>
<li>Benefits: no expert supervision needed, compact representations.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgf3d6086" class="outline-2">
<h2 id="orgf3d6086"><span class="section-number-2">13.</span> Patient Similarity</h2>
<div class="outline-text-2" id="text-13">
</div>
<div id="outline-container-orgbcbc22f" class="outline-3">
<h3 id="orgbcbc22f"><span class="section-number-3">13.1.</span> Introduction to Patient Similarity</h3>
<div class="outline-text-3" id="text-13-1">
<ul class="org-ul">
<li>Patient similarity introduces a new paradigm in medical practice.</li>
<li>Enables identifying similar patients for specific clinical contexts.</li>
<li>Supports pragmatic trials and practice-based medicine.</li>
<li>A supervised distance metric learning algorithm is introduced.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb5d3f2d" class="outline-3">
<h3 id="orgb5d3f2d"><span class="section-number-3">13.2.</span> Motivation</h3>
<div class="outline-text-3" id="text-13-2">
<ul class="org-ul">
<li>Traditional medicine relies on randomized clinical trials (RCT) for evidence (evidence-based medicine).</li>
<li>Limitations: patient heterogeneity, outdated/inapplicable guidelines.</li>
<li>EHR data enables a new paradigm: precision medicine.</li>
<li>Personalized care requires ability to measure patient similarity.</li>
</ul>
</div>
</div>

<div id="outline-container-org39c18c2" class="outline-3">
<h3 id="org39c18c2"><span class="section-number-3">13.3.</span> Traditional Paradigm</h3>
<div class="outline-text-3" id="text-13-3">
<ul class="org-ul">
<li>Evidence-based medicine uses:
<ul class="org-ul">
<li>RCTs to test hypotheses.</li>
<li>Publications and drug approvals.</li>
<li>Expert-created clinical guidelines.</li>
<li>Guideline application in clinical practice.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgd304aff" class="outline-3">
<h3 id="orgd304aff"><span class="section-number-3">13.4.</span> New Paradigm</h3>
<div class="outline-text-3" id="text-13-4">
<ul class="org-ul">
<li>Precision medicine promotes personalized care through:
<ul class="org-ul">
<li>Pragmatic trials using EHR data.</li>
<li>Patient similarity searches to inform treatment.</li>
<li>Practice-based evidence derived from real-world data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org4bd0f0a" class="outline-3">
<h3 id="org4bd0f0a"><span class="section-number-3">13.5.</span> Randomized Clinical Trials</h3>
<div class="outline-text-3" id="text-13-5">
<ul class="org-ul">
<li>Involves randomly assigning patients to control or treatment groups.</li>
<li>Compares average outcomes to assess treatment effectiveness.</li>
<li>Considered successful if treatment group shows better outcomes.</li>
</ul>
</div>
</div>

<div id="outline-container-org83773af" class="outline-3">
<h3 id="org83773af"><span class="section-number-3">13.6.</span> RCT Quiz Question</h3>
<div class="outline-text-3" id="text-13-6">
<ul class="org-ul">
<li>Asks about drawbacks of RCTs.</li>
<li>Lists controlled environment, testing one thing at a time, costs, etc.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb1b32f8" class="outline-3">
<h3 id="orgb1b32f8"><span class="section-number-3">13.7.</span> RCT Quiz Solution</h3>
<div class="outline-text-3" id="text-13-7">
<ul class="org-ul">
<li>RCTs:
<ul class="org-ul">
<li>Require controlled environments with selective populations.</li>
<li>Test only one intervention at a time.</li>
<li>Are expensive and time-consuming.</li>
<li>Discover causal relationships (not a drawback).</li>
<li>Do not typically deal with noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org2eee673" class="outline-3">
<h3 id="org2eee673"><span class="section-number-3">13.8.</span> Pragmatic Trials</h3>
<div class="outline-text-3" id="text-13-8">
<ul class="org-ul">
<li>Measure real-world treatment effectiveness using EHR data.</li>
<li>Use similarity searches to group patients by treatment and outcomes.</li>
<li>Aim to recommend best-performing treatments to current patients.</li>
<li>Reflect real-world patient variations and support individual care choices.</li>
</ul>
</div>
</div>

<div id="outline-container-orge02ff4d" class="outline-3">
<h3 id="orge02ff4d"><span class="section-number-3">13.9.</span> Pragmatic Trial Quiz Question</h3>
<div class="outline-text-3" id="text-13-9">
<ul class="org-ul">
<li>Questions benefits and limitations of pragmatic trials.</li>
</ul>
</div>
</div>

<div id="outline-container-org4d6a1fa" class="outline-3">
<h3 id="org4d6a1fa"><span class="section-number-3">13.10.</span> Pragmatic Trial Quiz Solution</h3>
<div class="outline-text-3" id="text-13-10">
<ul class="org-ul">
<li>Pragmatic trials:
<ul class="org-ul">
<li>Can't test new drugs.</li>
<li>Operate in real-world settings.</li>
<li>Can gather data continuously.</li>
<li>Are less costly and faster than RCTs.</li>
<li>Don't establish causality (no randomization).</li>
<li>Deal with noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org3cdb4d5" class="outline-3">
<h3 id="org3cdb4d5"><span class="section-number-3">13.11.</span> Utilize Patient Similarity</h3>
<div class="outline-text-3" id="text-13-11">
<ul class="org-ul">
<li>Practice-based medicine uses patient similarity to:
<ul class="org-ul">
<li>Generate hypotheses from retrospective evidence.</li>
<li>Confirm them through RCTs.</li>
<li>Update clinical guidelines.</li>
</ul></li>
<li>Decision process:
<ul class="org-ul">
<li>Apply guidelines if available.</li>
<li>Use similar patient data if guidelines aren't applicable.</li>
<li>Use professional judgment if similar data is lacking.</li>
</ul></li>
<li>Paradigms are complementary (evidence-based + practice-based).</li>
</ul>
</div>
</div>

<div id="outline-container-orgb03a606" class="outline-3">
<h3 id="orgb03a606"><span class="section-number-3">13.12.</span> Patient Similarity Approaches</h3>
<div class="outline-text-3" id="text-13-12">
<ul class="org-ul">
<li>Distance metric learning: supervised approach using feature vectors and labels.</li>
<li>Learn distance function to reflect similarity.</li>
<li>Graph-based methods use medical ontologies/disease networks to relate patients.</li>
</ul>
</div>
</div>

<div id="outline-container-org710b6fd" class="outline-3">
<h3 id="org710b6fd"><span class="section-number-3">13.13.</span> Patient Similarity Through LSML</h3>
<div class="outline-text-3" id="text-13-13">
<ul class="org-ul">
<li>LSML (Locally Supervised Metric Learning):
<ul class="org-ul">
<li>Learns a distance metric tailored to clinical context.</li>
<li>Uses baseline similarity to retrieve potential matches.</li>
<li>Refines metric to bring similar patients closer and dissimilar ones further.</li>
</ul></li>
<li>Objective: maximize margin between homogeneous and heterogeneous neighbors.</li>
<li>Mathematical formulation:
<ul class="org-ul">
<li>Learn a low-rank symmetric matrix W (via Mahalanobis distance).</li>
<li>Maximize trace(WᵀHW) where H captures margin information.</li>
<li>Solved via eigenvalue decomposition.</li>
</ul></li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org1a80890" class="outline-2">
<h2 id="org1a80890"><span class="section-number-2">14.</span> Deep Neural Network</h2>
<div class="outline-text-2" id="text-14">
</div>
<div id="outline-container-org703cee1" class="outline-3">
<h3 id="org703cee1"><span class="section-number-3">14.1.</span> Backward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-1">
<ul class="org-ul">
<li>A neuron processes inputs \( x_1, x_2, \dots, x_n \) with corresponding weights and a bias term to output a value \( y \).</li>
<li>Computation involves:
<ul class="org-ul">
<li>Linear combination: \( z = \sum w_i x_i + b \)</li>
<li>Nonlinear activation: \( y = g(z) \)</li>
</ul></li>
<li>Output type:
<ul class="org-ul">
<li>Binary (for classification) or continuous (for regression)</li>
</ul></li>
<li>Learning involves:
<ul class="org-ul">
<li>Specifying the activation function \( g \)</li>
<li>Learning weights and bias from data</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf59b2c5" class="outline-3">
<h3 id="orgf59b2c5"><span class="section-number-3">14.2.</span> Sigmoid Function</h3>
<div class="outline-text-3" id="text-14-2">
<ul class="org-ul">
<li>Nonlinear transformation specified by the modeler (not learned from data).</li>
<li>Popular activation functions: sigmoid, tanh, ReLU.</li>
<li>Sigmoid function: \( \sigma(x) = \frac{1}{1 + e^{-x}} \)</li>
<li>Output range: [0, 1]; interpretable as probability</li>
<li>Common for classification tasks</li>
<li>Problem: vanishing gradients when \( x \) is very small or very large</li>
</ul>
</div>
</div>

<div id="outline-container-org26f5bfc" class="outline-3">
<h3 id="org26f5bfc"><span class="section-number-3">14.3.</span> TANH Function</h3>
<div class="outline-text-3" id="text-14-3">
<ul class="org-ul">
<li>Defined as: \( \tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}} = 2\sigma(2x) - 1 \)</li>
<li>Output range: [–1, 1]; centered at 0</li>
<li>Larger gradients near zero than sigmoid</li>
<li>Still suffers from vanishing gradient when input is far from 0</li>
</ul>
</div>
</div>

<div id="outline-container-orgd344a12" class="outline-3">
<h3 id="orgd344a12"><span class="section-number-3">14.4.</span> Rectified Linear Function</h3>
<div class="outline-text-3" id="text-14-4">
<ul class="org-ul">
<li>ReLU: \( \max(0, x) \)</li>
<li>Output: [0, ∞); linear for \( x > 0 \), 0 otherwise</li>
<li>Does not suffer from vanishing gradient (gradient = 1 for \( x > 0 \))</li>
</ul>
</div>
</div>

<div id="outline-container-orge2a1e4a" class="outline-3">
<h3 id="orge2a1e4a"><span class="section-number-3">14.5.</span> Activation Functions_ Summary</h3>
<div class="outline-text-3" id="text-14-5">
<ul class="org-ul">
<li>Sigmoid and tanh are bounded; ReLU is unbounded above.</li>
<li>Choice of activation function is application dependent.</li>
</ul>
</div>
</div>

<div id="outline-container-org8f9e8f5" class="outline-3">
<h3 id="org8f9e8f5"><span class="section-number-3">14.6.</span> Train a Single Neuron</h3>
<div class="outline-text-3" id="text-14-6">
<ul class="org-ul">
<li>Goal: Learn weights and bias to minimize loss</li>
<li>Loss function measures difference between prediction \( y \) and target \( t \)</li>
<li>Example: squared Euclidean loss</li>
<li>Training = minimizing loss via optimization</li>
</ul>
</div>
</div>

<div id="outline-container-org5a4df1d" class="outline-3">
<h3 id="org5a4df1d"><span class="section-number-3">14.7.</span> Stochastic Gradient Descent</h3>
<div class="outline-text-3" id="text-14-7">
<ul class="org-ul">
<li>Efficient method to train neurons</li>
<li>Procedure:
<ol class="org-ol">
<li>Initialize weights and bias randomly</li>
<li>For each example:
<ul class="org-ul">
<li>Compute gradient of loss w.r.t. weights and bias</li>
<li>Update parameters using gradient and learning rate \( \eta \)</li>
</ul></li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-orga65194e" class="outline-3">
<h3 id="orga65194e"><span class="section-number-3">14.8.</span> Forward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-8">
<ul class="org-ul">
<li>Two-pass process: forward to compute \( y \), backward for gradients</li>
<li>Derivatives:
<ul class="org-ul">
<li>\( \frac{\partial z}{\partial w_i} = x_i \)</li>
<li>\( \frac{\partial y}{\partial z} = y(1 - y) \) (for sigmoid)</li>
</ul></li>
<li>Software often automates this process</li>
</ul>
</div>
</div>

<div id="outline-container-org80f90d6" class="outline-3">
<h3 id="org80f90d6"><span class="section-number-3">14.9.</span> Backward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-9">
<ul class="org-ul">
<li>Use chain rule to compute gradients:
<ul class="org-ul">
<li>\( \frac{\partial L}{\partial w_i} = (y - t) \cdot y(1 - y) \cdot x_i \)</li>
<li>\( \frac{\partial L}{\partial b} = (y - t) \cdot y(1 - y) \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orge581478" class="outline-3">
<h3 id="orge581478"><span class="section-number-3">14.10.</span> Train a Multilayer Neural Network</h3>
<div class="outline-text-3" id="text-14-10">
<ul class="org-ul">
<li>Similar process as single neuron</li>
<li>SGD updates weights and biases for each layer</li>
<li>Challenge: computing gradients for all layers</li>
</ul>
</div>
</div>

<div id="outline-container-org0589a95" class="outline-3">
<h3 id="org0589a95"><span class="section-number-3">14.11.</span> Forward Computation Part 1</h3>
<div class="outline-text-3" id="text-14-11">
<ul class="org-ul">
<li>Step-by-step example with 3 hidden units (h1, h2, h3) and 1 output</li>
<li>Each unit computes:
<ul class="org-ul">
<li>Linear combination: \( z \)</li>
<li>Activation: \( a = g(z) \)</li>
</ul></li>
<li>Final output is computed from hidden layer outputs</li>
</ul>
</div>
</div>

<div id="outline-container-org5c9a037" class="outline-3">
<h3 id="org5c9a037"><span class="section-number-3">14.12.</span> Forward Computation Part 2</h3>
<div class="outline-text-3" id="text-14-12">
<ul class="org-ul">
<li>Summary of previous computation</li>
<li>Network is fully connected</li>
<li>Shows the symmetry and modularity of neural networks</li>
</ul>
</div>
</div>

<div id="outline-container-orgfd5aba4" class="outline-3">
<h3 id="orgfd5aba4"><span class="section-number-3">14.13.</span> Forward Computation_ Vector Form</h3>
<div class="outline-text-3" id="text-14-13">
<ul class="org-ul">
<li>Computation represented in vector/matrix form:
<ul class="org-ul">
<li>\( z = Wx + b \)</li>
<li>\( a = g(z) \)</li>
</ul></li>
<li>Activation applied element-wise to vectors</li>
</ul>
</div>
</div>

<div id="outline-container-org53f7221" class="outline-3">
<h3 id="org53f7221"><span class="section-number-3">14.14.</span> Forward Computation_ More General Form</h3>
<div class="outline-text-3" id="text-14-14">
<ul class="org-ul">
<li>General forward pass for layer \( L \rightarrow L+1 \):
<ul class="org-ul">
<li>\( z^{(L+1)} = W^{(L)}a^{(L)} + b^{(L)} \)</li>
<li>\( a^{(L+1)} = g(z^{(L+1)}) \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc1d9a51" class="outline-3">
<h3 id="orgc1d9a51"><span class="section-number-3">14.15.</span> Forward Computation_ Summary</h3>
<div class="outline-text-3" id="text-14-15">
<ul class="org-ul">
<li>Computation can be expressed:
<ul class="org-ul">
<li>Unit-by-unit</li>
<li>Vectorized (preferred for efficiency)</li>
<li>Generalized for arbitrary depth</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orge96fd64" class="outline-3">
<h3 id="orge96fd64"><span class="section-number-3">14.16.</span> Gradient Descent for Neural Networks</h3>
<div class="outline-text-3" id="text-14-16">
<ul class="org-ul">
<li>Recap:
<ul class="org-ul">
<li>Initialize weights/biases</li>
<li>Compute gradients</li>
<li>Apply updates</li>
</ul></li>
<li>Efficient gradient computation is critical → leads to backpropagation</li>
</ul>
</div>
</div>

<div id="outline-container-org41d7fe6" class="outline-3">
<h3 id="org41d7fe6"><span class="section-number-3">14.17.</span> Backward Propagation Part 1</h3>
<div class="outline-text-3" id="text-14-17">
<ul class="org-ul">
<li>Gradient of loss w.r.t. \( w_{ji}^{(l)} \) is:
<ul class="org-ul">
<li>\( \delta_j^{(l)} \cdot a_i^{(l-1)} \)</li>
</ul></li>
<li>\( \delta_j^{(l)} \): error signal for unit j at layer l</li>
<li>Chain rule used to derive this</li>
</ul>
</div>
</div>

<div id="outline-container-org770436b" class="outline-3">
<h3 id="org770436b"><span class="section-number-3">14.18.</span> Backward Propagation Part 2</h3>
<div class="outline-text-3" id="text-14-18">
<ul class="org-ul">
<li>Gradient w.r.t. bias \( b_j^{(l)} = \delta_j^{(l)} \)</li>
<li>Bias derivatives are simpler (derivative of z w.r.t. bias = 1)</li>
</ul>
</div>
</div>

<div id="outline-container-org47e187e" class="outline-3">
<h3 id="org47e187e"><span class="section-number-3">14.19.</span> Backward Propagation Part 3</h3>
<div class="outline-text-3" id="text-14-19">
<ul class="org-ul">
<li>Need to compute all \( \delta_j^{(l)} \) efficiently</li>
<li>Done in backward pass from output to input layer</li>
</ul>
</div>
</div>

<div id="outline-container-org01d630f" class="outline-3">
<h3 id="org01d630f"><span class="section-number-3">14.20.</span> Backward Propagation Part 4</h3>
<div class="outline-text-3" id="text-14-20">
<ul class="org-ul">
<li>Compute final layer delta:
<ul class="org-ul">
<li>\( \delta_k^{(L)} = \frac{\partial L}{\partial z_k^{(L)}} = y - t \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf68ad05" class="outline-3">
<h3 id="orgf68ad05"><span class="section-number-3">14.21.</span> Backward Propagation Part 5</h3>
<div class="outline-text-3" id="text-14-21">
<ul class="org-ul">
<li>Compute delta for previous layer:
<ul class="org-ul">
<li>Chain rule involving gradient of activation and next layer delta</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org469385f" class="outline-3">
<h3 id="org469385f"><span class="section-number-3">14.22.</span> Backward Propagation Part 6</h3>
<div class="outline-text-3" id="text-14-22">
<ul class="org-ul">
<li>For hidden layers:
<ul class="org-ul">
<li>Delta includes:
<ul class="org-ul">
<li>Sum over deltas from next layer</li>
<li>Corresponding weights</li>
<li>Derivative of activation at current layer</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf0dc753" class="outline-3">
<h3 id="orgf0dc753"><span class="section-number-3">14.23.</span> Backward Propagation Summary</h3>
<div class="outline-text-3" id="text-14-23">
<ul class="org-ul">
<li>General form of delta:
<ul class="org-ul">
<li>\( \delta_j^{(l)} = g'(z_j^{(l)}) \sum_k w_{kj}^{(l)} \delta_k^{(l+1)} \)</li>
</ul></li>
<li>Backpropagation allows computing all gradients in one pass</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org36895ef" class="outline-2">
<h2 id="org36895ef"><span class="section-number-2">15.</span> Convolutional Neural Network</h2>
<div class="outline-text-2" id="text-15">
</div>
<div id="outline-container-orgb3b133c" class="outline-3">
<h3 id="orgb3b133c"><span class="section-number-3">15.1.</span> Convolution Local Networks</h3>
<div class="outline-text-3" id="text-15-1">
<ul class="org-ul">
<li>Feedforward networks connect all units between layers, causing high parameter counts for high-dimensional data.</li>
<li>Local connectivity can reduce parameters by limiting neuron connections to adjacent units.</li>
<li>Convolutional Neural Networks (CNNs) utilize this local connectivity across layers.</li>
<li><b><b>Weight sharing</b></b>:
<ul class="org-ul">
<li>Reuses the same parameters across locations, further reducing total parameters.</li>
<li>Inspired by convolution in signal processing.</li>
<li>Example: Sharing weights reduces 6 parameters to 3 for two hidden units.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgb0af3aa" class="outline-3">
<h3 id="orgb0af3aa"><span class="section-number-3">15.2.</span> Pooling Handling Distortion</h3>
<div class="outline-text-3" id="text-15-2">
<ul class="org-ul">
<li>Pooling layers add <b><b>translational invariance</b></b>, crucial for images and time series.</li>
<li><b><b>Max pooling</b></b> extracts maximum values, aiding stability under input shifts.</li>
<li>Example: Two shifted inputs can yield the same output after convolution + pooling.</li>
<li>This robustness is desirable for real-world applications where input variance is common.</li>
</ul>
</div>
</div>

<div id="outline-container-org3c6c7fa" class="outline-3">
<h3 id="org3c6c7fa"><span class="section-number-3">15.3.</span> Convolutional Neural Networks</h3>
<div class="outline-text-3" id="text-15-3">
<ul class="org-ul">
<li>CNNs are effective for grid-like data (e.g., images).</li>
<li>Advantages:
<ul class="org-ul">
<li>Sparse interconnections.</li>
<li>Parameter sharing.</li>
<li>Supports translational invariance.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgb829397" class="outline-3">
<h3 id="orgb829397"><span class="section-number-3">15.4.</span> Key Stages of CNN</h3>
<div class="outline-text-3" id="text-15-4">
<ul class="org-ul">
<li>A typical CNN architecture consists of:
<ul class="org-ul">
<li>Convolution layers</li>
<li>Pooling layers</li>
<li>Fully connected layers</li>
</ul></li>
<li>Example architecture:
<ul class="org-ul">
<li>Conv → Max Pool → Conv → Max Pool → 3×Conv → Pool → 3×Fully Connected</li>
</ul></li>
<li>More detail on design choices to follow.</li>
</ul>
</div>
</div>

<div id="outline-container-org4960d75" class="outline-3">
<h3 id="org4960d75"><span class="section-number-3">15.5.</span> 1-D Convolution</h3>
<div class="outline-text-3" id="text-15-5">
<ul class="org-ul">
<li>Illustrates convolution with 1D input: `[1, 2, -1, 1, -3]` and kernel `[1, 0, -1]`.</li>
<li>Output computed via sliding window and inner product.</li>
<li>Padding used at edges (zero-padding).</li>
<li>Introduces concept of <b><b>stride</b></b> (spacing between kernel applications).</li>
<li>Computationally intensive but highly parallelizable on GPUs.</li>
</ul>
</div>
</div>

<div id="outline-container-org8db390e" class="outline-3">
<h3 id="org8db390e"><span class="section-number-3">15.6.</span> Neural Network for 1-D Convolution</h3>
<div class="outline-text-3" id="text-15-6">
<ul class="org-ul">
<li>Convolution operation visualized as a neural network:
<ul class="org-ul">
<li>Applies shared weights across sliding input windows.</li>
<li>Two outputs H1 and H2 for two positions of the kernel.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc40f458" class="outline-3">
<h3 id="orgc40f458"><span class="section-number-3">15.7.</span> Convolution Layers</h3>
<div class="outline-text-3" id="text-15-7">
<ul class="org-ul">
<li>After convolution, <b><b>nonlinear activations</b></b> (e.g., ReLU) are applied.</li>
<li>Multiple filters (kernels) are often used to capture different patterns.</li>
<li>Output becomes a 3D tensor: height × width × number of filters.</li>
<li>Filters may extract features like average values or edges.</li>
</ul>
</div>
</div>

<div id="outline-container-orgae1ec35" class="outline-3">
<h3 id="orgae1ec35"><span class="section-number-3">15.8.</span> Pooling Layers</h3>
<div class="outline-text-3" id="text-15-8">
<ul class="org-ul">
<li>Pooling reduces spatial size of feature maps.</li>
<li>Types: Max pooling (most common), average pooling, sum pooling.</li>
<li>Example: 2×2 max pooling reduces 4D input to 2D output by taking max in each region.</li>
</ul>
</div>
</div>

<div id="outline-container-orga3bf11f" class="outline-3">
<h3 id="orga3bf11f"><span class="section-number-3">15.9.</span> Dimension Calculation of Convolution Layers</h3>
<div class="outline-text-3" id="text-15-9">
<ul class="org-ul">
<li>Input: `W1 × H1 × D1`</li>
<li>Output: `W2 × H2 × D2`
<ul class="org-ul">
<li>`W2 = (W1 - F + 2P) / S + 1`</li>
<li>`H2 = (H1 - F + 2P) / S + 1`</li>
<li>`D2 = K` (number of filters)</li>
</ul></li>
<li>Illustrated with example: input `227×227×3`, filter size `11×11`, stride `4`, output `55×55×96`.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd94a853" class="outline-3">
<h3 id="orgd94a853"><span class="section-number-3">15.10.</span> Dimension Calculation of Pooling Layer</h3>
<div class="outline-text-3" id="text-15-10">
<ul class="org-ul">
<li>Similar to convolution but without padding.</li>
<li>Output size:
<ul class="org-ul">
<li>`W2 = (W1 - F) / S + 1`</li>
<li>`H2 = (H1 - F) / S + 1`</li>
<li>`D2 = D1` (depth remains the same)</li>
</ul></li>
<li>Example: pooling `3×3` on input `55×55×96` with stride 2 → output `27×27×96`.</li>
</ul>
</div>
</div>

<div id="outline-container-org49cb9be" class="outline-3">
<h3 id="org49cb9be"><span class="section-number-3">15.11.</span> Parameter Count</h3>
<div class="outline-text-3" id="text-15-11">
<ul class="org-ul">
<li>Convolution layers often dominate the depth but not the parameter count.</li>
<li>Fully connected layers have the most parameters.</li>
<li>Example:
<ul class="org-ul">
<li>Conv layers: 3.7M parameters.</li>
<li>Final 3 FC layers: 58.6M parameters.</li>
</ul></li>
<li>Highlights computational efficiency of convolution layers.</li>
</ul>
</div>
</div>

<div id="outline-container-org3c0dee4" class="outline-3">
<h3 id="org3c0dee4"><span class="section-number-3">15.12.</span> Forward Calculation of Convolution</h3>
<div class="outline-text-3" id="text-15-12">
<ul class="org-ul">
<li>Number of computations = output size × filter size.</li>
<li>Formula: `W2 × H2 × D2 × F × F × D1`</li>
<li>Necessary to estimate computational cost.</li>
</ul>
</div>
</div>

<div id="outline-container-org6236e8f" class="outline-3">
<h3 id="org6236e8f"><span class="section-number-3">15.13.</span> Forward Calculation of Pooling Layers</h3>
<div class="outline-text-3" id="text-15-13">
<ul class="org-ul">
<li>Follows similar computation logic as convolution.</li>
<li>Number of operations = output size + filter size.</li>
</ul>
</div>
</div>

<div id="outline-container-org387bb8e" class="outline-3">
<h3 id="org387bb8e"><span class="section-number-3">15.14.</span> Forward Calculation of Fully-Connected Layers</h3>
<div class="outline-text-3" id="text-15-14">
<ul class="org-ul">
<li>Computation = input size × output size.</li>
<li>Converts 3D input tensor to a vector before FC layer.</li>
</ul>
</div>
</div>

<div id="outline-container-org1ad6655" class="outline-3">
<h3 id="org1ad6655"><span class="section-number-3">15.15.</span> CNN Architectures and Healthcare Applications</h3>
<div class="outline-text-3" id="text-15-15">
<ul class="org-ul">
<li>Discusses famous CNNs:
<ul class="org-ul">
<li><b><b>AlexNet</b></b>: Large improvement in image classification (2012).</li>
<li><b><b>Inception</b></b>: Deeper with parallel paths.</li>
<li><b><b>ResNet</b></b>: Introduced <b><b>skip connections</b></b>, mitigates vanishing gradients.</li>
</ul></li>
<li>Performance on ImageNet:
<ul class="org-ul">
<li>ResNet achieves &lt;4% error, near human-level.</li>
</ul></li>
<li>Important design traits:
<ul class="org-ul">
<li>Deeper networks.</li>
<li>Parallel paths and residual learning.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org837abcc" class="outline-3">
<h3 id="org837abcc"><span class="section-number-3">15.16.</span> Diabetic Retinopathy Diagnosis</h3>
<div class="outline-text-3" id="text-15-16">
<ul class="org-ul">
<li>CNNs applied to eye images for diabetic retinopathy detection.</li>
<li>Published in JAMA.</li>
<li>Achieved performance on par with domain experts.</li>
<li>ROC curve shows CNNs matching or exceeding doctors.</li>
</ul>
</div>
</div>

<div id="outline-container-org11a908e" class="outline-3">
<h3 id="org11a908e"><span class="section-number-3">15.17.</span> Dermatologist Classification of Skin Cancer with DNN</h3>
<div class="outline-text-3" id="text-15-17">
<ul class="org-ul">
<li>CNNs used to classify skin lesions (benign vs. malignant).</li>
<li>Published in <b>Nature</b>.</li>
<li>CNN performance comparable or superior to dermatologists.</li>
<li>Visualization: algorithm ROC curve often exceeds human expert dots.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org331eb2b" class="outline-2">
<h2 id="org331eb2b"><span class="section-number-2">16.</span> Recurrent Neural Network</h2>
<div class="outline-text-2" id="text-16">
</div>
<div id="outline-container-orgeee34ff" class="outline-3">
<h3 id="orgeee34ff"><span class="section-number-3">16.1.</span> Basic Concepts of RNN</h3>
<div class="outline-text-3" id="text-16-1">
<ul class="org-ul">
<li>Feedforward neural networks (FFNN):
<ul class="org-ul">
<li>Map input vector X → hidden state H → output Ŷ.</li>
<li>X can be a multi-hot vector of patient diseases.</li>
<li>H represents patient embedding; Ŷ is prediction (e.g. heart failure).</li>
</ul></li>
<li>Recurrent Neural Networks (RNN):
<ul class="org-ul">
<li>Similar structure to FFNN, with key difference: recursion in hidden state.</li>
<li>Hidden state H = function of input X and previous hidden state H.</li>
<li>Components:
<ul class="org-ul">
<li>U: weights from input to hidden</li>
<li>W: weights from previous hidden to current hidden</li>
<li>V: weights from hidden to output</li>
<li>f, g: activation functions</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org97ebede" class="outline-3">
<h3 id="org97ebede"><span class="section-number-3">16.2.</span> Basic RNN Structure</h3>
<div class="outline-text-3" id="text-16-2">
<ul class="org-ul">
<li>Unfolding RNN over time steps (t-1, t, t+1).</li>
<li>Each time step:
<ul class="org-ul">
<li>Input: Xₜ</li>
<li>Hidden state: Hₜ = f(UXₜ + WHₜ₋₁ + b₁)</li>
<li>Output: Ŷₜ = g(VHₜ + b₂)</li>
</ul></li>
<li>Parameters (U, W, V) shared across time steps.</li>
<li>Common RNN architectures:
<ul class="org-ul">
<li>One-to-many: e.g., image captioning.</li>
<li>Many-to-one: e.g., disease classification from clinical notes.</li>
<li>Many-to-many:
<ul class="org-ul">
<li>Seq-to-seq: e.g., machine translation.</li>
<li>Time-aligned: e.g., sequential disease prediction.</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org0c8403b" class="outline-3">
<h3 id="org0c8403b"><span class="section-number-3">16.3.</span> Forward Computation</h3>
<div class="outline-text-3" id="text-16-3">
<ul class="org-ul">
<li>Forward pass in training:
<ol class="org-ol">
<li>Compute Zₜ = UXₜ + WHₜ₋₁ + b₁</li>
<li>Hₜ = f(Zₜ)</li>
<li>oₜ = VHₜ + b₂</li>
<li>Ŷₜ = g(oₜ)</li>
</ol></li>
<li>Loss function:
<ul class="org-ul">
<li>Negative log likelihood (NLL) for sequence output.</li>
<li>f: e.g., tanh; g: e.g., softmax.</li>
<li>Total loss = sum over time steps.</li>
<li>For binary classification, simplified loss form shown.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org15ad563" class="outline-3">
<h3 id="org15ad563"><span class="section-number-3">16.4.</span> Backpropagation Through Time</h3>
<div class="outline-text-3" id="text-16-4">
<ul class="org-ul">
<li>Backward pass (BPTT):
<ul class="org-ul">
<li>Compute ∂L/∂oₜ, then ∂L/∂Hₜ.</li>
<li>Last time step: ∂L/∂Hₜ = Vᵀ ∂L/∂oₜ</li>
<li>Earlier steps: ∂L/∂Hₜ includes gradient through both oₜ and Hₜ₊₁.</li>
</ul></li>
<li>Parameter updates:
<ul class="org-ul">
<li>Gradients used to update U, W, V, b₁, b₂.</li>
</ul></li>
<li>Practical note:
<ul class="org-ul">
<li>Most modern libraries automate this.</li>
</ul></li>
<li>Vanishing gradient problem:
<ul class="org-ul">
<li>Gradients diminish through long sequences.</li>
<li>Makes learning long-term dependencies difficult.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org9de46bd" class="outline-3">
<h3 id="org9de46bd"><span class="section-number-3">16.5.</span> Standard RNN</h3>
<div class="outline-text-3" id="text-16-5">
<ul class="org-ul">
<li>Standard RNN cell:
<ul class="org-ul">
<li>Hₜ = tanh(WXₜ + UHₜ₋₁ + b)</li>
</ul></li>
<li>Suffers from vanishing gradient problem on long sequences.</li>
<li>Motivation for improved architectures like LSTM and GRU.</li>
</ul>
</div>
</div>

<div id="outline-container-org4ce52cb" class="outline-3">
<h3 id="org4ce52cb"><span class="section-number-3">16.6.</span> Gated Recurred Unit</h3>
<div class="outline-text-3" id="text-16-6">
<ul class="org-ul">
<li>GRU: simpler alternative to LSTM with good performance.</li>
<li>Key components:
<ul class="org-ul">
<li>Update gate (Zₜ): controls how much to keep from old H.</li>
<li>Reset gate (Rₜ): controls influence of old H on candidate state.</li>
</ul></li>
<li>Hidden state update:
<ul class="org-ul">
<li>New Hₜ = (1 - Zₜ) ∘ Hₜ₋₁ + Zₜ ∘ H̃ₜ</li>
</ul></li>
<li>Uses fewer parameters than LSTM, performs comparably in practice.</li>
</ul>
</div>
</div>

<div id="outline-container-org0fef65a" class="outline-3">
<h3 id="org0fef65a"><span class="section-number-3">16.7.</span> Bidirectional RNN</h3>
<div class="outline-text-3" id="text-16-7">
<ul class="org-ul">
<li>BiRNN = two RNNs:
<ul class="org-ul">
<li>One reads sequence forward, other backward.</li>
<li>Outputs are concatenated for each time step.</li>
</ul></li>
<li>Benefits:
<ul class="org-ul">
<li>Provides full context from both past and future.</li>
<li>Especially useful in NLP (e.g., semantic understanding).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org325bc7e" class="outline-3">
<h3 id="org325bc7e"><span class="section-number-3">16.8.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-16-8">
<ul class="org-ul">
<li>Application: Doctor AI (2016)
<ul class="org-ul">
<li>Predicts future patient events using RNNs.</li>
<li>Input: sequence of clinical visits (multi-hot encoded).</li>
<li>Output: next visit’s diagnosis codes.</li>
</ul></li>
<li>Dataset: 260,000 patients, 10 years of data, 38k unique codes.
<ul class="org-ul">
<li>Sparse input vectors; ~10–15 nonzero values per visit.</li>
<li>Output space: 1183 diagnosis codes.</li>
</ul></li>
<li>Evaluation:
<ul class="org-ul">
<li>Top-k recall: measures how many true labels are in top-k predictions.</li>
<li>RNN outperformed baselines (heuristics, logistic regression).</li>
</ul></li>
<li>Transfer learning:
<ul class="org-ul">
<li>Pretrained RNNs on large datasets can improve performance on smaller hospital datasets.
<ul class="org-ul">
<li>Warm-start initialization leads to better generalization.</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgd19363f" class="outline-2">
<h2 id="orgd19363f"><span class="section-number-2">17.</span> Focus</h2>
<div class="outline-text-2" id="text-17">
<blockquote>
<p>
Concepts, ideas, and algorithms mentioned in lectures, labs, and assignments.
</p>

<p>
Mostly conceptual level questions, the remaining = simple computation/coding syntax/complexity questions.
</p>

<p>
Likely topics: MapReduce, Hadoop, Spark, Scala, KNN, Kmeans, random forest, EM, CNN, linear regression, logistics regression.
</p>
</blockquote>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: W</p>
<p class="date">Created: 2025-03-31 Mon 09:58</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
