<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2025-03-29 Sat 11:52 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>BD4H</title>
<meta name="author" content="W" />
<meta name="generator" content="Org Mode" />
<link rel="stylesheet" type="text/css" href="../src/readtheorg_theme/css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="../src/readtheorg_theme/css/readtheorg.css"/>
<script type="text/javascript" src="../src/lib/js/jquery.min.js"></script>
<script type="text/javascript" src="../src/lib/js/bootstrap.min.js"></script>
<script type="text/javascript" src="../src/lib/js/jquery.stickytableheaders.min.js"></script>
<script type="text/javascript" src="../src/readtheorg_theme/js/readtheorg.js"></script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content" class="content">
<h1 class="title">BD4H</h1>
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#org1f7db0f">1. Introduction to Big Data</a>
<ul>
<li><a href="#org74383b9">1.1. Introduction</a></li>
<li><a href="#org4199f86">1.2. About BDAH</a></li>
<li><a href="#org5b60c92">1.3. Learning Goals</a></li>
<li><a href="#org197c55b">1.4. Current Problems in Healthcare</a></li>
<li><a href="#orgcdb753a">1.5. The Four Vs</a></li>
<li><a href="#orgbb1185c">1.6. More About The Four Vs</a></li>
<li><a href="#orge39bfa6">1.7. Data Science is Sexy</a></li>
<li><a href="#orgaf6a4d0">1.8. BDAH Quiz 1 Question</a></li>
<li><a href="#orgbe13993">1.9. BDAH Quiz 1 Solution</a></li>
<li><a href="#orgcf983d3">1.10. BDAH Quiz 2 Question</a></li>
<li><a href="#orgd2c8c75">1.11. BDAH Quiz 2 Solution</a></li>
<li><a href="#orgf35f567">1.12. BDAH Quiz 3 Question</a></li>
<li><a href="#org914d725">1.13. BDAH Quiz 3 Solution</a></li>
</ul>
</li>
<li><a href="#orgedbc918">2. Big Data Course Overview</a>
<ul>
<li><a href="#orgf18a92e">2.1. Introduction to Course Overview</a></li>
<li><a href="#orge399b0e">2.2. Big Data Big Picture</a></li>
<li><a href="#org1e2fde5">2.3. Healthcare Applications</a></li>
<li><a href="#org8b87c25">2.4. Predictive Modeling Quiz Question</a></li>
<li><a href="#org98389b4">2.5. Predictive Modeling Quiz Solution</a></li>
<li><a href="#org000ee58">2.6. Predictive Modeling Challenges</a></li>
<li><a href="#orgc398c97">2.7. Computational Phenotyping</a></li>
<li><a href="#org17c7f98">2.8. Computational Phenotyping Quiz Question</a></li>
<li><a href="#orgd9b8106">2.9. Computational Phenotyping Quiz Solution</a></li>
<li><a href="#org4439a60">2.10. Phenotyping Algorithm</a></li>
<li><a href="#org531b539">2.11. Patient Similarity Quiz Question</a></li>
<li><a href="#org6e9d9ed">2.12. Patient Similarity Quiz Solution</a></li>
<li><a href="#org07ab028">2.13. Patient Similarity</a></li>
<li><a href="#org4eeb7ec">2.14. Algorithms</a></li>
<li><a href="#org5f56817">2.15. Systems</a></li>
<li><a href="#org2d06e9b">2.16. Summary</a></li>
</ul>
</li>
<li><a href="#org1c90d4b">3. Predictive Modeling</a>
<ul>
<li><a href="#org98658b5">3.1. Introduction to Predictive Modeling</a></li>
<li><a href="#orgabd0ae8">3.2. Predictive Modeling vs EHR</a></li>
<li><a href="#orgcecd22d">3.3. Predictive Modeling Pipeline</a></li>
<li><a href="#orgde0d506">3.4. Prediction Target</a></li>
<li><a href="#org372ee0c">3.5. Heart Failure Quiz</a></li>
<li><a href="#org2add229">3.6. Motivations for Early Detection</a></li>
<li><a href="#orgd2a6fbf">3.7. Cohort Construction</a></li>
<li><a href="#org27c6794">3.8. Prospective vs. Retrospective Studies</a></li>
<li><a href="#org2b4d6be">3.9. Prospective vs. Retrospective Quiz</a></li>
<li><a href="#org1f302a6">3.10. Cohort Study</a></li>
<li><a href="#orgfad8e76">3.11. Case-Control Study</a></li>
<li><a href="#orgb96d78f">3.12. Feature Construction</a></li>
<li><a href="#org86f749e">3.13. Feature Construction Quizzes</a></li>
<li><a href="#org43cbd0b">3.14. Prediction Performance on Windows</a></li>
<li><a href="#orgf35498b">3.15. Feature Selection</a></li>
<li><a href="#org301b471">3.16. Predictive Model</a></li>
<li><a href="#org817d346">3.17. Performance Evaluation</a></li>
<li><a href="#orgaa3fec5">3.18. Cross-Validation</a></li>
<li><a href="#orgc5618f5">3.19. Conclusion</a></li>
</ul>
</li>
<li><a href="#org8d1c04f">4. MapReduce</a>
<ul>
<li><a href="#org83be4fb">4.1. Introduction to MapReduce</a></li>
<li><a href="#org68f3374">4.2. What is MapReduce</a></li>
<li><a href="#org8385004">4.3. Computational Process</a></li>
<li><a href="#orge7cabf1">4.4. Learning Via Aggregation Statistics</a></li>
<li><a href="#org57b1d62">4.5. MapReduce Abstraction</a></li>
<li><a href="#org3bb834c">4.6. MapReduce System</a></li>
<li><a href="#org8fc267b">4.7. MapReduce Fault Recovery</a></li>
<li><a href="#org8e5e268">4.8. Distributed File Systems</a></li>
<li><a href="#org976c247">4.9. MapReduce Design Choice</a></li>
<li><a href="#orgb4f2e47">4.10. Analytics with MapReduce</a></li>
<li><a href="#org58701a3">4.11. MapReduce KNN</a></li>
<li><a href="#org7c68944">4.12. Linear Regression</a></li>
<li><a href="#org7e31046">4.13. MapReduce for Linear Regression Quiz Question</a></li>
<li><a href="#orgabab1f5">4.14. MapReduce for Linear Regression Quiz Solution</a></li>
<li><a href="#org32fa747">4.15. Limitations of MapReduce</a></li>
<li><a href="#orgac0616d">4.16. MapReduce Summary Quiz Question</a></li>
<li><a href="#orgc78ca8b">4.17. MapReduce Summary Quiz Solution</a></li>
</ul>
</li>
<li><a href="#orge4ae145">5. Classification Model Metrics</a>
<ul>
<li><a href="#orgf226510">5.1. Predictive Model review</a></li>
<li><a href="#orgedefc00">5.2. Confusion matrix</a></li>
<li><a href="#orgc283160">5.3. Accuracy metrics</a>
<ul>
<li><a href="#orgbcf10f7">5.3.1. Other notes</a></li>
</ul>
</li>
<li><a href="#orgc32b30d">5.4. Predictive metrics</a></li>
<li><a href="#orge5832cf">5.5. F1 score</a></li>
<li><a href="#org016e156">5.6. Classifier quiz</a></li>
<li><a href="#orgf10fb0c">5.7. Reversing predictions</a></li>
<li><a href="#org75aab64">5.8. Receiver operating characteristic</a></li>
<li><a href="#orgf6f12b6">5.9. "Best classifier threshold" quiz</a></li>
<li><a href="#orgac6e6e5">5.10. Regression metrics</a>
<ul>
<li><a href="#org0efdf41">5.10.1. \(R^2\)</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#org86ce29a">6. Ensemble methods</a>
<ul>
<li><a href="#org6ca7412">6.1. Gradient Descent Method (GDM) for Linear Regression</a></li>
<li><a href="#org1c3b667">6.2. Stochastic Gradient Descent (SGD) Method</a></li>
<li><a href="#orgbe1c3cc">6.3. SGD for Linear Regression</a></li>
<li><a href="#orgc830d93">6.4. Ensemble Method Pt 1</a></li>
<li><a href="#orge9adf2e">6.5. Ensemble Method Pt 2</a></li>
<li><a href="#org17388b0">6.6. Bias Variance Tradeoff</a></li>
<li><a href="#org8215a54">6.7. Bias Variance Tradeoff Quiz Question</a></li>
<li><a href="#org6ba8eda">6.8. Bias Variance Tradeoff Quiz Solution</a></li>
<li><a href="#org24d6c20">6.9. Bias Variance Tradeoff Quiz 2 Question</a></li>
<li><a href="#orgf28f988">6.10. Bias Variance Tradeoff Quiz 2 Solution</a></li>
<li><a href="#org86cf9b3">6.11. Bagging</a></li>
<li><a href="#org4e887a7">6.12. Random Forest</a></li>
<li><a href="#org4f26b0a">6.13. Why Bagging Works</a></li>
<li><a href="#org9600e34">6.14. Boosting</a></li>
<li><a href="#org6189eb8">6.15. Bagging vs Boosting Quiz Question</a></li>
<li><a href="#org7a9cf9a">6.16. Bagging vs Boosting Quiz Solution</a></li>
<li><a href="#org17a1633">6.17. Summary for Ensemble Methods</a></li>
<li><a href="#orgacf2391">6.18. Introduction to Ensemble Method</a></li>
<li><a href="#org244c378">6.19. Gradient Descent Method</a></li>
</ul>
</li>
<li><a href="#orgb34656a">7. Computational Phenotyping</a>
<ul>
<li><a href="#org8881adf">7.1. Introduction to Phenotyping</a></li>
<li><a href="#orgb772c73">7.2. Computational Phenotyping</a></li>
<li><a href="#orge2ba6f4">7.3. Phenotyping Algorithm</a></li>
<li><a href="#orga87cb9b">7.4. Applications of Phenotyping</a></li>
<li><a href="#org09351ce">7.5. Genomic Wide Association Study</a></li>
<li><a href="#org7407d71">7.6. Why Do We Care About Phenotyping</a></li>
<li><a href="#org288d3fa">7.7. Clinical Predictive Modeling</a></li>
<li><a href="#org6eadc1d">7.8. Pragmatic Clinical Trials</a></li>
<li><a href="#orgc590ecb">7.9. Healthcare Quality Measurement</a></li>
<li><a href="#org3bea276">7.10. Phenotyping Methods Part 1</a></li>
<li><a href="#org9a0712f">7.11. Phenotyping Methods Part 2</a></li>
<li><a href="#orgda607fe">7.12. Phenotyping Quiz Question</a></li>
<li><a href="#org951b4c6">7.13. Phenotyping Quiz Solution</a></li>
</ul>
</li>
<li><a href="#org23eb419">8. Clustering</a>
<ul>
<li><a href="#orgee4c1ec">8.1. Introduction to Clustering</a></li>
<li><a href="#orgb2132f6">8.2. Healthcare Applications</a></li>
<li><a href="#orgc9fe745">8.3. What is Clustering</a></li>
<li><a href="#orgabc499d">8.4. Algorithm Overview</a></li>
<li><a href="#orgd0d30b1">8.5. K Means</a></li>
<li><a href="#org98ac64b">8.6. K Means Quiz Question</a></li>
<li><a href="#org0609258">8.7. K Means Quiz Solution</a></li>
<li><a href="#orgc15e933">8.8. Hierarchical Clustering</a></li>
<li><a href="#org8d35181">8.9. Agglomerative Clustering</a></li>
<li><a href="#org8a18607">8.10. Gaussian Mixture Model</a></li>
<li><a href="#org8e70134">8.11. GMM Expectation Maximization</a></li>
<li><a href="#org7c856e5">8.12. GMM Steps</a></li>
<li><a href="#orgc768231">8.13. GMM Visual Illustration</a></li>
<li><a href="#org41d33b7">8.14. K Means Vs GMM</a></li>
<li><a href="#orgf492ce1">8.15. Mini Batch K Means</a></li>
<li><a href="#org48f28dc">8.16. Mini Batch K Means Quiz Question</a></li>
<li><a href="#org7b11b77">8.17. Mini Batch K Means Quiz Solution</a></li>
<li><a href="#org58b8401">8.18. DBScan</a></li>
<li><a href="#org9a8f114">8.19. DBScan Key Concepts</a></li>
<li><a href="#orgfc8d9c5">8.20. DBScan Algorithm</a></li>
<li><a href="#org0706ca1">8.21. DBScan Example</a></li>
<li><a href="#org32ec1d5">8.22. DBScan Quiz Question</a></li>
<li><a href="#org28a0899">8.23. DBScan Quiz Solution</a></li>
<li><a href="#orge5a4e6a">8.24. Clustering Evaluation Metrics</a></li>
<li><a href="#org51148f0">8.25. Rand Index</a></li>
<li><a href="#org92faca6">8.26. Mutual Information</a></li>
<li><a href="#org2656fb0">8.27. Summary of RI and MI</a></li>
<li><a href="#orge75fc57">8.28. Silhouette Coefficient</a></li>
<li><a href="#org8f106b9">8.29. Silhouette Coefficient Pros and Cons</a></li>
</ul>
</li>
<li><a href="#orgc6c26a3">9. Spark</a>
<ul>
<li><a href="#orgd030e5f">9.1. Introduction to Spark</a></li>
<li><a href="#org8ce1ea5">9.2. Environment</a></li>
<li><a href="#org00714f1">9.3. Motivation</a></li>
<li><a href="#org56d5fc8">9.4. Iteration in Map Reduce</a></li>
<li><a href="#orgbc7767f">9.5. Workload Illustration</a></li>
<li><a href="#org8cd0668">9.6. Challenge</a></li>
<li><a href="#orgf639a9e">9.7. Solution  RDDs</a></li>
<li><a href="#org9807cd6">9.8. RDD Recovery</a></li>
<li><a href="#orga866770">9.9. Spark Stack</a></li>
<li><a href="#orge636dfc">9.10. Spark Programming Interface</a></li>
<li><a href="#org4bf6605">9.11. RDD Transformations</a></li>
<li><a href="#orgdc39326">9.12. RDD Transformations Quiz Question</a></li>
<li><a href="#org4f34576">9.13. RDD Transformations Quiz Solution</a></li>
<li><a href="#orgc2adbbf">9.14. Spark Operations</a></li>
<li><a href="#org4d2da13">9.15. Shared Variable</a></li>
<li><a href="#orgf7e2505">9.16. Fault Tolerance</a></li>
<li><a href="#orgcbb1d0d">9.17. Example  Logistic Regression</a></li>
<li><a href="#org74b871a">9.18. Example  Disease Risk Prediction</a></li>
<li><a href="#orgb6cca43">9.19. Serial ALS</a></li>
<li><a href="#orge27a33c">9.20. Naive Spark ALS</a></li>
</ul>
</li>
<li><a href="#orgc70f685">10. Medical Ontology</a>
<ul>
<li><a href="#org8ae7750">10.1. Introduction to Medical Ontology</a></li>
<li><a href="#org07d71b8">10.2. Health Data Standards</a></li>
<li><a href="#orgf217b0f">10.3. ICD</a></li>
<li><a href="#org95492e6">10.4. ICD 9 to ICD 10 Mapping</a></li>
<li><a href="#org048df9e">10.5. ICD 9 Quiz Question</a></li>
<li><a href="#org13aa6f8">10.6. ICD 9 Quiz Solution</a></li>
<li><a href="#org2046b6b">10.7. ICD Code Quiz Question</a></li>
<li><a href="#org814542c">10.8. ICD Code Quiz Solution</a></li>
<li><a href="#org3431a3d">10.9. CPT</a></li>
<li><a href="#org43bc6e8">10.10. CPT Code Quiz Question</a></li>
<li><a href="#orgcb42e54">10.11. CPT Code Quiz Solution</a></li>
<li><a href="#org554e698">10.12. LOINC</a></li>
<li><a href="#orgb1c3ddd">10.13. LOINC Code Quiz Question</a></li>
<li><a href="#org48eddea">10.14. LOINC Code Quiz Solution</a></li>
<li><a href="#orgfc4405b">10.15. NDC</a></li>
<li><a href="#orgb9ae90e">10.16. NDC Code Quiz Question</a></li>
<li><a href="#orgded009e">10.17. NDC Code Quiz Solution</a></li>
<li><a href="#orgf989e85">10.18. SNOMED</a></li>
<li><a href="#org5b77bdb">10.19. Logical Model of SNOMED CT</a></li>
<li><a href="#orgf918cbf">10.20. SNOMED Example</a></li>
<li><a href="#org9b2c032">10.21. SNOMED Relationships</a></li>
<li><a href="#org6e41fb1">10.22. SNOMED Design</a></li>
<li><a href="#org5f403b8">10.23. SNOMED Code Quiz Question</a></li>
<li><a href="#orgbb0e50b">10.24. SNOMED Code Quiz Solution</a></li>
<li><a href="#org1decd74">10.25. SNOMED Quiz Question</a></li>
<li><a href="#org3f8dfbb">10.26. SNOMED Quiz Solution</a></li>
<li><a href="#org6961461">10.27. UMLS</a></li>
<li><a href="#orgbb87bf0">10.28. Metathesaurus Concepts</a></li>
<li><a href="#org215f722">10.29. Semantic Network</a></li>
<li><a href="#orge848990">10.30. Specialist Lexicon</a></li>
</ul>
</li>
<li><a href="#org9ef6167">11. Graph Analysis</a>
<ul>
<li><a href="#org501aa1f">11.1. Introduction to Graph Analysis</a></li>
<li><a href="#org70c2c64">11.2. Agenda</a></li>
<li><a href="#org56ab366">11.3. PageRank</a></li>
<li><a href="#orgee70169">11.4. MapReduce  PageRank</a></li>
<li><a href="#orgd8e5454">11.5. PageRank Quiz Question</a></li>
<li><a href="#org37373f0">11.6. PageRank Quiz Solution</a></li>
<li><a href="#orgec0e666">11.7. Spectral Clustering</a></li>
<li><a href="#org84a8d0e">11.8. Similarity Graph Construction</a></li>
<li><a href="#orgc5407bd">11.9. E Neighborhood Graph</a></li>
<li><a href="#org4c8ba94">11.10. K Nearest Neighbor Graph</a></li>
<li><a href="#org28278e8">11.11. Fully Connected Graph</a></li>
<li><a href="#orgcf8120f">11.12. E Neighborhood Graph Quiz Question</a></li>
<li><a href="#orgfd99d3c">11.13. E Neighborhood Graph Quiz Solution</a></li>
<li><a href="#org26452e6">11.14. Spectral Clustering Algorithm</a></li>
<li><a href="#org53e5025">11.15. Big Data Conclusion</a></li>
</ul>
</li>
<li><a href="#orgf89255b">12. Dimensionality Reduction</a>
<ul>
<li><a href="#org776be0d">12.1. Introduction to Dimensionality Reduction</a></li>
<li><a href="#orgadde29c">12.2. Dimensionality Reduction</a></li>
<li><a href="#org9ae41a3">12.3. Singular Value Decomposition</a></li>
<li><a href="#org2bac861">12.4. SVD Example</a></li>
<li><a href="#orgc72e07f">12.5. SVD Properties</a></li>
<li><a href="#org2d630bb">12.6. Quiz  SVD Interpretation Question</a></li>
<li><a href="#org10de0ca">12.7. Quiz  SVD Interpretation Solution</a></li>
<li><a href="#org8c7921f">12.8. Principal Component Analysis</a></li>
<li><a href="#org02fe2cc">12.9. PCA Interpretation</a></li>
<li><a href="#orgafc643c">12.10. Sparsity Problem with SVD</a></li>
<li><a href="#orgdcb5d8f">12.11. CUR Decomposition</a></li>
<li><a href="#org4915ec5">12.12. CUR Algorithm</a></li>
<li><a href="#org4f7b716">12.13. CUR Quiz Question</a></li>
<li><a href="#org8b3e80c">12.14. CUR Quiz Solution</a></li>
<li><a href="#orgcb85f0b">12.15. Tensor for EHR</a></li>
<li><a href="#org164a943">12.16. Tensor Slicing</a></li>
<li><a href="#org73cd67f">12.17. Rank 1 Tensor</a></li>
<li><a href="#orgc3a92a5">12.18. Example Phenotype</a></li>
<li><a href="#org03feab0">12.19. Phenotyping Through Tensor Factorization</a></li>
<li><a href="#org9cfb443">12.20. CP Decomposition</a></li>
<li><a href="#orgdbc728e">12.21. Phenotyping Process Using Tensor Factorization</a></li>
<li><a href="#org210c1ac">12.22. Phenotyping for Predictive Modeling</a></li>
<li><a href="#orgf6dc06b">12.23. Predictive Performance</a></li>
<li><a href="#orgac933bb">12.24. Major Disease Phenotypes</a></li>
<li><a href="#org24d9c04">12.25. Tensor vs NMF</a></li>
<li><a href="#orgc6c7d1e">12.26. Summary  Phenotyping Via Tensor Factorization</a></li>
</ul>
</li>
<li><a href="#orge890cd6">13. Patient Similarity</a>
<ul>
<li><a href="#orgeb17e19">13.1. Introduction to Patient Similarity</a></li>
<li><a href="#orgaa1c659">13.2. Motivation</a></li>
<li><a href="#org4c6ed91">13.3. Traditional Paradigm</a></li>
<li><a href="#orgd3c0f30">13.4. New Paradigm</a></li>
<li><a href="#org46196e3">13.5. Randomized Clinical Trials</a></li>
<li><a href="#orgcfa6b9f">13.6. RCT Quiz Question</a></li>
<li><a href="#orgfef50e9">13.7. RCT Quiz Solution</a></li>
<li><a href="#orga5d4bcd">13.8. Pragmatic Trials</a></li>
<li><a href="#org80ed9ea">13.9. Pragmatic Trial Quiz Question</a></li>
<li><a href="#org7465824">13.10. Pragmatic Trial Quiz Solution</a></li>
<li><a href="#org2d1d61b">13.11. Utilize Patient Similarity</a></li>
<li><a href="#org4294595">13.12. Patient Similarity Approaches</a></li>
<li><a href="#org1e1fb5f">13.13. Patient Similarity Through LSML</a></li>
</ul>
</li>
<li><a href="#orgb9b1c82">14. Deep Neural Network</a>
<ul>
<li><a href="#orgcef6028">14.1. Backward Computation for a Neuron</a></li>
<li><a href="#orgc486986">14.2. Sigmoid Function</a></li>
<li><a href="#org9cb7cf1">14.3. TANH Function</a></li>
<li><a href="#org8cea475">14.4. Rectified Linear Function</a></li>
<li><a href="#org45385c9">14.5. Activation Functions_ Summary</a></li>
<li><a href="#org13c92b2">14.6. Train a Single Neuron</a></li>
<li><a href="#orga6e786a">14.7. Stochastic Gradient Descent</a></li>
<li><a href="#org92b16f8">14.8. Forward Computation for a Neuron</a></li>
<li><a href="#org9cbd531">14.9. Backward Computation for a Neuron</a></li>
<li><a href="#org7be7c1a">14.10. Train a Multilayer Neural Network</a></li>
<li><a href="#orge279baf">14.11. Forward Computation Part 1</a></li>
<li><a href="#orgd6223c0">14.12. Forward Computation Part 2</a></li>
<li><a href="#org53ef4c8">14.13. Forward Computation_ Vector Form</a></li>
<li><a href="#org5900422">14.14. Forward Computation_ More General Form</a></li>
<li><a href="#org7b37361">14.15. Forward Computation_ Summary</a></li>
<li><a href="#org23e6b6a">14.16. Gradient Descent for Neural Networks</a></li>
<li><a href="#org3395083">14.17. Backward Propagation Part 1</a></li>
<li><a href="#org45521e8">14.18. Backward Propagation Part 2</a></li>
<li><a href="#orgc014ec4">14.19. Backward Propagation Part 3</a></li>
<li><a href="#org87747d5">14.20. Backward Propagation Part 4</a></li>
<li><a href="#org8f3fcee">14.21. Backward Propagation Part 5</a></li>
<li><a href="#org1434215">14.22. Backward Propagation Part 6</a></li>
<li><a href="#orgbfc8198">14.23. Backward Propagation Summary</a></li>
</ul>
</li>
<li><a href="#org328e135">15. Convolutional Neural Network</a>
<ul>
<li><a href="#orgf501402">15.1. Convolution Local Networks</a></li>
<li><a href="#org02be772">15.2. Pooling Handling Distortion</a></li>
<li><a href="#orgaad049f">15.3. Convolutional Neural Networks</a></li>
<li><a href="#orgdce997e">15.4. Key Stages of CNN</a></li>
<li><a href="#org6c269c0">15.5. 1-D Convolution</a></li>
<li><a href="#orgc1adde8">15.6. Neural Network for 1-D Convolution</a></li>
<li><a href="#org467b4b1">15.7. Convolution Layers</a></li>
<li><a href="#orga5ae2af">15.8. Pooling Layers</a></li>
<li><a href="#org4533173">15.9. Dimension Calculation of Convolution Layers</a></li>
<li><a href="#org10e80c7">15.10. Dimension Calculation of Pooling Layer</a></li>
<li><a href="#org77dc5c8">15.11. Parameter Count</a></li>
<li><a href="#org8022846">15.12. Forward Calculation of Convolution</a></li>
<li><a href="#org0c5bde8">15.13. Forward Calculation of Pooling Layers</a></li>
<li><a href="#orged4147e">15.14. Forward Calculation of Fully-Connected Layers</a></li>
<li><a href="#org3a5a288">15.15. CNN Architectures and Healthcare Applications</a></li>
<li><a href="#orgcd7c08b">15.16. Diabetic Retinopathy Diagnosis</a></li>
<li><a href="#orgd9c2bfd">15.17. Dermatologist Classification of Skin Cancer with DNN</a></li>
</ul>
</li>
<li><a href="#orgbbb9693">16. Recurrent Neural Network</a>
<ul>
<li><a href="#org0c8a137">16.1. Basic Concepts of RNN</a></li>
<li><a href="#org67193ac">16.2. Basic RNN Structure</a></li>
<li><a href="#org415755a">16.3. Forward Computation</a></li>
<li><a href="#orgafdfe3a">16.4. Backpropagation Through Time</a></li>
<li><a href="#org0f2b320">16.5. Standard RNN</a></li>
<li><a href="#orgd5f7b5b">16.6. Gated Recurred Unit</a></li>
<li><a href="#orga1ec563">16.7. Bidirectional RNN</a></li>
<li><a href="#org44d6f02">16.8. Healthcare Applications</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div id="outline-container-org1f7db0f" class="outline-2">
<h2 id="org1f7db0f"><span class="section-number-2">1.</span> Introduction to Big Data</h2>
<div class="outline-text-2" id="text-1">
</div>
<div id="outline-container-org74383b9" class="outline-3">
<h3 id="org74383b9"><span class="section-number-3">1.1.</span> Introduction</h3>
<div class="outline-text-3" id="text-1-1">
<ul class="org-ul">
<li>Course: Big Data Analytics for Healthcare (BDAH)</li>
<li>Instructor: Jimeng Sun, associate professor at Georgia Tech</li>
<li>Background: Expertise in healthcare analytics and data mining, previous work at IBM TJ Watson Research Center</li>
<li>Course focus: What students will learn and why it matters</li>
</ul>
</div>
</div>

<div id="outline-container-org4199f86" class="outline-3">
<h3 id="org4199f86"><span class="section-number-3">1.2.</span> About BDAH</h3>
<div class="outline-text-3" id="text-1-2">
<ul class="org-ul">
<li>Intersection of healthcare and big data</li>
<li>Covers healthcare applications, data, analytics, and big data processing</li>
<li>Focus on how data science is applied to healthcare</li>
</ul>
</div>
</div>

<div id="outline-container-org5b60c92" class="outline-3">
<h3 id="org5b60c92"><span class="section-number-3">1.3.</span> Learning Goals</h3>
<div class="outline-text-3" id="text-1-3">
<ul class="org-ul">
<li>Understand healthcare data</li>
<li>Learn analytics algorithms</li>
<li>Work with big data systems</li>
<li>Application: Build models for disease risk prediction, treatment recommendation, patient clustering, and similarity analysis</li>
<li>Assessments: Homework using big data tools, project with system building, report writing, and presentations</li>
</ul>
</div>
</div>

<div id="outline-container-org197c55b" class="outline-3">
<h3 id="org197c55b"><span class="section-number-3">1.4.</span> Current Problems in Healthcare</h3>
<div class="outline-text-3" id="text-1-4">
<ul class="org-ul">
<li>High costs: $3.8 trillion/year in the U.S.</li>
<li>Massive waste: $765 billion/year</li>
<li>Poor quality: 200,000-400,000 preventable deaths annually</li>
<li>Preventable deaths ranked as the third leading cause of death</li>
<li>Hope: Big data can improve care and reduce costs</li>
</ul>
</div>
</div>

<div id="outline-container-orgcdb753a" class="outline-3">
<h3 id="orgcdb753a"><span class="section-number-3">1.5.</span> The Four Vs</h3>
<div class="outline-text-3" id="text-1-5">
<ul class="org-ul">
<li>Volume: Large healthcare data amounts</li>
<li>Variety: Different types of healthcare data sources</li>
<li>Velocity: Real-time data processing needs</li>
<li>Veracity: Issues with data quality (noise, missing data, errors)</li>
</ul>
</div>
</div>

<div id="outline-container-orgbb1185c" class="outline-3">
<h3 id="orgbb1185c"><span class="section-number-3">1.6.</span> More About The Four Vs</h3>
<div class="outline-text-3" id="text-1-6">
<ul class="org-ul">
<li>Examples of data volume:
<ul class="org-ul">
<li>Genomic data: 200 GB per genome</li>
<li>fMRI scan: ~300 GB per scan</li>
<li>US medical imaging data per year: ~100 petabytes</li>
</ul></li>
<li>Data variety:
<ul class="org-ul">
<li>Clinical (demographics, diagnosis, procedures, etc.)</li>
<li>Patient-generated (wearables, sensors)</li>
<li>Real-time (ICU monitoring data)</li>
</ul></li>
<li>Focus on managing diverse data types in the course</li>
</ul>
</div>
</div>

<div id="outline-container-orge39bfa6" class="outline-3">
<h3 id="orge39bfa6"><span class="section-number-3">1.7.</span> Data Science is Sexy</h3>
<div class="outline-text-3" id="text-1-7">
<ul class="org-ul">
<li>Data science is a critical field</li>
<li>Harvard Business Review: "Data Scientist, The Sexiest Job in the 21st Century"</li>
<li>Data scientists:
<ul class="org-ul">
<li>Capitalize on big data</li>
<li>Overcome technical limitations</li>
<li>Develop key tools like Hadoop and Spark</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgaf6a4d0" class="outline-3">
<h3 id="orgaf6a4d0"><span class="section-number-3">1.8.</span> BDAH Quiz 1 Question</h3>
<div class="outline-text-3" id="text-1-8">
<ul class="org-ul">
<li>True/False: A graduate-level degree is necessary to become a data scientist.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbe13993" class="outline-3">
<h3 id="orgbe13993"><span class="section-number-3">1.9.</span> BDAH Quiz 1 Solution</h3>
<div class="outline-text-3" id="text-1-9">
<ul class="org-ul">
<li>Answer: True</li>
</ul>
</div>
</div>

<div id="outline-container-orgcf983d3" class="outline-3">
<h3 id="orgcf983d3"><span class="section-number-3">1.10.</span> BDAH Quiz 2 Question</h3>
<div class="outline-text-3" id="text-1-10">
<ul class="org-ul">
<li>Question: What is the average salary of a data scientist?</li>
</ul>
</div>
</div>

<div id="outline-container-orgd2c8c75" class="outline-3">
<h3 id="orgd2c8c75"><span class="section-number-3">1.11.</span> BDAH Quiz 2 Solution</h3>
<div class="outline-text-3" id="text-1-11">
<ul class="org-ul">
<li>Answer: $120,000 per year (experienced data scientists earn over $150,000)</li>
</ul>
</div>
</div>

<div id="outline-container-orgf35f567" class="outline-3">
<h3 id="orgf35f567"><span class="section-number-3">1.12.</span> BDAH Quiz 3 Question</h3>
<div class="outline-text-3" id="text-1-12">
<ul class="org-ul">
<li>Question: What skills do data scientists need?</li>
</ul>
</div>
</div>

<div id="outline-container-org914d725" class="outline-3">
<h3 id="org914d725"><span class="section-number-3">1.13.</span> BDAH Quiz 3 Solution</h3>
<div class="outline-text-3" id="text-1-13">
<ul class="org-ul">
<li>Answer:
<ul class="org-ul">
<li>Math and statistics</li>
<li>Domain knowledge</li>
<li>Programming and databases</li>
<li>Communication and visualization</li>
</ul></li>
</ul>
</div>
</div>
</div>

<div id="outline-container-orgedbc918" class="outline-2">
<h2 id="orgedbc918"><span class="section-number-2">2.</span> Big Data Course Overview</h2>
<div class="outline-text-2" id="text-2">
</div>
<div id="outline-container-orgf18a92e" class="outline-3">
<h3 id="orgf18a92e"><span class="section-number-3">2.1.</span> Introduction to Course Overview</h3>
<div class="outline-text-3" id="text-2-1">
<ul class="org-ul">
<li>Overview of big data analytics for healthcare.</li>
<li>Topics covered:
<ul class="org-ul">
<li>Healthcare applications of big data.</li>
<li>Algorithms used in applications.</li>
<li>Software systems for implementation.</li>
</ul></li>
<li>Course structure alternates among these topics.</li>
</ul>
</div>
</div>

<div id="outline-container-orge399b0e" class="outline-3">
<h3 id="orge399b0e"><span class="section-number-3">2.2.</span> Big Data Big Picture</h3>
<div class="outline-text-3" id="text-2-2">
<ul class="org-ul">
<li>Course focuses on three key areas:
<ul class="org-ul">
<li>Big data systems.</li>
<li>Scalable machine learning algorithms.</li>
<li>Healthcare applications.</li>
</ul></li>
<li>Integration of these elements to solve healthcare problems.</li>
</ul>
</div>
</div>

<div id="outline-container-org1e2fde5" class="outline-3">
<h3 id="org1e2fde5"><span class="section-number-3">2.3.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-2-3">
<ul class="org-ul">
<li>Three types of healthcare applications:
<ol class="org-ol">
<li><b><b>Predictive Modeling</b></b>: Uses historical data to predict future outcomes.</li>
<li><b><b>Computational Phenotyping</b></b>: Converts electronic health records (EHRs) into meaningful clinical concepts.</li>
<li><b><b>Patient Similarity</b></b>: Identifies groups of patients with similar characteristics.</li>
</ol></li>
<li>Starts with predictive modeling.</li>
</ul>
</div>
</div>

<div id="outline-container-org8b87c25" class="outline-3">
<h3 id="org8b87c25"><span class="section-number-3">2.4.</span> Predictive Modeling Quiz Question</h3>
<div class="outline-text-3" id="text-2-4">
<ul class="org-ul">
<li>Predictive modeling example: predicting treatment effectiveness for epilepsy patients.</li>
<li>Quiz: Estimate percentage of patients responding to different treatments.</li>
</ul>
</div>
</div>

<div id="outline-container-org98389b4" class="outline-3">
<h3 id="org98389b4"><span class="section-number-3">2.5.</span> Predictive Modeling Quiz Solution</h3>
<div class="outline-text-3" id="text-2-5">
<ul class="org-ul">
<li>Solution breakdown:
<ul class="org-ul">
<li><b><b>Group A</b></b> (responded within 2 years): 32%.</li>
<li><b><b>Group B</b></b> (responded between 2-5 years): 24%.</li>
<li><b><b>Group C</b></b> (did not respond after 5 years): 44%.</li>
</ul></li>
<li>Goal of predictive modeling:
<ul class="org-ul">
<li>Improve early response rates.</li>
<li>Identify non-responders for alternative treatments.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org000ee58" class="outline-3">
<h3 id="org000ee58"><span class="section-number-3">2.6.</span> Predictive Modeling Challenges</h3>
<div class="outline-text-3" id="text-2-6">
<ul class="org-ul">
<li>Challenges in predictive modeling:
<ul class="org-ul">
<li>Handling large patient datasets with various types of data.</li>
<li>Managing multiple models in a complex computational pipeline.</li>
<li>Evaluating and comparing multiple predictive pipelines.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc398c97" class="outline-3">
<h3 id="orgc398c97"><span class="section-number-3">2.7.</span> Computational Phenotyping</h3>
<div class="outline-text-3" id="text-2-7">
<ul class="org-ul">
<li>Converts raw patient data (e.g., demographics, diagnoses, medications, procedures, lab tests, clinical notes) into clinical concepts or phenotypes.</li>
</ul>
</div>
</div>

<div id="outline-container-org17c7f98" class="outline-3">
<h3 id="org17c7f98"><span class="section-number-3">2.8.</span> Computational Phenotyping Quiz Question</h3>
<div class="outline-text-3" id="text-2-8">
<ul class="org-ul">
<li>Identifying data issues in phenotyping.</li>
<li>Quiz: List potential "waste products" in raw data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd9b8106" class="outline-3">
<h3 id="orgd9b8106"><span class="section-number-3">2.9.</span> Computational Phenotyping Quiz Solution</h3>
<div class="outline-text-3" id="text-2-9">
<ul class="org-ul">
<li>Common data issues:
<ul class="org-ul">
<li><b><b>Missing values</b></b>: Some important data may be absent.</li>
<li><b><b>Duplicates</b></b>: Patient records may appear multiple times.</li>
<li><b><b>Irrelevant data</b></b>: Not all raw information is useful.</li>
<li><b><b>Redundant information</b></b>: Different records may indicate the same condition (e.g., diagnosis and medication for diabetes).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org4439a60" class="outline-3">
<h3 id="org4439a60"><span class="section-number-3">2.10.</span> Phenotyping Algorithm</h3>
<div class="outline-text-3" id="text-2-10">
<ul class="org-ul">
<li>Example: Identifying Type 2 Diabetes from EHR data.</li>
<li>Decision process:
<ul class="org-ul">
<li>Check for Type 1 Diabetes diagnosis.</li>
<li>Check for Type 2 Diabetes diagnosis.</li>
<li>Verify medication records and abnormal lab results.</li>
</ul></li>
<li>Importance:
<ul class="org-ul">
<li>EHR data is often unreliable.</li>
<li>Multiple data sources improve diagnostic accuracy.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org531b539" class="outline-3">
<h3 id="org531b539"><span class="section-number-3">2.11.</span> Patient Similarity Quiz Question</h3>
<div class="outline-text-3" id="text-2-11">
<ul class="org-ul">
<li>Different types of reasoning doctors use:
<ul class="org-ul">
<li>Flowchart-based reasoning (like phenotyping algorithms).</li>
<li>Instinct and intuition.</li>
<li>Case-based reasoning (comparing patients to past cases).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org6e9d9ed" class="outline-3">
<h3 id="org6e9d9ed"><span class="section-number-3">2.12.</span> Patient Similarity Quiz Solution</h3>
<div class="outline-text-3" id="text-2-12">
<ul class="org-ul">
<li>Correct answer: <b><b>Case-based reasoning</b></b>.</li>
<li>Doctors often compare current patients to previous similar cases.</li>
</ul>
</div>
</div>

<div id="outline-container-org07ab028" class="outline-3">
<h3 id="org07ab028"><span class="section-number-3">2.13.</span> Patient Similarity</h3>
<div class="outline-text-3" id="text-2-13">
<ul class="org-ul">
<li>Simulating case-based reasoning using algorithms.</li>
<li>Process:
<ul class="org-ul">
<li>Search for similar patients in a database.</li>
<li>Identify treatment outcomes for similar cases.</li>
<li>Recommend best treatment based on historical data.</li>
</ul></li>
<li>Objective: Utilize full database knowledge instead of relying on a single doctor's experience.</li>
</ul>
</div>
</div>

<div id="outline-container-org4eeb7ec" class="outline-3">
<h3 id="org4eeb7ec"><span class="section-number-3">2.14.</span> Algorithms</h3>
<div class="outline-text-3" id="text-2-14">
<ul class="org-ul">
<li>Introduction to machine learning algorithms in healthcare.</li>
<li>Covered topics:
<ol class="org-ol">
<li><b><b>Classification</b></b>: Mapping patient data to target variables (e.g., predicting heart attack risk).</li>
<li><b><b>Clustering</b></b>: Grouping similar patients based on health conditions.</li>
<li><b><b>Dimensionality Reduction</b></b>: Reducing large patient datasets to essential features.</li>
<li><b><b>Graph Analysis</b></b>: Analyzing relationships between patients and diseases.</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org5f56817" class="outline-3">
<h3 id="org5f56817"><span class="section-number-3">2.15.</span> Systems</h3>
<div class="outline-text-3" id="text-2-15">
<ul class="org-ul">
<li>Introduction to big data systems for healthcare applications.</li>
<li>Covered systems:
<ul class="org-ul">
<li><b><b>Hadoop</b></b>: Disk-based distributed system.</li>
<li><b><b>Spark</b></b>: In-memory distributed system (faster than Hadoop).</li>
</ul></li>
<li>Topics include:
<ul class="org-ul">
<li>Hadoop infrastructure (MapReduce, HDFS, Pig, Hive, HBase).</li>
<li>Spark infrastructure (Spark SQL, Spark Streaming, MLlib, GraphX).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org2d06e9b" class="outline-3">
<h3 id="org2d06e9b"><span class="section-number-3">2.16.</span> Summary</h3>
<div class="outline-text-3" id="text-2-16">
<ul class="org-ul">
<li>Recap of three key areas:
<ul class="org-ul">
<li>Healthcare applications.</li>
<li>Machine learning algorithms.</li>
<li>Big data systems.</li>
</ul></li>
<li>Course integrates these areas:
<ul class="org-ul">
<li>Example: Using logistic regression on Hadoop to predict heart failure.</li>
</ul></li>
<li>Next step: Begin applying concepts.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org1c90d4b" class="outline-2">
<h2 id="org1c90d4b"><span class="section-number-2">3.</span> Predictive Modeling</h2>
<div class="outline-text-2" id="text-3">
</div>
<div id="outline-container-org98658b5" class="outline-3">
<h3 id="org98658b5"><span class="section-number-3">3.1.</span> Introduction to Predictive Modeling</h3>
<div class="outline-text-3" id="text-3-1">
<dl class="org-dl">
<dt>Predictive modeling</dt><dd>using historical data to predict future events.

<ul class="org-ul">
<li>Example: Using electronic health records (EHR) to model heart failure</li>
<li>Key Goal: Develop a good predictive model using EHR efficiently.</li>
</ul></dd>
</dl>
</div>
</div>

<div id="outline-container-orgabd0ae8" class="outline-3">
<h3 id="orgabd0ae8"><span class="section-number-3">3.2.</span> Predictive Modeling vs EHR</h3>
<div class="outline-text-3" id="text-3-2">
<ul class="org-ul">
<li>Importance: Increased research interest in using EHR for clinical predictive modeling.</li>
<li>Data Sources: EHR has become a major source for predictive modeling research.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcecd22d" class="outline-3">
<h3 id="orgcecd22d"><span class="section-number-3">3.3.</span> Predictive Modeling Pipeline</h3>
<div class="outline-text-3" id="text-3-3">
<ul class="org-ul">
<li>Predictive modeling is a multi-step computational process:
<ol class="org-ol">
<li>Define the prediction target.</li>
<li>Construct the relevant patient cohort.</li>
<li>Identify potentially relevant features.</li>
<li>Select the most relevant features.</li>
<li>Compute the predictive model.</li>
<li>Evaluate the predictive model.</li>
</ol></li>
<li>Iterative process until a satisfactory model is obtained.</li>
</ul>
</div>
</div>

<div id="outline-container-orgde0d506" class="outline-3">
<h3 id="orgde0d506"><span class="section-number-3">3.4.</span> Prediction Target</h3>
<div class="outline-text-3" id="text-3-4">
<ul class="org-ul">
<li>Investigators may have many targets, but only some are feasible.</li>
<li>Selection criteria: The target should be interesting and possible with the available data.</li>
<li>Example: Predicting the onset of heart failure.</li>
</ul>
</div>
</div>

<div id="outline-container-org372ee0c" class="outline-3">
<h3 id="org372ee0c"><span class="section-number-3">3.5.</span> Heart Failure Quiz</h3>
<div class="outline-text-3" id="text-3-5">
<ul class="org-ul">
<li>Question: How many new heart failure cases occur annually in the U.S.?</li>
<li>Options: 17,000; 260,000; 550,000; 1,250,000.</li>
<li>Answer: 550,000 cases per year.</li>
</ul>
</div>
</div>

<div id="outline-container-org2add229" class="outline-3">
<h3 id="org2add229"><span class="section-number-3">3.6.</span> Motivations for Early Detection</h3>
<div class="outline-text-3" id="text-3-6">
<ul class="org-ul">
<li>Heart failure is complex with diverse symptoms and subsets.</li>
<li>Early detection can:
<ul class="org-ul">
<li>Reduce hospitalization costs.</li>
<li>Introduce early interventions to slow progression.</li>
<li>Improve clinical guidelines for heart failure prevention.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgd2a6fbf" class="outline-3">
<h3 id="orgd2a6fbf"><span class="section-number-3">3.7.</span> Cohort Construction</h3>
<div class="outline-text-3" id="text-3-7">
<ul class="org-ul">
<li>Defines the study population for predictive modeling.</li>
<li>Study population is a subset of all patients.</li>
<li>Four study designs based on two axes:
<ul class="org-ul">
<li>Prospective vs. Retrospective studies.</li>
<li>Cohort vs. Case-Control studies.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org27c6794" class="outline-3">
<h3 id="org27c6794"><span class="section-number-3">3.8.</span> Prospective vs. Retrospective Studies</h3>
<div class="outline-text-3" id="text-3-8">
<ul class="org-ul">
<li><b><b>Prospective Study</b></b>: Define cohort first, then collect data.</li>
<li><b><b>Retrospective Study</b></b>: Use existing data from past records.</li>
</ul>
</div>
</div>

<div id="outline-container-org2b4d6be" class="outline-3">
<h3 id="org2b4d6be"><span class="section-number-3">3.9.</span> Prospective vs. Retrospective Quiz</h3>
<div class="outline-text-3" id="text-3-9">
<ul class="org-ul">
<li>Comparison of study properties:
<ul class="org-ul">
<li>Retrospective studies have more noise.</li>
<li>Prospective studies are more expensive and time-consuming.</li>
<li>Retrospective studies can handle larger datasets.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org1f302a6" class="outline-3">
<h3 id="org1f302a6"><span class="section-number-3">3.10.</span> Cohort Study</h3>
<div class="outline-text-3" id="text-3-10">
<ul class="org-ul">
<li>Includes patients exposed to a particular risk (e.g., heart failure readmission).</li>
<li>Defines inclusion and exclusion criteria.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfad8e76" class="outline-3">
<h3 id="orgfad8e76"><span class="section-number-3">3.11.</span> Case-Control Study</h3>
<div class="outline-text-3" id="text-3-11">
<ul class="org-ul">
<li>Compares patients with a condition (cases) to similar patients without it (controls).</li>
<li>Matching criteria: Age, gender, clinic visits.</li>
<li>Cases are rarer than controls in real-world data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb96d78f" class="outline-3">
<h3 id="orgb96d78f"><span class="section-number-3">3.12.</span> Feature Construction</h3>
<div class="outline-text-3" id="text-3-12">
<ul class="org-ul">
<li>Defines patient features for predicting outcomes.</li>
<li>Data sequences from EHR:
<ul class="org-ul">
<li>Events (diagnoses, medications, lab results).</li>
<li>Observation window (used for feature extraction).</li>
<li>Prediction window (future period to predict).</li>
</ul></li>
<li>Impact of window sizes on model accuracy.</li>
</ul>
</div>
</div>

<div id="outline-container-org86f749e" class="outline-3">
<h3 id="org86f749e"><span class="section-number-3">3.13.</span> Feature Construction Quizzes</h3>
<div class="outline-text-3" id="text-3-13">
<ul class="org-ul">
<li>Quiz 1: Best timeline for modeling → Large observation, small prediction window.</li>
<li>Quiz 2: Most useful model → Small observation, large prediction window (ideal but difficult).</li>
</ul>
</div>
</div>

<div id="outline-container-org43cbd0b" class="outline-3">
<h3 id="org43cbd0b"><span class="section-number-3">3.14.</span> Prediction Performance on Windows</h3>
<div class="outline-text-3" id="text-3-14">
<ul class="org-ul">
<li><b><b>Prediction Window</b></b>: Longer window reduces accuracy.</li>
<li><b><b>Observation Window</b></b>: Longer window improves model performance until plateau.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf35498b" class="outline-3">
<h3 id="orgf35498b"><span class="section-number-3">3.15.</span> Feature Selection</h3>
<div class="outline-text-3" id="text-3-15">
<ul class="org-ul">
<li>Identifies relevant features from EHR data.</li>
<li>EHR provides a large number of potential features (e.g., demographics, vitals).</li>
<li>Different targets require different feature subsets.</li>
</ul>
</div>
</div>

<div id="outline-container-org301b471" class="outline-3">
<h3 id="org301b471"><span class="section-number-3">3.16.</span> Predictive Model</h3>
<div class="outline-text-3" id="text-3-16">
<ul class="org-ul">
<li>Maps input features to predicted outcomes.</li>
<li>Regression models for continuous targets (e.g., healthcare costs).</li>
<li>Classification models for categorical targets (e.g., heart failure).</li>
<li>Common methods: Logistic regression, decision trees, random forests.</li>
</ul>
</div>
</div>

<div id="outline-container-org817d346" class="outline-3">
<h3 id="org817d346"><span class="section-number-3">3.17.</span> Performance Evaluation</h3>
<div class="outline-text-3" id="text-3-17">
<ul class="org-ul">
<li>Key metric: Testing error (not training error).</li>
<li>Models must generalize to unseen data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaa3fec5" class="outline-3">
<h3 id="orgaa3fec5"><span class="section-number-3">3.18.</span> Cross-Validation</h3>
<div class="outline-text-3" id="text-3-18">
<ul class="org-ul">
<li>Splits data into training and validation sets.</li>
<li>Common methods:
<ul class="org-ul">
<li>Leave-One-Out Cross-Validation.</li>
<li>K-Fold Cross-Validation.</li>
<li>Randomized Cross-Validation.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc5618f5" class="outline-3">
<h3 id="orgc5618f5"><span class="section-number-3">3.19.</span> Conclusion</h3>
<div class="outline-text-3" id="text-3-19">
<ul class="org-ul">
<li>Summary of the predictive modeling pipeline:
<ul class="org-ul">
<li>Define the prediction target.</li>
<li>Construct the patient cohort.</li>
<li>Generate relevant features.</li>
<li>Select important features.</li>
<li>Train the predictive model.</li>
<li>Evaluate model performance.</li>
</ul></li>
<li>Goal: Design high-level predictive modeling studies using EHR data.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org8d1c04f" class="outline-2">
<h2 id="org8d1c04f"><span class="section-number-2">4.</span> MapReduce</h2>
<div class="outline-text-2" id="text-4">
</div>
<div id="outline-container-org83be4fb" class="outline-3">
<h3 id="org83be4fb"><span class="section-number-3">4.1.</span> Introduction to MapReduce</h3>
<div class="outline-text-3" id="text-4-1">
<ul class="org-ul">
<li>Overview of MapReduce as a big data processing tool.</li>
<li>Utilizes distributed computation and storage.</li>
<li>Covers:
<ul class="org-ul">
<li>What MapReduce is</li>
<li>Fault tolerance in distributed environments</li>
<li>Analytical use cases and limitations</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org68f3374" class="outline-3">
<h3 id="org68f3374"><span class="section-number-3">4.2.</span> What is MapReduce</h3>
<div class="outline-text-3" id="text-4-2">
<ul class="org-ul">
<li>Hadoop/MapReduce is:
<ul class="org-ul">
<li>A programming model for parallel computation</li>
<li>An execution environment (Hadoop with HDFS)</li>
<li>A software package with various tools</li>
</ul></li>
<li>Capabilities:
<ul class="org-ul">
<li>Distributed storage via HDFS</li>
<li>Distributed computation via MapReduce</li>
<li>Built-in fault tolerance</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8385004" class="outline-3">
<h3 id="org8385004"><span class="section-number-3">4.3.</span> Computational Process</h3>
<div class="outline-text-3" id="text-4-3">
<ul class="org-ul">
<li>Originated at Google (2004), open-sourced via Apache Hadoop.</li>
<li>Java-based platform.</li>
<li>Programming constrained to Map and Reduce functions for scalability.</li>
<li>Designed for parallel, fault-tolerant processing.</li>
<li>Emphasis on mastering computational patterns for analytics.</li>
</ul>
</div>
</div>

<div id="outline-container-orge7cabf1" class="outline-3">
<h3 id="orge7cabf1"><span class="section-number-3">4.4.</span> Learning Via Aggregation Statistics</h3>
<div class="outline-text-3" id="text-4-4">
<ul class="org-ul">
<li>Core design: express ML algorithms as aggregation tasks.</li>
<li>Example: Heart failure risk factor frequency analysis.</li>
<li>Map: extract risk factors per patient.</li>
<li>Reduce: aggregate frequencies across population.</li>
<li>Leads into deeper abstraction needs and trade-offs.</li>
</ul>
</div>
</div>

<div id="outline-container-org57b1d62" class="outline-3">
<h3 id="org57b1d62"><span class="section-number-3">4.5.</span> MapReduce Abstraction</h3>
<div class="outline-text-3" id="text-4-5">
<ul class="org-ul">
<li>Example task: count disease cases from patient records.</li>
<li>Map:
<ul class="org-ul">
<li>Emit (disease, 1) for each mention</li>
</ul></li>
<li>Shuffle:
<ul class="org-ul">
<li>Group by disease</li>
</ul></li>
<li>Reduce:
<ul class="org-ul">
<li>Sum values for each disease</li>
</ul></li>
<li>Emphasizes two-phase logic for scalability.</li>
</ul>
</div>
</div>

<div id="outline-container-org3bb834c" class="outline-3">
<h3 id="org3bb834c"><span class="section-number-3">4.6.</span> MapReduce System</h3>
<div class="outline-text-3" id="text-4-6">
<ul class="org-ul">
<li>Real-world data is too large for single machines.</li>
<li>Data is partitioned and processed by multiple mappers.</li>
<li>Intermediate results are shuffled and passed to reducers.</li>
<li>Final results computed from reduce function.</li>
<li>Three stages:
<ul class="org-ul">
<li>Map</li>
<li>Shuffle</li>
<li>Reduce</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8fc267b" class="outline-3">
<h3 id="org8fc267b"><span class="section-number-3">4.7.</span> MapReduce Fault Recovery</h3>
<div class="outline-text-3" id="text-4-7">
<ul class="org-ul">
<li>Fault tolerance is a built-in system feature.</li>
<li>On failure:
<ul class="org-ul">
<li>Mappers/reducers are restarted with minimal recomputation.</li>
</ul></li>
<li>Goal: Only failed components are recomputed.</li>
<li>System handles all failure recovery.</li>
</ul>
</div>
</div>

<div id="outline-container-org8e5e268" class="outline-3">
<h3 id="org8e5e268"><span class="section-number-3">4.8.</span> Distributed File Systems</h3>
<div class="outline-text-3" id="text-4-8">
<ul class="org-ul">
<li>HDFS: Hadoop's storage system.</li>
<li>Splits large files into partitions.</li>
<li>Partitions stored on multiple machines with redundancy.</li>
<li>Benefits:
<ul class="org-ul">
<li>Faster concurrent access</li>
<li>Fault tolerance (recover from worker failures)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org976c247" class="outline-3">
<h3 id="org976c247"><span class="section-number-3">4.9.</span> MapReduce Design Choice</h3>
<div class="outline-text-3" id="text-4-9">
<ul class="org-ul">
<li>Design principle: minimal functionality for reliability and scalability.</li>
<li>Restricted computation model (e.g., aggregation queries).</li>
<li>Map: operate on individual records</li>
<li>Reduce: aggregate results</li>
<li>Supports straggler mitigation (slow mappers duplicated)</li>
</ul>
</div>
</div>

<div id="outline-container-orgb4f2e47" class="outline-3">
<h3 id="orgb4f2e47"><span class="section-number-3">4.10.</span> Analytics with MapReduce</h3>
<div class="outline-text-3" id="text-4-10">
<ul class="org-ul">
<li>Application examples:
<ul class="org-ul">
<li>K-Nearest Neighbors (KNN)</li>
<li>Linear Regression</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org58701a3" class="outline-3">
<h3 id="org58701a3"><span class="section-number-3">4.11.</span> MapReduce KNN</h3>
<div class="outline-text-3" id="text-4-11">
<ul class="org-ul">
<li>KNN implementation:
<ul class="org-ul">
<li>Map:
<ul class="org-ul">
<li>Find K nearest neighbors per partition</li>
</ul></li>
<li>Reduce:
<ul class="org-ul">
<li>Combine local results to find global nearest neighbors</li>
</ul></li>
</ul></li>
<li>Patient data partitioned, processed in parallel</li>
</ul>
</div>
</div>

<div id="outline-container-org7c68944" class="outline-3">
<h3 id="org7c68944"><span class="section-number-3">4.12.</span> Linear Regression</h3>
<div class="outline-text-3" id="text-4-12">
<ul class="org-ul">
<li>Goal: map patient features to heart disease risk</li>
<li>Normal equation:
<ul class="org-ul">
<li>β = (XᵗX)⁻¹Xᵗy</li>
</ul></li>
<li>MapReduce:
<ul class="org-ul">
<li>Map f1: compute xi * xiᵗ</li>
<li>Map f2: compute xi * yi</li>
<li>Reduce: aggregate sums</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7e31046" class="outline-3">
<h3 id="org7e31046"><span class="section-number-3">4.13.</span> MapReduce for Linear Regression Quiz Question</h3>
<div class="outline-text-3" id="text-4-13">
<ul class="org-ul">
<li>Quiz: specify map/reduce pseudo code for computing Xᵗy</li>
</ul>
</div>
</div>

<div id="outline-container-orgabab1f5" class="outline-3">
<h3 id="orgabab1f5"><span class="section-number-3">4.14.</span> MapReduce for Linear Regression Quiz Solution</h3>
<div class="outline-text-3" id="text-4-14">
<ul class="org-ul">
<li>Solution:
<ul class="org-ul">
<li>Map: compute xi * yi</li>
<li>Reduce: aggregate results</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org32fa747" class="outline-3">
<h3 id="org32fa747"><span class="section-number-3">4.15.</span> Limitations of MapReduce</h3>
<div class="outline-text-3" id="text-4-15">
<ul class="org-ul">
<li>Example: Logistic Regression via Gradient Descent</li>
<li>Challenge: Requires iterative computation
<ul class="org-ul">
<li>Each iteration loads data twice</li>
</ul></li>
<li>MapReduce not efficient for:
<ul class="org-ul">
<li>Iterative, multi-stage computation</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgac0616d" class="outline-3">
<h3 id="orgac0616d"><span class="section-number-3">4.16.</span> MapReduce Summary Quiz Question</h3>
<div class="outline-text-3" id="text-4-16">
<ul class="org-ul">
<li>Quiz on ideal conditions for MapReduce:
<ul class="org-ul">
<li>Single vs. multiple passes</li>
<li>Skewed vs. uniform key distribution</li>
<li>Synchronization needs</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc78ca8b" class="outline-3">
<h3 id="orgc78ca8b"><span class="section-number-3">4.17.</span> MapReduce Summary Quiz Solution</h3>
<div class="outline-text-3" id="text-4-17">
<ul class="org-ul">
<li>Best suited for:
<ul class="org-ul">
<li>Single-pass jobs (e.g., histograms)</li>
<li>Both skewed and uniform key distributions</li>
<li>Minimal synchronization (only between Map and Reduce phases)</li>
</ul></li>
</ul>
</div>
</div>
</div>
<div id="outline-container-orge4ae145" class="outline-2">
<h2 id="orge4ae145"><span class="section-number-2">5.</span> Classification Model Metrics</h2>
<div class="outline-text-2" id="text-5">
</div>
<div id="outline-container-orgf226510" class="outline-3">
<h3 id="orgf226510"><span class="section-number-3">5.1.</span> Predictive Model review</h3>
<div class="outline-text-3" id="text-5-1">
<dl class="org-dl">
<dt>Predictive model</dt><dd>mapping function between model inputs and outputs (pred's)</dd>
<dt>Model metrics</dt><dd>needed to know how well they're performing</dd>
</dl>
</div>
</div>
<div id="outline-container-orgedefc00" class="outline-3">
<h3 id="orgedefc00"><span class="section-number-3">5.2.</span> Confusion matrix</h3>
<div class="outline-text-3" id="text-5-2">
<ul class="org-ul">
<li>True positive/negative also known as "Condition" positive/negative</li>
<li>Predicted positive/negative also known as "Prediction Outcome" positive/negative</li>
</ul>
</div>
</div>
<div id="outline-container-orgc283160" class="outline-3">
<h3 id="orgc283160"><span class="section-number-3">5.3.</span> Accuracy metrics</h3>
<div class="outline-text-3" id="text-5-3">
<p>
All divided by ground truth values.
</p>
<dl class="org-dl">
<dt>Accuracy</dt><dd>(TP+TN)/TP. Not most useful for imbalanced class</dd>
<dt>True positive rate</dt><dd>TP/CP (Sensitivity, Recall)</dd>
<dt>False positive rate</dt><dd>FP/CN</dd>
<dt>False negative</dt><dd>FN/CP</dd>
<dt>True negative rate</dt><dd>TN/CN (Specificity)</dd>
</dl>
</div>
<div id="outline-container-orgbcf10f7" class="outline-4">
<h4 id="orgbcf10f7"><span class="section-number-4">5.3.1.</span> Other notes</h4>
<div class="outline-text-4" id="text-5-3-1">
<ul class="org-ul">
<li>FP is a type I error.</li>
<li>FN is a type II error.</li>
<li>Hard to perform well on all metrics</li>
<li>Important to choose the right metrics</li>
</ul>
</div>
</div>
</div>
<div id="outline-container-orgc32b30d" class="outline-3">
<h3 id="orgc32b30d"><span class="section-number-3">5.4.</span> Predictive metrics</h3>
<div class="outline-text-3" id="text-5-4">
<p>
All divided by prediction outcomes.
</p>
<dl class="org-dl">
<dt>Prevalence</dt><dd>CP/Total population. How likely disease occurs in population</dd>
<dt>Positive predictive value (Precision)</dt><dd>TP/Pred outcome positive.</dd>
<dt>False discovery rate</dt><dd>FP/Pred outcome positive</dd>
<dt>Negative predictive value</dt><dd>TN/Pred outcome negative</dd>
<dt>False omission rate</dt><dd>FN/Pred outcome negative</dd>
</dl>
</div>
</div>
<div id="outline-container-orge5832cf" class="outline-3">
<h3 id="orge5832cf"><span class="section-number-3">5.5.</span> F1 score</h3>
<div class="outline-text-3" id="text-5-5">
<p>
Harmonic mean of PPV and TPR
\[
F_1 = 2 \times \frac{PPV \times TPR}{PPV + TPR}
\]
</p>
</div>
</div>
<div id="outline-container-org016e156" class="outline-3">
<h3 id="org016e156"><span class="section-number-3">5.6.</span> Classifier quiz</h3>
<div class="outline-text-3" id="text-5-6">
<p>
Which is the best classifier?
</p>
<ul class="org-ul">
<li>Highest F1, PPV and Accuracy is the best classifier</li>
</ul>
</div>
</div>
<div id="outline-container-orgf10fb0c" class="outline-3">
<h3 id="orgf10fb0c"><span class="section-number-3">5.7.</span> Reversing predictions</h3>
<div class="outline-text-3" id="text-5-7">
<p>
It's always possible to reverse the predictions so 0.21 might perform better than 0.69/0.50.
</p>
</div>
</div>
<div id="outline-container-org75aab64" class="outline-3">
<h3 id="org75aab64"><span class="section-number-3">5.8.</span> Receiver operating characteristic</h3>
<div class="outline-text-3" id="text-5-8">
<ul class="org-ul">
<li>Predictive models generally output continuous score.</li>
<li>Threshold is needed as precision bound to force to a certain category</li>
<li>ROC provides a way to compare performance of classifiers as the decision boundary is varied</li>
<li>ROC curve is the plot of TP rate vs FP rate at various threshold values
<ul class="org-ul">
<li>We sort by prediction score (highest first)</li>
<li>Use prediction score as threshold values</li>
<li>Plot on the chart and see how many are misclassified (needs True value to be known)</li>
</ul></li>
<li>AUROC does not depend on the choice of threshold.</li>
<li>AUROC is the most popular metric for classification</li>
</ul>
</div>
</div>
<div id="outline-container-orgf6f12b6" class="outline-3">
<h3 id="orgf6f12b6"><span class="section-number-3">5.9.</span> "Best classifier threshold" quiz</h3>
<div class="outline-text-3" id="text-5-9">
<ul class="org-ul">
<li>There isn't a standard answer, it depends on whether TP, TN or other metric is prioritized</li>
</ul>
</div>
</div>
<div id="outline-container-orgac6e6e5" class="outline-3">
<h3 id="orgac6e6e5"><span class="section-number-3">5.10.</span> Regression metrics</h3>
<div class="outline-text-3" id="text-5-10">
<dl class="org-dl">
<dt>MAE</dt><dd>average of absolute errors, harder to work with since the absolute value is not differentiable</dd>
<dt>MSE</dt><dd>average of squared errors, easier to work with as the derivative of squared term is linear. Increases a lot faster than MAE</dd>
<dt>\(R^2\)</dt><dd>bounded by \((-\infty,1)\). AKA coefficient of determination.</dd>
</dl>
</div>
<div id="outline-container-org0efdf41" class="outline-4">
<h4 id="org0efdf41"><span class="section-number-4">5.10.1.</span> \(R^2\)</h4>
<div class="outline-text-4" id="text-5-10-1">
<p>
\[
R^2 = 1-\frac{\sum_i (y_i - \hat{y}_i)^2}{\sum_i(y_i - \bar{y})^2}
\]
i.e. 1-MSE/Variance
</p>

<p>
Negative \(R^2\) values means they perform worse than a simple average of raw data.
As noise increases, \(R^2\) decreases.
</p>
</div>
</div>
</div>
</div>
<div id="outline-container-org86ce29a" class="outline-2">
<h2 id="org86ce29a"><span class="section-number-2">6.</span> Ensemble methods</h2>
<div class="outline-text-2" id="text-6">
</div>
<div id="outline-container-org6ca7412" class="outline-3">
<h3 id="org6ca7412"><span class="section-number-3">6.1.</span> Gradient Descent Method (GDM) for Linear Regression</h3>
<div class="outline-text-3" id="text-6-1">
<ul class="org-ul">
<li>Illustrates gradient descent using linear regression.</li>
<li>Dataset: each row = patient; features in X, outcome Y (e.g., hospital cost).</li>
<li>Objective: learn linear mapping from features X to outcome Y.</li>
<li>Assumes Gaussian distribution → log-likelihood leads to squared error minimization.</li>
<li>Gradient: derived by taking derivative of log-likelihood w.r.t. coefficients β.</li>
<li>Update rule: move β in direction of gradient with step size η.</li>
<li>Requires multiple iterations over full dataset → computationally expensive.</li>
</ul>
</div>
</div>

<div id="outline-container-org1c3b667" class="outline-3">
<h3 id="org1c3b667"><span class="section-number-3">6.2.</span> Stochastic Gradient Descent (SGD) Method</h3>
<div class="outline-text-3" id="text-6-2">
<ul class="org-ul">
<li>SGD is a scalable variant of gradient descent for large datasets.</li>
<li>Instead of using full dataset, computes gradients on random subsets (mini-batches).</li>
<li>One-point update = SGD; larger batches = mini-batch gradient descent.</li>
<li>Faster per-iteration updates compared to full batch gradient descent.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbe1c3cc" class="outline-3">
<h3 id="orgbe1c3cc"><span class="section-number-3">6.3.</span> SGD for Linear Regression</h3>
<div class="outline-text-3" id="text-6-3">
<ul class="org-ul">
<li>Applies SGD to same regression problem.</li>
<li>Each update uses one patient:
<ul class="org-ul">
<li>Compute individual log-likelihood and gradient.</li>
<li>Update β using gradient.</li>
</ul></li>
<li>More efficient than full gradient computation.</li>
<li>Key idea: updates based on single or small group of samples.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc830d93" class="outline-3">
<h3 id="orgc830d93"><span class="section-number-3">6.4.</span> Ensemble Method Pt 1</h3>
<div class="outline-text-3" id="text-6-4">
<ul class="org-ul">
<li>Ensemble methods combine multiple models for improved performance.</li>
<li>Can use same (e.g., Random Forest) or different base models.</li>
<li>Real-world example: Netflix Prize → ensemble models won.</li>
<li>Many teams merged, forming more powerful ensembles.</li>
</ul>
</div>
</div>

<div id="outline-container-orge9adf2e" class="outline-3">
<h3 id="orge9adf2e"><span class="section-number-3">6.5.</span> Ensemble Method Pt 2</h3>
<div class="outline-text-3" id="text-6-5">
<ul class="org-ul">
<li>General ensemble steps:
<ol class="org-ol">
<li>Generate multiple datasets (via bagging or boosting).</li>
<li>Train separate models on each.</li>
<li>Aggregate outputs using functions like averaging or weighted sum.</li>
</ol></li>
<li>Outcome: different ensemble strategies.</li>
</ul>
</div>
</div>

<div id="outline-container-org17388b0" class="outline-3">
<h3 id="org17388b0"><span class="section-number-3">6.6.</span> Bias Variance Tradeoff</h3>
<div class="outline-text-3" id="text-6-6">
<ul class="org-ul">
<li>Key insight: model error = bias² + variance.</li>
<li>Bias: error from wrong assumptions (e.g., assuming linear when it's not).</li>
<li>Variance: error from sensitivity to training data.</li>
<li>Ideal model: low bias and low variance.</li>
<li>Graphical example: dartboard showing different combinations.</li>
<li>As model complexity increases:
<ul class="org-ul">
<li>Bias ↓</li>
<li>Variance ↑</li>
<li>Total error = minimum at an optimal complexity.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8215a54" class="outline-3">
<h3 id="org8215a54"><span class="section-number-3">6.7.</span> Bias Variance Tradeoff Quiz Question</h3>
<div class="outline-text-3" id="text-6-7">
<ul class="org-ul">
<li>Task: rank four models (flat, linear, smooth curve, wiggly) by complexity.</li>
<li>A → D: D (flat) &lt; A (linear) &lt; B (curve) &lt; C (wiggly)</li>
</ul>
</div>
</div>

<div id="outline-container-org6ba8eda" class="outline-3">
<h3 id="org6ba8eda"><span class="section-number-3">6.8.</span> Bias Variance Tradeoff Quiz Solution</h3>
<div class="outline-text-3" id="text-6-8">
<ul class="org-ul">
<li>Explanation:
<ul class="org-ul">
<li>D: lowest complexity (flat line).</li>
<li>A: next, simple linear model.</li>
<li>B: more complex curve.</li>
<li>C: most complex (wiggly).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org24d6c20" class="outline-3">
<h3 id="org24d6c20"><span class="section-number-3">6.9.</span> Bias Variance Tradeoff Quiz 2 Question</h3>
<div class="outline-text-3" id="text-6-9">
<ul class="org-ul">
<li>Given same models, asked which best fits data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf28f988" class="outline-3">
<h3 id="orgf28f988"><span class="section-number-3">6.10.</span> Bias Variance Tradeoff Quiz 2 Solution</h3>
<div class="outline-text-3" id="text-6-10">
<ul class="org-ul">
<li>Answer: B (smooth curve)</li>
<li>Balances bias and variance well.</li>
</ul>
</div>
</div>

<div id="outline-container-org86cf9b3" class="outline-3">
<h3 id="org86cf9b3"><span class="section-number-3">6.11.</span> Bagging</h3>
<div class="outline-text-3" id="text-6-11">
<ul class="org-ul">
<li>Bagging = Bootstrap Aggregation.</li>
<li>Method:
<ul class="org-ul">
<li>Sample data with replacement.</li>
<li>Train models on each sample.</li>
<li>Combine predictions via majority vote or averaging.</li>
</ul></li>
<li>Reduces variance of the final model.</li>
</ul>
</div>
</div>

<div id="outline-container-org4e887a7" class="outline-3">
<h3 id="org4e887a7"><span class="section-number-3">6.12.</span> Random Forest</h3>
<div class="outline-text-3" id="text-6-12">
<ul class="org-ul">
<li>Special case of bagging with decision trees.</li>
<li>Steps:
<ul class="org-ul">
<li>Randomly sample both rows (patients) and columns (features).</li>
<li>Train simple decision trees on these subsets.</li>
</ul></li>
<li>Final prediction = average of all trees.</li>
<li>Simple trees help maintain diversity and lower computation.</li>
</ul>
</div>
</div>

<div id="outline-container-org4f26b0a" class="outline-3">
<h3 id="org4f26b0a"><span class="section-number-3">6.13.</span> Why Bagging Works</h3>
<div class="outline-text-3" id="text-6-13">
<ul class="org-ul">
<li>Reduces variance without increasing bias.</li>
<li>Variance reduction from averaging independent model predictions.</li>
<li>Based on statistical principle: variance of mean = variance / T for T models.</li>
</ul>
</div>
</div>

<div id="outline-container-org9600e34" class="outline-3">
<h3 id="org9600e34"><span class="section-number-3">6.14.</span> Boosting</h3>
<div class="outline-text-3" id="text-6-14">
<ul class="org-ul">
<li>Builds models sequentially.</li>
<li>Each new model focuses on mistakes of previous ones.</li>
<li>Final model = weighted combination of all models.</li>
<li>Pros: better accuracy.</li>
<li>Cons: prone to overfitting.</li>
<li>Example: AdaBoost is most popular.</li>
</ul>
</div>
</div>

<div id="outline-container-org6189eb8" class="outline-3">
<h3 id="org6189eb8"><span class="section-number-3">6.15.</span> Bagging vs Boosting Quiz Question</h3>
<div class="outline-text-3" id="text-6-15">
<ul class="org-ul">
<li>Quiz to compare characteristics:
<ul class="org-ul">
<li>Combination method (simple vs weighted average)</li>
<li>Parallelism</li>
<li>Noise sensitivity</li>
<li>Accuracy</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7a9cf9a" class="outline-3">
<h3 id="org7a9cf9a"><span class="section-number-3">6.16.</span> Bagging vs Boosting Quiz Solution</h3>
<div class="outline-text-3" id="text-6-16">
<ul class="org-ul">
<li>Bagging:
<ul class="org-ul">
<li>Simple average, parallel-friendly, less sensitive to noise, reliable.</li>
</ul></li>
<li>Boosting:
<ul class="org-ul">
<li>Weighted average, sequential (hard to parallelize), sensitive to noise, potentially higher accuracy but less reliable.</li>
</ul></li>
<li>Analogy:
<ul class="org-ul">
<li>Bagging = reliable Japanese car.</li>
<li>Boosting = high-performance but risky sports car.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org17a1633" class="outline-3">
<h3 id="org17a1633"><span class="section-number-3">6.17.</span> Summary for Ensemble Methods</h3>
<div class="outline-text-3" id="text-6-17">
<ul class="org-ul">
<li>Pros:
<ul class="org-ul">
<li>Simple, flexible, few parameters, theoretical backing.</li>
</ul></li>
<li>Cons:
<ul class="org-ul">
<li>Computational cost (training + inference).</li>
<li>Harder interpretation (due to model complexity).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgacf2391" class="outline-3">
<h3 id="orgacf2391"><span class="section-number-3">6.18.</span> Introduction to Ensemble Method</h3>
<div class="outline-text-3" id="text-6-18">
<ul class="org-ul">
<li>Sets the stage:
<ul class="org-ul">
<li>Recap on model evaluation.</li>
<li>Overview: SGD, ensemble methods, bias-variance tradeoff, bagging &amp; boosting.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org244c378" class="outline-3">
<h3 id="org244c378"><span class="section-number-3">6.19.</span> Gradient Descent Method</h3>
<div class="outline-text-3" id="text-6-19">
<ul class="org-ul">
<li>General overview of gradient descent for optimization.</li>
<li>Applies to regression/classification.</li>
<li>Process:
<ol class="org-ol">
<li>Define likelihood (or log-likelihood).</li>
<li>Compute gradient w.r.t. parameters.</li>
<li>Update parameters iteratively using gradient.</li>
</ol></li>
<li>Requires step size (learning rate).</li>
<li>Can be extended using more advanced techniques (e.g., conjugate gradient).</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgb34656a" class="outline-2">
<h2 id="orgb34656a"><span class="section-number-2">7.</span> Computational Phenotyping</h2>
<div class="outline-text-2" id="text-7">
</div>
<div id="outline-container-org8881adf" class="outline-3">
<h3 id="org8881adf"><span class="section-number-3">7.1.</span> Introduction to Phenotyping</h3>
<div class="outline-text-3" id="text-7-1">
<ul class="org-ul">
<li>Introduction to clustering in healthcare, specifically "phenotyping."</li>
<li>Phenotypes refer to diseases or conditions; known phenotypes exist, but many more are undiscovered.</li>
<li>Computational phenotyping helps discover novel phenotypes using available data.</li>
<li>Applications include disease diagnosis, cost prediction, readmission risk, and genomic studies.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb772c73" class="outline-3">
<h3 id="orgb772c73"><span class="section-number-3">7.2.</span> Computational Phenotyping</h3>
<div class="outline-text-3" id="text-7-2">
<ul class="org-ul">
<li>Converts raw EHR data into meaningful phenotypes using algorithms.</li>
<li>EHR data includes demographics, diagnosis codes, medications, procedures, labs, notes.</li>
<li>Challenges:
<ul class="org-ul">
<li>Noisy and incomplete data.</li>
<li>Primarily designed for clinical use, not research.</li>
<li>Redundancy across data sources.</li>
</ul></li>
<li>Goal: derive research-grade phenotypes from raw clinical data.</li>
</ul>
</div>
</div>

<div id="outline-container-orge2ba6f4" class="outline-3">
<h3 id="orge2ba6f4"><span class="section-number-3">7.3.</span> Phenotyping Algorithm</h3>
<div class="outline-text-3" id="text-7-3">
<ul class="org-ul">
<li>Example: Type 2 diabetes phenotyping algorithm.</li>
<li>Sequential logic:
<ol class="org-ol">
<li>Exclude Type 1 diagnosis.</li>
<li>Check for Type 2 diagnosis.</li>
<li>Check medication and lab results.</li>
</ol></li>
<li>Multiple confirmation paths possible.</li>
<li>Developed manually by clinical experts.</li>
</ul>
</div>
</div>

<div id="outline-container-orga87cb9b" class="outline-3">
<h3 id="orga87cb9b"><span class="section-number-3">7.4.</span> Applications of Phenotyping</h3>
<div class="outline-text-3" id="text-7-4">
<ul class="org-ul">
<li>Genomic studies: link genomic and phenotypic data.</li>
<li>Clinical predictive modeling: forecast disease onset, hospitalizations.</li>
<li>Pragmatic clinical trials: assess treatment effectiveness in real-world settings.</li>
<li>Healthcare quality measurement: compare care quality across institutions.</li>
<li>All rely on robust phenotyping algorithms.</li>
</ul>
</div>
</div>

<div id="outline-container-org09351ce" class="outline-3">
<h3 id="org09351ce"><span class="section-number-3">7.5.</span> Genomic Wide Association Study</h3>
<div class="outline-text-3" id="text-7-5">
<ul class="org-ul">
<li>GWAS scans SNPs to associate genetic variations with diseases.</li>
<li>Steps:
<ol class="org-ol">
<li>Identify phenotypes.</li>
<li>Divide subjects into cases/controls.</li>
<li>Collect DNA and scan for SNPs.</li>
<li>Calculate odds ratios and p-values.</li>
</ol></li>
<li>Statistically significant SNPs may indicate disease relevance.</li>
<li>High-quality phenotyping critical for accurate GWAS.</li>
</ul>
</div>
</div>

<div id="outline-container-org7407d71" class="outline-3">
<h3 id="org7407d71"><span class="section-number-3">7.6.</span> Why Do We Care About Phenotyping</h3>
<div class="outline-text-3" id="text-7-6">
<ul class="org-ul">
<li>Genomic data is rapidly becoming cheaper and more abundant.</li>
<li>Phenotypic data remains complex and costly to acquire.</li>
<li>Demand for better phenotyping algorithms to support scalable, high-quality genomic research.</li>
</ul>
</div>
</div>

<div id="outline-container-org288d3fa" class="outline-3">
<h3 id="org288d3fa"><span class="section-number-3">7.7.</span> Clinical Predictive Modeling</h3>
<div class="outline-text-3" id="text-7-7">
<ul class="org-ul">
<li>Predictive modeling using raw EHR data is problematic (noise, complexity, lack of portability).</li>
<li>Phenotyping transforms raw data into simpler, meaningful concepts.</li>
<li>Benefits:
<ul class="org-ul">
<li>Better model accuracy.</li>
<li>Easier interpretation.</li>
<li>Generalizability across hospitals.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org6eadc1d" class="outline-3">
<h3 id="org6eadc1d"><span class="section-number-3">7.8.</span> Pragmatic Clinical Trials</h3>
<div class="outline-text-3" id="text-7-8">
<ul class="org-ul">
<li>Traditional vs. pragmatic trials:
<ul class="org-ul">
<li>Traditional: controlled, single condition/drug, strict selection, randomized.</li>
<li>Pragmatic: real-world, multiple conditions/drugs, minimal selection, often not randomized.</li>
</ul></li>
<li>High-quality phenotyping essential for identifying patient conditions and treatments in pragmatic settings.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc590ecb" class="outline-3">
<h3 id="orgc590ecb"><span class="section-number-3">7.9.</span> Healthcare Quality Measurement</h3>
<div class="outline-text-3" id="text-7-9">
<ul class="org-ul">
<li>Challenge: hospitals use diverse EHR formats.</li>
<li>Centralized processing of raw data is complex.</li>
<li>Solution: local phenotyping at hospitals → send standardized phenotype data to central site.</li>
<li>Enables scalable and fair quality comparison across institutions.</li>
</ul>
</div>
</div>

<div id="outline-container-org3bea276" class="outline-3">
<h3 id="org3bea276"><span class="section-number-3">7.10.</span> Phenotyping Methods Part 1</h3>
<div class="outline-text-3" id="text-7-10">
<ul class="org-ul">
<li>Two main approaches:
<ul class="org-ul">
<li>Supervised learning: uses labeled data (function approximation).</li>
<li>Unsupervised learning: identifies patterns/structures without labels (description/summarization).</li>
</ul></li>
<li>Illustrative examples and playful dialogue explore these ideas in depth.</li>
</ul>
</div>
</div>

<div id="outline-container-org9a0712f" class="outline-3">
<h3 id="org9a0712f"><span class="section-number-3">7.11.</span> Phenotyping Methods Part 2</h3>
<div class="outline-text-3" id="text-7-11">
<ul class="org-ul">
<li>Supervised: expert-defined rules (Boolean logic, decision trees), or machine learning classifiers.
<ul class="org-ul">
<li>Pros: interpretable, low validation effort.</li>
<li>Cons: high development cost, not good for unknown phenotypes, potential poor transferability.</li>
</ul></li>
<li>Unsupervised: clustering, dimensionality reduction.
<ul class="org-ul">
<li>Pros: less manual effort, no need for labels.</li>
<li>Cons: hard to validate, large data requirements.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgda607fe" class="outline-3">
<h3 id="orgda607fe"><span class="section-number-3">7.12.</span> Phenotyping Quiz Question</h3>
<div class="outline-text-3" id="text-7-12">
<ul class="org-ul">
<li>Quiz:
<ul class="org-ul">
<li>Which approach requires more human effort during evaluation?</li>
<li>Which is easier to interpret?</li>
<li>Choices: expert-defined rules vs. classification models.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org951b4c6" class="outline-3">
<h3 id="org951b4c6"><span class="section-number-3">7.13.</span> Phenotyping Quiz Solution</h3>
<div class="outline-text-3" id="text-7-13">
<ul class="org-ul">
<li>Answer:
<ul class="org-ul">
<li>Classification models require more evaluation effort (need more labeled data).</li>
<li>Expert-defined rules are easier to interpret (clinician-derived, intuitive).</li>
</ul></li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org23eb419" class="outline-2">
<h2 id="org23eb419"><span class="section-number-2">8.</span> Clustering</h2>
<div class="outline-text-2" id="text-8">
</div>
<div id="outline-container-orgee4c1ec" class="outline-3">
<h3 id="orgee4c1ec"><span class="section-number-3">8.1.</span> Introduction to Clustering</h3>
<div class="outline-text-3" id="text-8-1">
<ul class="org-ul">
<li>Introduction of clustering as a method distinct from classification.</li>
<li>Classification: Grouping by known labels.</li>
<li>Clustering: Discovering groups from raw data.</li>
<li>Topics covered:
<ul class="org-ul">
<li>Definition of clustering.</li>
<li>Algorithms: K-means, Gaussian Mixture Models (GMM).</li>
<li>Scalable algorithms for large datasets.</li>
<li>Applications in healthcare.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgb2132f6" class="outline-3">
<h3 id="orgb2132f6"><span class="section-number-3">8.2.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-8-2">
<ul class="org-ul">
<li>Clustering in healthcare:
<ul class="org-ul">
<li>Patient stratification: group patients with similar characteristics.</li>
<li>Disease hierarchy discovery: learn structure/relationships among diseases.</li>
<li>Phenotyping: convert raw data into meaningful phenotypes (clinical concepts).</li>
<li>Phenotypes = clusters of similar patients.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc9fe745" class="outline-3">
<h3 id="orgc9fe745"><span class="section-number-3">8.3.</span> What is Clustering</h3>
<div class="outline-text-3" id="text-8-3">
<ul class="org-ul">
<li>Clustering explained via patient-disease matrix:
<ul class="org-ul">
<li>Rows = patients, Columns = diseases.</li>
<li>Apply clustering to rows → patient clusters (e.g., P1, P2, P3).</li>
<li>Apply clustering to columns → disease clusters (e.g., D1, D2, D3).</li>
<li>Supports applications like phenotyping and disease discovery.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgabc499d" class="outline-3">
<h3 id="orgabc499d"><span class="section-number-3">8.4.</span> Algorithm Overview</h3>
<div class="outline-text-3" id="text-8-4">
<ul class="org-ul">
<li>Two categories of clustering algorithms:
<ul class="org-ul">
<li>Classical: K-means, Hierarchical, Gaussian Mixture Model.</li>
<li>Scalable: Mini-batch K-means, DBSCAN.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgd0d30b1" class="outline-3">
<h3 id="orgd0d30b1"><span class="section-number-3">8.5.</span> K Means</h3>
<div class="outline-text-3" id="text-8-5">
<ul class="org-ul">
<li>Inputs: data points X₁ to Xₙ, number of clusters K.</li>
<li>Outputs: K clusters S₁ to Sₖ.</li>
<li>Objective: Minimize distance of data points to cluster centers (µᵢ).</li>
<li>Algorithm steps:
<ol class="org-ol">
<li>Initialize K centers.</li>
<li>Assign points to nearest center.</li>
<li>Update centers based on mean of assigned points.</li>
<li>Repeat until convergence.</li>
</ol></li>
<li>Example illustrated using K=2.</li>
</ul>
</div>
</div>

<div id="outline-container-org98ac64b" class="outline-3">
<h3 id="org98ac64b"><span class="section-number-3">8.6.</span> K Means Quiz Question</h3>
<div class="outline-text-3" id="text-8-6">
<ul class="org-ul">
<li>Quiz: Given n, k, d (dimensions), i (iterations), what's the complexity?</li>
</ul>
</div>
</div>

<div id="outline-container-org0609258" class="outline-3">
<h3 id="org0609258"><span class="section-number-3">8.7.</span> K Means Quiz Solution</h3>
<div class="outline-text-3" id="text-8-7">
<ul class="org-ul">
<li>Complexity: O(n × k × d × i)
<ul class="org-ul">
<li>Cost driven by assignment and update operations.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc15e933" class="outline-3">
<h3 id="orgc15e933"><span class="section-number-3">8.8.</span> Hierarchical Clustering</h3>
<div class="outline-text-3" id="text-8-8">
<ul class="org-ul">
<li>Builds a hierarchy of data points.</li>
<li>Two approaches:
<ul class="org-ul">
<li>Agglomerative (bottom-up): Merge smallest clusters until one remains.</li>
<li>Divisive (top-down): Split one big cluster into smaller ones.</li>
</ul></li>
<li>Agglomerative more commonly used.</li>
</ul>
</div>
</div>

<div id="outline-container-org8d35181" class="outline-3">
<h3 id="org8d35181"><span class="section-number-3">8.9.</span> Agglomerative Clustering</h3>
<div class="outline-text-3" id="text-8-9">
<ul class="org-ul">
<li>Steps:
<ol class="org-ol">
<li>Compute n×n distance matrix.</li>
<li>Initialize each point as its own cluster.</li>
<li>Iteratively merge closest clusters and update distance matrix.</li>
<li>Stop when one cluster remains (produces a hierarchy).</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org8a18607" class="outline-3">
<h3 id="org8a18607"><span class="section-number-3">8.10.</span> Gaussian Mixture Model</h3>
<div class="outline-text-3" id="text-8-10">
<ul class="org-ul">
<li>Soft clustering: points belong to multiple clusters with probabilities.</li>
<li>Each cluster is a Gaussian (normal distribution).</li>
<li>Parameters: mean (µₖ), variance (σₖ), mixing coefficient (πₖ).</li>
<li>Goal: Estimate GMM parameters from data.</li>
</ul>
</div>
</div>

<div id="outline-container-org8e70134" class="outline-3">
<h3 id="org8e70134"><span class="section-number-3">8.11.</span> GMM Expectation Maximization</h3>
<div class="outline-text-3" id="text-8-11">
<ul class="org-ul">
<li>Uses Expectation-Maximization (EM) algorithm:
<ul class="org-ul">
<li>Initialize parameters.</li>
<li>E-step: Compute assignment scores (γₙₖ).</li>
<li>M-step: Update parameters using γₙₖ.</li>
<li>Iterate until convergence.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7c856e5" class="outline-3">
<h3 id="org7c856e5"><span class="section-number-3">8.12.</span> GMM Steps</h3>
<div class="outline-text-3" id="text-8-12">
<ul class="org-ul">
<li>Details of EM:
<ul class="org-ul">
<li>Initialization: K-means can be used for better results.</li>
<li>E-step: Compute γₙₖ using cluster probabilities.</li>
<li>M-step:
<ul class="org-ul">
<li>Compute cluster size (Nₖ), mean (µₖ), covariance (σₖ), mixing coefficient (πₖ).</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc768231" class="outline-3">
<h3 id="orgc768231"><span class="section-number-3">8.13.</span> GMM Visual Illustration</h3>
<div class="outline-text-3" id="text-8-13">
<ul class="org-ul">
<li>Visual demo of GMM clustering.</li>
<li>Iteratively updates Gaussians until convergence.</li>
<li>Final model reflects soft cluster assignments.</li>
</ul>
</div>
</div>

<div id="outline-container-org41d33b7" class="outline-3">
<h3 id="org41d33b7"><span class="section-number-3">8.14.</span> K Means Vs GMM</h3>
<div class="outline-text-3" id="text-8-14">
<ul class="org-ul">
<li>Comparison:
<ul class="org-ul">
<li>K-means: hard assignment, single center (µₖ).</li>
<li>GMM: soft assignment, parameters include µₖ, σₖ, πₖ.</li>
<li>Both use iterative updates.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf492ce1" class="outline-3">
<h3 id="orgf492ce1"><span class="section-number-3">8.15.</span> Mini Batch K Means</h3>
<div class="outline-text-3" id="text-8-15">
<ul class="org-ul">
<li>Scalable version of K-means.</li>
<li>Works with small batches (M) instead of full dataset.</li>
<li>Updates centers incrementally.</li>
<li>Reduces computational cost for large data.</li>
</ul>
</div>
</div>

<div id="outline-container-org48f28dc" class="outline-3">
<h3 id="org48f28dc"><span class="section-number-3">8.16.</span> Mini Batch K Means Quiz Question</h3>
<div class="outline-text-3" id="text-8-16">
<ul class="org-ul">
<li>Quiz: Given k, t (iterations), b (batch size), d (dimensionality), compute complexity.</li>
</ul>
</div>
</div>

<div id="outline-container-org7b11b77" class="outline-3">
<h3 id="org7b11b77"><span class="section-number-3">8.17.</span> Mini Batch K Means Quiz Solution</h3>
<div class="outline-text-3" id="text-8-17">
<ul class="org-ul">
<li>Complexity: O(t × b × k × d)</li>
</ul>
</div>
</div>

<div id="outline-container-org58b8401" class="outline-3">
<h3 id="org58b8401"><span class="section-number-3">8.18.</span> DBScan</h3>
<div class="outline-text-3" id="text-8-18">
<ul class="org-ul">
<li>Density-Based Spatial Clustering.</li>
<li>Clusters = dense areas, separated by sparse regions.</li>
<li>Can identify clusters of arbitrary shape.</li>
<li>Effective for noisy data.</li>
</ul>
</div>
</div>

<div id="outline-container-org9a8f114" class="outline-3">
<h3 id="org9a8f114"><span class="section-number-3">8.19.</span> DBScan Key Concepts</h3>
<div class="outline-text-3" id="text-8-19">
<ul class="org-ul">
<li>Density = number of points within ε-distance.</li>
<li>Key concepts:
<ul class="org-ul">
<li>Core points: high-density.</li>
<li>Border points: near core points.</li>
<li>Noise: not in dense region.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgfc8d9c5" class="outline-3">
<h3 id="orgfc8d9c5"><span class="section-number-3">8.20.</span> DBScan Algorithm</h3>
<div class="outline-text-3" id="text-8-20">
<ul class="org-ul">
<li>Steps:
<ol class="org-ol">
<li>Build graph connecting core points to neighbors.</li>
<li>Identify connected components as clusters.</li>
<li>Remove clustered points and repeat.</li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org0706ca1" class="outline-3">
<h3 id="org0706ca1"><span class="section-number-3">8.21.</span> DBScan Example</h3>
<div class="outline-text-3" id="text-8-21">
<ul class="org-ul">
<li>Visual example: finds 6 clusters, some noise points remain.</li>
</ul>
</div>
</div>

<div id="outline-container-org32ec1d5" class="outline-3">
<h3 id="org32ec1d5"><span class="section-number-3">8.22.</span> DBScan Quiz Question</h3>
<div class="outline-text-3" id="text-8-22">
<ul class="org-ul">
<li>Quiz: How many clusters can a point belong to in DBSCAN?</li>
</ul>
</div>
</div>

<div id="outline-container-org28a0899" class="outline-3">
<h3 id="org28a0899"><span class="section-number-3">8.23.</span> DBScan Quiz Solution</h3>
<div class="outline-text-3" id="text-8-23">
<ul class="org-ul">
<li>Answer: 0 or 1 (noise = 0, core/border = 1)</li>
</ul>
</div>
</div>

<div id="outline-container-orge5a4e6a" class="outline-3">
<h3 id="orge5a4e6a"><span class="section-number-3">8.24.</span> Clustering Evaluation Metrics</h3>
<div class="outline-text-3" id="text-8-24">
<ul class="org-ul">
<li>Metrics:
<ul class="org-ul">
<li>Rand Index (RI)</li>
<li>Mutual Information (MI)</li>
<li>Silhouette Coefficient</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org51148f0" class="outline-3">
<h3 id="org51148f0"><span class="section-number-3">8.25.</span> Rand Index</h3>
<div class="outline-text-3" id="text-8-25">
<ul class="org-ul">
<li>Measures agreement between predicted and true clusters.</li>
<li>RI = (number of correct pairs) / (total pairs)</li>
<li>Value between 0 (bad) and 1 (perfect).</li>
</ul>
</div>
</div>

<div id="outline-container-org92faca6" class="outline-3">
<h3 id="org92faca6"><span class="section-number-3">8.26.</span> Mutual Information</h3>
<div class="outline-text-3" id="text-8-26">
<ul class="org-ul">
<li>From information theory: measures shared info between predicted and true clusters.</li>
<li>Normalized Mutual Information (NMI) between 0 and 1.</li>
</ul>
</div>
</div>

<div id="outline-container-org2656fb0" class="outline-3">
<h3 id="org2656fb0"><span class="section-number-3">8.27.</span> Summary of RI and MI</h3>
<div class="outline-text-3" id="text-8-27">
<ul class="org-ul">
<li>Pros:
<ul class="org-ul">
<li>Bounded values, interpretable.</li>
<li>No assumptions about cluster shapes.</li>
</ul></li>
<li>Cons:
<ul class="org-ul">
<li>Require ground truth, which may be unavailable.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orge75fc57" class="outline-3">
<h3 id="orge75fc57"><span class="section-number-3">8.28.</span> Silhouette Coefficient</h3>
<div class="outline-text-3" id="text-8-28">
<ul class="org-ul">
<li>Measures how similar a point is to its own cluster vs. others.</li>
<li>Formula: (b - a) / max(a, b)</li>
<li>Does not require ground truth.</li>
</ul>
</div>
</div>

<div id="outline-container-org8f106b9" class="outline-3">
<h3 id="org8f106b9"><span class="section-number-3">8.29.</span> Silhouette Coefficient Pros and Cons</h3>
<div class="outline-text-3" id="text-8-29">
<ul class="org-ul">
<li>Range: -1 (bad) to 1 (good), 0 = overlapping clusters.</li>
<li>Assumes spherical clusters → less effective for complex shapes (e.g. DBSCAN).</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgc6c26a3" class="outline-2">
<h2 id="orgc6c26a3"><span class="section-number-2">9.</span> Spark</h2>
<div class="outline-text-2" id="text-9">
</div>
<div id="outline-container-orgd030e5f" class="outline-3">
<h3 id="orgd030e5f"><span class="section-number-3">9.1.</span> Introduction to Spark</h3>
<div class="outline-text-3" id="text-9-1">
<ul class="org-ul">
<li>Spark is introduced as an alternative to MapReduce, better suited for iterative workloads.</li>
<li>Leverages distributed memory across machines.</li>
<li>Core concept: Resilient Distributed Dataset (RDD).</li>
<li>Promotes efficiency for machine learning and healthcare applications.</li>
</ul>
</div>
</div>

<div id="outline-container-org8ce1ea5" class="outline-3">
<h3 id="org8ce1ea5"><span class="section-number-3">9.2.</span> Environment</h3>
<div class="outline-text-3" id="text-9-2">
<ul class="org-ul">
<li>Big data analytics typically run in data centers or cloud platforms like AWS, GCP, Azure.</li>
<li>Environment includes interconnected racks of servers.</li>
<li>Context set to understand the need for systems like Spark.</li>
</ul>
</div>
</div>

<div id="outline-container-org00714f1" class="outline-3">
<h3 id="org00714f1"><span class="section-number-3">9.3.</span> Motivation</h3>
<div class="outline-text-3" id="text-9-3">
<ul class="org-ul">
<li>Hadoop &amp; MapReduce explained with acyclic data flow using stable storage.</li>
<li>Strength: fault-tolerance via disk storage.</li>
<li>Weakness: inefficiency with iterative or interactive workloads (e.g. ML, graph analysis).</li>
<li>Need for in-memory computation system like Spark.</li>
</ul>
</div>
</div>

<div id="outline-container-org56d5fc8" class="outline-3">
<h3 id="org56d5fc8"><span class="section-number-3">9.4.</span> Iteration in Map Reduce</h3>
<div class="outline-text-3" id="text-9-4">
<ul class="org-ul">
<li>Machine learning models need repeated computation.</li>
<li>Hadoop repeatedly reads/writes from disk, causing inefficiency.</li>
<li>Highlights the need for in-memory operations.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbc7767f" class="outline-3">
<h3 id="orgbc7767f"><span class="section-number-3">9.5.</span> Workload Illustration</h3>
<div class="outline-text-3" id="text-9-5">
<ul class="org-ul">
<li>Iterative workloads: same computation on same data (e.g., model training).</li>
<li>Interactive workloads: varying computations on evolving data (e.g., data exploration).</li>
<li>Key: keep working sets in memory for speed.</li>
</ul>
</div>
</div>

<div id="outline-container-org8cd0668" class="outline-3">
<h3 id="org8cd0668"><span class="section-number-3">9.6.</span> Challenge</h3>
<div class="outline-text-3" id="text-9-6">
<ul class="org-ul">
<li>Challenge: memory is not fault-tolerant.</li>
<li>Existing storage abstractions rely on fine-grained updates, which are expensive to track.</li>
<li>Need for a better balance between efficiency and fault tolerance.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf639a9e" class="outline-3">
<h3 id="orgf639a9e"><span class="section-number-3">9.7.</span> Solution  RDDs</h3>
<div class="outline-text-3" id="text-9-7">
<ul class="org-ul">
<li>Spark uses RDDs to balance fault-tolerance and efficiency.</li>
<li>RDDs support coarse-grained transformations (e.g., map, group-by, join).</li>
<li>Lineage tracking enables fault recovery without excessive replication.</li>
</ul>
</div>
</div>

<div id="outline-container-org9807cd6" class="outline-3">
<h3 id="org9807cd6"><span class="section-number-3">9.8.</span> RDD Recovery</h3>
<div class="outline-text-3" id="text-9-8">
<ul class="org-ul">
<li>Recovery through recomputation based on lineage.</li>
<li>Interactive: rerun only failed iterations.</li>
<li>Iterative: reload or recompute chunks from previous iterations.</li>
</ul>
</div>
</div>

<div id="outline-container-orga866770" class="outline-3">
<h3 id="orga866770"><span class="section-number-3">9.9.</span> Spark Stack</h3>
<div class="outline-text-3" id="text-9-9">
<ul class="org-ul">
<li>Core: scheduling, memory management, fault recovery.</li>
<li>APIs: Spark SQL, Streaming, MLib (ML), GraphX.</li>
<li>Compatible with YARN, Mesos, or standalone.</li>
</ul>
</div>
</div>

<div id="outline-container-orge636dfc" class="outline-3">
<h3 id="orge636dfc"><span class="section-number-3">9.10.</span> Spark Programming Interface</h3>
<div class="outline-text-3" id="text-9-10">
<ul class="org-ul">
<li>Interfaces: Scala, Python, Java.</li>
<li>RDD operations: transformations and actions.</li>
<li>Support for shared variables (broadcast, accumulator).</li>
</ul>
</div>
</div>

<div id="outline-container-org4bf6605" class="outline-3">
<h3 id="org4bf6605"><span class="section-number-3">9.11.</span> RDD Transformations</h3>
<div class="outline-text-3" id="text-9-11">
<ul class="org-ul">
<li>Examples: map vs flatMap, distinct, union, intersection, subtract.</li>
<li>flatMap flattens nested lists from map.</li>
<li>Intersection and subtract are more costly due to set operations.</li>
</ul>
</div>
</div>

<div id="outline-container-orgdc39326" class="outline-3">
<h3 id="orgdc39326"><span class="section-number-3">9.12.</span> RDD Transformations Quiz Question</h3>
<div class="outline-text-3" id="text-9-12">
<ul class="org-ul">
<li>Quiz: Apply map and filter on input RDD to understand transformation output.</li>
</ul>
</div>
</div>

<div id="outline-container-org4f34576" class="outline-3">
<h3 id="org4f34576"><span class="section-number-3">9.13.</span> RDD Transformations Quiz Solution</h3>
<div class="outline-text-3" id="text-9-13">
<ul class="org-ul">
<li>Map squares each number: [1, 4, 9, 16].</li>
<li>Filter excludes ‘1’: resulting in [2, 3, 4].</li>
</ul>
</div>
</div>

<div id="outline-container-orgc2adbbf" class="outline-3">
<h3 id="orgc2adbbf"><span class="section-number-3">9.14.</span> Spark Operations</h3>
<div class="outline-text-3" id="text-9-14">
<ul class="org-ul">
<li>Transformations: map, filter, groupByKey, etc.</li>
<li>Actions: reduce, collect, count, etc.</li>
<li>Broader set than Hadoop MapReduce.</li>
</ul>
</div>
</div>

<div id="outline-container-org4d2da13" class="outline-3">
<h3 id="org4d2da13"><span class="section-number-3">9.15.</span> Shared Variable</h3>
<div class="outline-text-3" id="text-9-15">
<ul class="org-ul">
<li>Broadcast variables: efficient read-only shared data.</li>
<li>Example Spark job with clinical notes:
<ul class="org-ul">
<li>Load data → filter → split → cache.</li>
<li>Perform actions like counting specific symptoms (e.g., fever, cough).</li>
<li>Benefits from caching and distributed execution.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf7e2505" class="outline-3">
<h3 id="orgf7e2505"><span class="section-number-3">9.16.</span> Fault Tolerance</h3>
<div class="outline-text-3" id="text-9-16">
<ul class="org-ul">
<li>RDD tracks transformations (lineage).</li>
<li>Lost partitions can be recomputed using lineage.</li>
<li>Visuals: sequence of filters and maps reconstruct lost data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcbb1d0d" class="outline-3">
<h3 id="orgcbb1d0d"><span class="section-number-3">9.17.</span> Example  Logistic Regression</h3>
<div class="outline-text-3" id="text-9-17">
<ul class="org-ul">
<li>Spark efficiently supports logistic regression via iteration.</li>
<li>Data cached in memory → quick updates.</li>
<li>Uses map + reduce to compute gradient.</li>
<li>Comparison: Spark (6s/iter) vs Hadoop (127s/iter).</li>
</ul>
</div>
</div>

<div id="outline-container-org74b871a" class="outline-3">
<h3 id="org74b871a"><span class="section-number-3">9.18.</span> Example  Disease Risk Prediction</h3>
<div class="outline-text-3" id="text-9-18">
<ul class="org-ul">
<li>Predict patient disease risks using collaborative filtering.</li>
<li>Matrix R = patient × disease.</li>
<li>ALS algorithm: alternates updating patient and disease features.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb6cca43" class="outline-3">
<h3 id="orgb6cca43"><span class="section-number-3">9.19.</span> Serial ALS</h3>
<div class="outline-text-3" id="text-9-19">
<ul class="org-ul">
<li>Serial implementation of ALS.</li>
<li>Iterative updates of A (patient) and B (disease) matrices.</li>
<li>Uses map functions over patient/disease ranges.</li>
</ul>
</div>
</div>

<div id="outline-container-orge27a33c" class="outline-3">
<h3 id="orge27a33c"><span class="section-number-3">9.20.</span> Naive Spark ALS</h3>
<div class="outline-text-3" id="text-9-20">
<ul class="org-ul">
<li>Parallel ALS using `spark.parallelize`.</li>
<li>Broadcast used for efficient data sharing.</li>
<li>Broadcast improves performance 3× over naive version.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-orgc70f685" class="outline-2">
<h2 id="orgc70f685"><span class="section-number-2">10.</span> Medical Ontology</h2>
<div class="outline-text-2" id="text-10">
</div>
<div id="outline-container-org8ae7750" class="outline-3">
<h3 id="org8ae7750"><span class="section-number-3">10.1.</span> Introduction to Medical Ontology</h3>
<div class="outline-text-3" id="text-10-1">
<ul class="org-ul">
<li>Healthcare benefits from structured medical knowledge.</li>
<li>Ontologies and knowledge graphs are common in this domain.</li>
<li>Historically, healthcare has developed ontologies for:
<ul class="org-ul">
<li>Diseases</li>
<li>Medical procedures</li>
<li>Medications</li>
<li>Lab tests</li>
</ul></li>
<li>Ontologies aid understanding of healthcare data and validate data models.</li>
<li>The lesson discusses various ontologies and their role in analytics.</li>
</ul>
</div>
</div>

<div id="outline-container-org07d71b8" class="outline-3">
<h3 id="org07d71b8"><span class="section-number-3">10.2.</span> Health Data Standards</h3>
<div class="outline-text-3" id="text-10-2">
<ul class="org-ul">
<li>Overview of major health data standards:
<ul class="org-ul">
<li>LOINC: Lab test codes</li>
<li>ICD: Diagnostic codes</li>
<li>CPT: Procedure codes</li>
<li>NDC: Medication codes</li>
</ul></li>
<li>SNOMED: A comprehensive ontology integrating all the above.</li>
<li>UMLS: Software system that accesses integrated medical knowledge.</li>
<li>Use case: Patient encounter example illustrating all standard usages.</li>
<li>Importance:
<ul class="org-ul">
<li>Insurance claim processing</li>
<li>Research and data analysis</li>
</ul></li>
<li>Preview of upcoming detailed discussions on individual standards.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf217b0f" class="outline-3">
<h3 id="orgf217b0f"><span class="section-number-3">10.3.</span> ICD</h3>
<div class="outline-text-3" id="text-10-3">
<ul class="org-ul">
<li>ICD = International Classification of Diseases.</li>
<li>Developed by WHO, categorizes diseases.</li>
<li>ICD-9: 17,000+ codes; 3-5 digits; includes main and supplementary categories (e.g., E, V codes).</li>
<li>ICD-10: 141,000+ codes; 7-character alphanumeric structure.</li>
<li>Detailed coding for category, etiology, body part, severity, and extension.</li>
<li>Examples:
<ul class="org-ul">
<li>ICD-9: 250.01 = Type 1 diabetes, no complications</li>
<li>ICD-10: M1A.31X2 = Chronic gout with renal impairment in left shoulder, no tophus</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org95492e6" class="outline-3">
<h3 id="org95492e6"><span class="section-number-3">10.4.</span> ICD 9 to ICD 10 Mapping</h3>
<div class="outline-text-3" id="text-10-4">
<ul class="org-ul">
<li>Mapping is mostly one-to-many due to ICD-10's higher specificity.</li>
<li>Some mappings are one-to-one (e.g., Tietze's Syndrome).</li>
<li>Complicated mappings: e.g., ICD-9 code maps to 2,530 ICD-10 codes.</li>
<li>Complexity highlights the increased detail of ICD-10.</li>
</ul>
</div>
</div>

<div id="outline-container-org048df9e" class="outline-3">
<h3 id="org048df9e"><span class="section-number-3">10.5.</span> ICD 9 Quiz Question</h3>
<div class="outline-text-3" id="text-10-5">
<ul class="org-ul">
<li>Identifies valid and invalid ICD-9 codes.</li>
</ul>
</div>
</div>

<div id="outline-container-org13aa6f8" class="outline-3">
<h3 id="org13aa6f8"><span class="section-number-3">10.6.</span> ICD 9 Quiz Solution</h3>
<div class="outline-text-3" id="text-10-6">
<ul class="org-ul">
<li>Valid: 501, 802.3, V70, E820.0</li>
<li>Invalid: U80.1, 5A0.01</li>
</ul>
</div>
</div>

<div id="outline-container-org2046b6b" class="outline-3">
<h3 id="org2046b6b"><span class="section-number-3">10.7.</span> ICD Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-7">
<ul class="org-ul">
<li>Task: Find ICD-9 and ICD-10 codes for Influenza.</li>
</ul>
</div>
</div>

<div id="outline-container-org814542c" class="outline-3">
<h3 id="org814542c"><span class="section-number-3">10.8.</span> ICD Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-8">
<ul class="org-ul">
<li>ICD-9: 487.1</li>
<li>ICD-10: J11.1</li>
</ul>
</div>
</div>

<div id="outline-container-org3431a3d" class="outline-3">
<h3 id="org3431a3d"><span class="section-number-3">10.9.</span> CPT</h3>
<div class="outline-text-3" id="text-10-9">
<ul class="org-ul">
<li>CPT = Current Procedural Terminology (US standard).</li>
<li>Used for medical, surgical, diagnostic procedures.</li>
<li>Maintained by the AMA.</li>
<li>Key in insurance reimbursement.</li>
<li>Categories:
<ul class="org-ul">
<li>Category 1: Common procedures (e.g., surgery, radiology)</li>
<li>Category 2: Quality metrics</li>
<li>Category 3: Experimental procedures (e.g., codes ending in T)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org43bc6e8" class="outline-3">
<h3 id="org43bc6e8"><span class="section-number-3">10.10.</span> CPT Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-10">
<ul class="org-ul">
<li>Task: Find CPT code for detailed office visit.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcb42e54" class="outline-3">
<h3 id="orgcb42e54"><span class="section-number-3">10.11.</span> CPT Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-11">
<ul class="org-ul">
<li>Office visit range: 99201–99205</li>
<li>Time-based codes (e.g., 99201 = 10 mins, 99205 = 1 hour)</li>
</ul>
</div>
</div>

<div id="outline-container-org554e698" class="outline-3">
<h3 id="org554e698"><span class="section-number-3">10.12.</span> LOINC</h3>
<div class="outline-text-3" id="text-10-12">
<ul class="org-ul">
<li>LOINC = Logical Observation Identifiers Names and Codes.</li>
<li>Created by Regenstrief Institute.</li>
<li>Standard for lab tests and clinical observations.</li>
<li>Example: 2865-4 = a specific lab test</li>
<li>Attributes:
<ul class="org-ul">
<li>Component</li>
<li>Property</li>
<li>Time aspect (e.g., Pt = point in time)</li>
<li>Sample type</li>
<li>Scale (quantitative, ordinal, etc.)</li>
<li>Method</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgb1c3ddd" class="outline-3">
<h3 id="orgb1c3ddd"><span class="section-number-3">10.13.</span> LOINC Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-13">
<ul class="org-ul">
<li>Task: Find LOINC code for Creatinine lab test.</li>
</ul>
</div>
</div>

<div id="outline-container-org48eddea" class="outline-3">
<h3 id="org48eddea"><span class="section-number-3">10.14.</span> LOINC Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-14">
<ul class="org-ul">
<li>Answer: 2160-0</li>
</ul>
</div>
</div>

<div id="outline-container-orgfc4405b" class="outline-3">
<h3 id="orgfc4405b"><span class="section-number-3">10.15.</span> NDC</h3>
<div class="outline-text-3" id="text-10-15">
<ul class="org-ul">
<li>NDC = National Drug Code (for medications).</li>
<li>Maintained by the FDA.</li>
<li>Structure: 3 segments
<ul class="org-ul">
<li>Labeler code (company)</li>
<li>Product code (drug type)</li>
<li>Package code (package details)</li>
</ul></li>
<li>Different digit combinations exist but total is 10 digits.</li>
</ul>
</div>
</div>

<div id="outline-container-orgb9ae90e" class="outline-3">
<h3 id="orgb9ae90e"><span class="section-number-3">10.16.</span> NDC Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-16">
<ul class="org-ul">
<li>Task: Find NDC code for metformin hydrochloride, 500mg.</li>
</ul>
</div>
</div>

<div id="outline-container-orgded009e" class="outline-3">
<h3 id="orgded009e"><span class="section-number-3">10.17.</span> NDC Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-17">
<ul class="org-ul">
<li>Example code: 0093-7214-01 (labeler-dependent)</li>
</ul>
</div>
</div>

<div id="outline-container-orgf989e85" class="outline-3">
<h3 id="orgf989e85"><span class="section-number-3">10.18.</span> SNOMED</h3>
<div class="outline-text-3" id="text-10-18">
<ul class="org-ul">
<li>SNOMED = Systematized Nomenclature of Medicine.</li>
<li>Maintained by IHTSDO (Denmark).</li>
<li>Supports clinical documentation and semantic interoperability.</li>
<li>National editions and implementation subsets exist.</li>
<li>Broad user base: clinicians, researchers, analysts.</li>
</ul>
</div>
</div>

<div id="outline-container-org5b77bdb" class="outline-3">
<h3 id="org5b77bdb"><span class="section-number-3">10.19.</span> Logical Model of SNOMED CT</h3>
<div class="outline-text-3" id="text-10-19">
<ul class="org-ul">
<li>Core components:
<ul class="org-ul">
<li>Concepts (with unique ID)</li>
<li>Descriptions (FSN and synonyms)</li>
<li>Relationships (e.g., “is a”, attributes)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgf918cbf" class="outline-3">
<h3 id="orgf918cbf"><span class="section-number-3">10.20.</span> SNOMED Example</h3>
<div class="outline-text-3" id="text-10-20">
<ul class="org-ul">
<li>Example concept: 22298006 = Myocardial infarction disorder</li>
<li>Multiple descriptions including synonyms</li>
<li>Preferred vs. acceptable terms are defined by reference sets.</li>
</ul>
</div>
</div>

<div id="outline-container-org9b2c032" class="outline-3">
<h3 id="org9b2c032"><span class="section-number-3">10.21.</span> SNOMED Relationships</h3>
<div class="outline-text-3" id="text-10-21">
<ul class="org-ul">
<li>“Is a” relationship creates concept hierarchies.</li>
<li>Example: Cellulitis of foot “is a” cellulitis, and “is a” disorder of foot.</li>
<li>Relationships are directional and acyclic.</li>
</ul>
</div>
</div>

<div id="outline-container-org6e41fb1" class="outline-3">
<h3 id="org6e41fb1"><span class="section-number-3">10.22.</span> SNOMED Design</h3>
<div class="outline-text-3" id="text-10-22">
<ul class="org-ul">
<li>19-level hierarchy of concepts.</li>
<li>Concepts range from general to highly specific.</li>
<li>Relationships include “is a” and others with attributes.</li>
<li>Each concept has a unique ID and descriptions.</li>
</ul>
</div>
</div>

<div id="outline-container-org5f403b8" class="outline-3">
<h3 id="org5f403b8"><span class="section-number-3">10.23.</span> SNOMED Code Quiz Question</h3>
<div class="outline-text-3" id="text-10-23">
<ul class="org-ul">
<li>Task: Find SNOMED code for chronic gouty arthritis.</li>
</ul>
</div>
</div>

<div id="outline-container-orgbb0e50b" class="outline-3">
<h3 id="orgbb0e50b"><span class="section-number-3">10.24.</span> SNOMED Code Quiz Solution</h3>
<div class="outline-text-3" id="text-10-24">
<ul class="org-ul">
<li>Answer: 68451005</li>
</ul>
</div>
</div>

<div id="outline-container-org1decd74" class="outline-3">
<h3 id="org1decd74"><span class="section-number-3">10.25.</span> SNOMED Quiz Question</h3>
<div class="outline-text-3" id="text-10-25">
<ul class="org-ul">
<li>Question: What is the structure of "is a" relationships in SNOMED?</li>
</ul>
</div>
</div>

<div id="outline-container-org3f8dfbb" class="outline-3">
<h3 id="org3f8dfbb"><span class="section-number-3">10.26.</span> SNOMED Quiz Solution</h3>
<div class="outline-text-3" id="text-10-26">
<ul class="org-ul">
<li>Answer: Directed acyclic graph (not a tree).</li>
</ul>
</div>
</div>

<div id="outline-container-org6961461" class="outline-3">
<h3 id="org6961461"><span class="section-number-3">10.27.</span> UMLS</h3>
<div class="outline-text-3" id="text-10-27">
<ul class="org-ul">
<li>UMLS = Unified Medical Language System.</li>
<li>Maintained by US National Library of Medicine.</li>
<li>Integrates multiple standards into a unified system.</li>
<li>Components:
<ul class="org-ul">
<li>Metathesaurus</li>
<li>Semantic network</li>
<li>Specialist lexicon and tools</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgbb87bf0" class="outline-3">
<h3 id="orgbb87bf0"><span class="section-number-3">10.28.</span> Metathesaurus Concepts</h3>
<div class="outline-text-3" id="text-10-28">
<ul class="org-ul">
<li>Concept hierarchy:
<ul class="org-ul">
<li>Atom (AUI) → String (SUI) → Term (LUI) → Concept (CUI)</li>
</ul></li>
<li>Multiple ontologies/sources integrated.</li>
</ul>
</div>
</div>

<div id="outline-container-org215f722" class="outline-3">
<h3 id="org215f722"><span class="section-number-3">10.29.</span> Semantic Network</h3>
<div class="outline-text-3" id="text-10-29">
<ul class="org-ul">
<li>135+ semantic types (e.g., diseases, drugs)</li>
<li>54+ semantic relationships (e.g., cause, treat)</li>
<li>Organizes concepts and their interrelations.</li>
</ul>
</div>
</div>

<div id="outline-container-orge848990" class="outline-3">
<h3 id="orge848990"><span class="section-number-3">10.30.</span> Specialist Lexicon</h3>
<div class="outline-text-3" id="text-10-30">
<ul class="org-ul">
<li>Lexicon of 300,000+ biomedical terms.</li>
<li>Includes syntax, morphology, orthography.</li>
<li>Used in NLP tools like MetaMap and MMTx.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org9ef6167" class="outline-2">
<h2 id="org9ef6167"><span class="section-number-2">11.</span> Graph Analysis</h2>
<div class="outline-text-2" id="text-11">
</div>
<div id="outline-container-org501aa1f" class="outline-3">
<h3 id="org501aa1f"><span class="section-number-3">11.1.</span> Introduction to Graph Analysis</h3>
<div class="outline-text-3" id="text-11-1">
<ul class="org-ul">
<li>Overview of graph analysis and its applications.</li>
<li>Used in search engines, social networks, and healthcare.</li>
<li>Identifies clusters of related entities (webpages, patients, conditions).</li>
<li>Will cover similarity graphs and spectral clustering.</li>
</ul>
</div>
</div>

<div id="outline-container-org70c2c64" class="outline-3">
<h3 id="org70c2c64"><span class="section-number-3">11.2.</span> Agenda</h3>
<div class="outline-text-3" id="text-11-2">
<ul class="org-ul">
<li>Two main graph-based algorithms introduced:
<ul class="org-ul">
<li>PageRank: Ranks nodes in a directed graph based on importance.</li>
<li>Spectral Clustering: A graph partitioning-based clustering algorithm.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org56ab366" class="outline-3">
<h3 id="org56ab366"><span class="section-number-3">11.3.</span> PageRank</h3>
<div class="outline-text-3" id="text-11-3">
<ul class="org-ul">
<li>Developed by Google co-founders for webpage ranking.</li>
<li>Traditional content-based ranking is prone to spam.</li>
<li>PageRank relies on the graph structure (links between pages).</li>
<li>Intuition: More high-quality pages linking to a page → higher rank.</li>
<li>Represent graph using an adjacency matrix:
<ul class="org-ul">
<li>Normalize matrix rows to represent transition probabilities.</li>
<li>PageRank computed using recursion: browsing + teleporting.</li>
<li>Browsing redistributes rank via adjacency matrix; teleporting randomly jumps.</li>
<li>Equation: `q = cAq + (1 - c)e/N`, where `c` is damping factor (e.g., 0.85).</li>
</ul></li>
<li>Next steps: Implement PageRank using big data systems like Hadoop MapReduce.</li>
</ul>
</div>
</div>

<div id="outline-container-orgee70169" class="outline-3">
<h3 id="orgee70169"><span class="section-number-3">11.4.</span> MapReduce  PageRank</h3>
<div class="outline-text-3" id="text-11-4">
<ul class="org-ul">
<li>Implementation of PageRank in MapReduce:
<ul class="org-ul">
<li><b><b>Map Phase</b></b>:
<ul class="org-ul">
<li>Emit current page and its PageRank.</li>
<li>Distribute PageRank to each outgoing link equally.</li>
</ul></li>
<li><b><b>Reduce Phase</b></b>:
<ul class="org-ul">
<li>Aggregate partial PageRanks from map phase.</li>
<li>Add teleporting component.</li>
<li>Emit updated PageRank.</li>
</ul></li>
<li>Hadoop shuffles partial sums to reducers.</li>
<li>Iterative process: repeat map-reduce steps to converge.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgd8e5454" class="outline-3">
<h3 id="orgd8e5454"><span class="section-number-3">11.5.</span> PageRank Quiz Question</h3>
<div class="outline-text-3" id="text-11-5">
<ul class="org-ul">
<li>Quiz: Rank webpages in a directed graph by PageRank (based on structure).</li>
</ul>
</div>
</div>

<div id="outline-container-org37373f0" class="outline-3">
<h3 id="org37373f0"><span class="section-number-3">11.6.</span> PageRank Quiz Solution</h3>
<div class="outline-text-3" id="text-11-6">
<ul class="org-ul">
<li>Highest: YouTube (3 incoming links).</li>
<li>Lowest: Google (no incoming links).</li>
<li>Recursive PageRank shows differences despite similar local structures.</li>
<li>Amazon &gt; Wikipedia &gt; Facebook in mid-tier due to recursion effects.</li>
</ul>
</div>
</div>

<div id="outline-container-orgec0e666" class="outline-3">
<h3 id="orgec0e666"><span class="section-number-3">11.7.</span> Spectral Clustering</h3>
<div class="outline-text-3" id="text-11-7">
<ul class="org-ul">
<li>Spectral Clustering vs Traditional Clustering:
<ul class="org-ul">
<li>Uses similarity graph of patients based on features.</li>
<li>Steps:
<ol class="org-ol">
<li>Construct similarity graph (nodes = patients, edges = similarity).</li>
<li>Convert graph to adjacency matrix.</li>
<li>Compute top-k eigenvectors of the matrix.</li>
<li>Cluster eigenvectors into groups.</li>
</ol></li>
<li>Effective for high-dimensional or noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org84a8d0e" class="outline-3">
<h3 id="org84a8d0e"><span class="section-number-3">11.8.</span> Similarity Graph Construction</h3>
<div class="outline-text-3" id="text-11-8">
<ul class="org-ul">
<li>Several ways to construct similarity graphs:
<ul class="org-ul">
<li>Epsilon-neighborhood.</li>
<li>K-nearest neighbors.</li>
<li>Fully connected with edge weights (e.g., Gaussian kernel).</li>
</ul></li>
<li>Purpose: capture local similarity between patients.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc5407bd" class="outline-3">
<h3 id="orgc5407bd"><span class="section-number-3">11.9.</span> E Neighborhood Graph</h3>
<div class="outline-text-3" id="text-11-9">
<ul class="org-ul">
<li>Connect nodes within ε distance.</li>
<li>No edge if distance &gt; ε.</li>
<li>Defines sparsity based on ε.</li>
</ul>
</div>
</div>

<div id="outline-container-org4c8ba94" class="outline-3">
<h3 id="org4c8ba94"><span class="section-number-3">11.10.</span> K Nearest Neighbor Graph</h3>
<div class="outline-text-3" id="text-11-10">
<ul class="org-ul">
<li>Each node connects to its k-nearest neighbors (directed).</li>
<li>Results in sparse graphs for small k.</li>
<li>Variants: binary edges vs weighted edges based on distance.</li>
</ul>
</div>
</div>

<div id="outline-container-org28278e8" class="outline-3">
<h3 id="org28278e8"><span class="section-number-3">11.11.</span> Fully Connected Graph</h3>
<div class="outline-text-3" id="text-11-11">
<ul class="org-ul">
<li>Every node connected to every other.</li>
<li>Edge weights determined by similarity (e.g., Gaussian kernel).</li>
<li>Most dense graph representation.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcf8120f" class="outline-3">
<h3 id="orgcf8120f"><span class="section-number-3">11.12.</span> E Neighborhood Graph Quiz Question</h3>
<div class="outline-text-3" id="text-11-12">
<ul class="org-ul">
<li>Quiz: Select best ε value to reflect clustering structure in 2D patient data.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfd99d3c" class="outline-3">
<h3 id="orgfd99d3c"><span class="section-number-3">11.13.</span> E Neighborhood Graph Quiz Solution</h3>
<div class="outline-text-3" id="text-11-13">
<ul class="org-ul">
<li>Small ε: too many disconnected clusters.</li>
<li>Large ε: single cluster.</li>
<li>Optimal ε (e.g., B or C) reveals natural clusters in data.</li>
</ul>
</div>
</div>

<div id="outline-container-org26452e6" class="outline-3">
<h3 id="org26452e6"><span class="section-number-3">11.14.</span> Spectral Clustering Algorithm</h3>
<div class="outline-text-3" id="text-11-14">
<ul class="org-ul">
<li>Variations exist:
<ul class="org-ul">
<li>Unnormalized vs Normalized Spectral Clustering.</li>
</ul></li>
<li>Example: simplest uses eigenvectors + k-means.</li>
<li>Spectral clustering transforms data to eigenspace where clusters are clearer.</li>
<li>Recommended reading for theoretical underpinnings.</li>
</ul>
</div>
</div>

<div id="outline-container-org53e5025" class="outline-3">
<h3 id="org53e5025"><span class="section-number-3">11.15.</span> Big Data Conclusion</h3>
<div class="outline-text-3" id="text-11-15">
<ul class="org-ul">
<li>Wrap-up of the video series.</li>
<li>Presenter: Jimmy Son.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-orgf89255b" class="outline-2">
<h2 id="orgf89255b"><span class="section-number-2">12.</span> Dimensionality Reduction</h2>
<div class="outline-text-2" id="text-12">
</div>
<div id="outline-container-org776be0d" class="outline-3">
<h3 id="org776be0d"><span class="section-number-3">12.1.</span> Introduction to Dimensionality Reduction</h3>
<div class="outline-text-3" id="text-12-1">
<ul class="org-ul">
<li>Overview of dimensionality reduction as a method for simplifying noisy, high-dimensional data.</li>
<li>Methods covered:
<ul class="org-ul">
<li>Singular Value Decomposition (SVD)</li>
<li>Principal Component Analysis (PCA)</li>
<li>CUR Decomposition</li>
<li>Tensor Factorization</li>
</ul></li>
<li>Applications in predictive modeling and healthcare phenotyping.</li>
</ul>
</div>
</div>

<div id="outline-container-orgadde29c" class="outline-3">
<h3 id="orgadde29c"><span class="section-number-3">12.2.</span> Dimensionality Reduction</h3>
<div class="outline-text-3" id="text-12-2">
<ul class="org-ul">
<li>Recap of clustering algorithms (hard, soft, and scalable).</li>
<li>Motivation for dimensionality reduction: efficiency and robustness.</li>
<li>Classical methods:
<ul class="org-ul">
<li>SVD and PCA: summarize features via linear combinations.</li>
<li>CUR: uses actual data rows/columns, preserves sparsity.</li>
</ul></li>
<li>Introduction to Tensor Factorization for higher-order data.</li>
</ul>
</div>
</div>

<div id="outline-container-org9ae41a3" class="outline-3">
<h3 id="org9ae41a3"><span class="section-number-3">12.3.</span> Singular Value Decomposition</h3>
<div class="outline-text-3" id="text-12-3">
<ul class="org-ul">
<li>Definition: SVD decomposes a matrix X into U, Σ, Vᵗ.</li>
<li>U and V: orthogonal matrices (singular vectors).</li>
<li>Σ: diagonal matrix with singular values (in descending order).</li>
<li>Visual interpretation with patients (rows) and diseases (columns).</li>
</ul>
</div>
</div>

<div id="outline-container-org2bac861" class="outline-3">
<h3 id="org2bac861"><span class="section-number-3">12.4.</span> SVD Example</h3>
<div class="outline-text-3" id="text-12-4">
<ul class="org-ul">
<li>Spectral view: rank-1 approximations via σᵢUᵢVᵢᵗ.</li>
<li>Example with documents and terms:
<ul class="org-ul">
<li>CS vs. Medical documents.</li>
<li>Each rank-1 component captures topic-specific vocabulary.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc72e07f" class="outline-3">
<h3 id="orgc72e07f"><span class="section-number-3">12.5.</span> SVD Properties</h3>
<div class="outline-text-3" id="text-12-5">
<ul class="org-ul">
<li>V: eigenvectors of XᵗX with eigenvalues Σ².</li>
<li>U: eigenvectors of XXᵗ.</li>
<li>Reinforces relationship between SVD and eigen-decomposition.</li>
</ul>
</div>
</div>

<div id="outline-container-org2d630bb" class="outline-3">
<h3 id="org2d630bb"><span class="section-number-3">12.6.</span> Quiz  SVD Interpretation Question</h3>
<div class="outline-text-3" id="text-12-6">
<ul class="org-ul">
<li>Question about AᵗA matrix from a document-term matrix.</li>
</ul>
</div>
</div>

<div id="outline-container-org10de0ca" class="outline-3">
<h3 id="org10de0ca"><span class="section-number-3">12.7.</span> Quiz  SVD Interpretation Solution</h3>
<div class="outline-text-3" id="text-12-7">
<ul class="org-ul">
<li>Answer: AᵗA is a term-to-term similarity matrix.</li>
</ul>
</div>
</div>

<div id="outline-container-org8c7921f" class="outline-3">
<h3 id="org8c7921f"><span class="section-number-3">12.8.</span> Principal Component Analysis</h3>
<div class="outline-text-3" id="text-12-8">
<ul class="org-ul">
<li>PCA is derived from SVD: UΣ = principal components, Vᵗ = loadings.</li>
<li>Visualized with a patient-by-disease matrix.</li>
<li>Principal components capture the main variance direction.</li>
</ul>
</div>
</div>

<div id="outline-container-org02fe2cc" class="outline-3">
<h3 id="org02fe2cc"><span class="section-number-3">12.9.</span> PCA Interpretation</h3>
<div class="outline-text-3" id="text-12-9">
<ul class="org-ul">
<li>PCA reduces dimensions by projecting onto principal component directions.</li>
<li>Example with weight and height to one dimension.</li>
</ul>
</div>
</div>

<div id="outline-container-orgafc643c" class="outline-3">
<h3 id="orgafc643c"><span class="section-number-3">12.10.</span> Sparsity Problem with SVD</h3>
<div class="outline-text-3" id="text-12-10">
<ul class="org-ul">
<li>SVD and PCA generate dense outputs, losing sparsity of original data.</li>
<li>CUR decomposition preserves sparsity and reduces storage/computation cost.</li>
</ul>
</div>
</div>

<div id="outline-container-orgdcb5d8f" class="outline-3">
<h3 id="orgdcb5d8f"><span class="section-number-3">12.11.</span> CUR Decomposition</h3>
<div class="outline-text-3" id="text-12-11">
<ul class="org-ul">
<li>CUR approximates A ≈ CUR using actual rows and columns.</li>
<li>Comparison with SVD: CUR retains interpretability and sparsity.</li>
</ul>
</div>
</div>

<div id="outline-container-org4915ec5" class="outline-3">
<h3 id="org4915ec5"><span class="section-number-3">12.12.</span> CUR Algorithm</h3>
<div class="outline-text-3" id="text-12-12">
<ul class="org-ul">
<li>Algorithm steps:
<ol class="org-ol">
<li>Perform SVD.</li>
<li>Sample rows/columns based on vector norms.</li>
<li>Compute U as pseudo-inverse operations.</li>
</ol></li>
<li>Sampling based on row norms of U and V.</li>
</ul>
</div>
</div>

<div id="outline-container-org4f7b716" class="outline-3">
<h3 id="org4f7b716"><span class="section-number-3">12.13.</span> CUR Quiz Question</h3>
<div class="outline-text-3" id="text-12-13">
<ul class="org-ul">
<li>Question: what do C and R represent in CUR decomposition?</li>
</ul>
</div>
</div>

<div id="outline-container-org8b3e80c" class="outline-3">
<h3 id="org8b3e80c"><span class="section-number-3">12.14.</span> CUR Quiz Solution</h3>
<div class="outline-text-3" id="text-12-14">
<ul class="org-ul">
<li>Answer: C = actual diagnoses (columns), R = actual patients (rows).</li>
<li>Emphasizes CUR’s interpretability and sparsity preservation.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcb85f0b" class="outline-3">
<h3 id="orgcb85f0b"><span class="section-number-3">12.15.</span> Tensor for EHR</h3>
<div class="outline-text-3" id="text-12-15">
<ul class="org-ul">
<li>Tensor = generalized matrix (e.g., 3D tensor: patient × diagnosis × medication).</li>
<li>Captures multi-dimensional interactions in EHR data.</li>
</ul>
</div>
</div>

<div id="outline-container-org164a943" class="outline-3">
<h3 id="org164a943"><span class="section-number-3">12.16.</span> Tensor Slicing</h3>
<div class="outline-text-3" id="text-12-16">
<ul class="org-ul">
<li>Extract subtensors (matrices) by fixing all but two modes.</li>
<li>Examples:
<ul class="org-ul">
<li>Diagnosis-medication matrix for a patient.</li>
<li>Patients and diseases by medication.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org73cd67f" class="outline-3">
<h3 id="org73cd67f"><span class="section-number-3">12.17.</span> Rank 1 Tensor</h3>
<div class="outline-text-3" id="text-12-17">
<ul class="org-ul">
<li>Rank-1 tensor = outer product of vectors from each mode.</li>
<li>Element-wise: aᵢ * bⱼ * cₖ.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc3a92a5" class="outline-3">
<h3 id="orgc3a92a5"><span class="section-number-3">12.18.</span> Example Phenotype</h3>
<div class="outline-text-3" id="text-12-18">
<ul class="org-ul">
<li>Phenotypes modeled as rank-1 tensors.</li>
<li>Example: hypertension phenotype with specific meds and patients.</li>
</ul>
</div>
</div>

<div id="outline-container-org03feab0" class="outline-3">
<h3 id="org03feab0"><span class="section-number-3">12.19.</span> Phenotyping Through Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-19">
<ul class="org-ul">
<li>Tensor = sum of rank-1 tensors (each = phenotype).</li>
<li>Each component has patient, diagnosis, and medication factors.</li>
<li>λ = importance of each phenotype.</li>
</ul>
</div>
</div>

<div id="outline-container-org9cfb443" class="outline-3">
<h3 id="org9cfb443"><span class="section-number-3">12.20.</span> CP Decomposition</h3>
<div class="outline-text-3" id="text-12-20">
<ul class="org-ul">
<li>Canonical Polyadic (CP) Decomposition = sum of R rank-1 tensors.</li>
<li>Each tensor has one λ and one vector per mode.</li>
<li>Output: three factor matrices and λ vector.</li>
<li>Often used with non-negativity constraints.</li>
</ul>
</div>
</div>

<div id="outline-container-orgdbc728e" class="outline-3">
<h3 id="orgdbc728e"><span class="section-number-3">12.21.</span> Phenotyping Process Using Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-21">
<ul class="org-ul">
<li>Process:
<ol class="org-ol">
<li>Factorize training tensor to learn phenotype factors.</li>
<li>Project new patients onto phenotype space.</li>
</ol></li>
<li>Result: low-dimensional patient representation.</li>
</ul>
</div>
</div>

<div id="outline-container-org210c1ac" class="outline-3">
<h3 id="org210c1ac"><span class="section-number-3">12.22.</span> Phenotyping for Predictive Modeling</h3>
<div class="outline-text-3" id="text-12-22">
<ul class="org-ul">
<li>Build tensor from EHR events within an observation window.</li>
<li>Target = predict heart failure.</li>
<li>Tensor = patient × diagnosis × medication.</li>
<li>Phenotypes used as features for prediction.</li>
</ul>
</div>
</div>

<div id="outline-container-orgf6dc06b" class="outline-3">
<h3 id="orgf6dc06b"><span class="section-number-3">12.23.</span> Predictive Performance</h3>
<div class="outline-text-3" id="text-12-23">
<ul class="org-ul">
<li>Compared PCA, NMF, and Tensor Factorization.</li>
<li>All outperformed baseline with fewer features (phenotypes).</li>
<li>Tensor and NMF had better performance than PCA.</li>
</ul>
</div>
</div>

<div id="outline-container-orgac933bb" class="outline-3">
<h3 id="orgac933bb"><span class="section-number-3">12.24.</span> Major Disease Phenotypes</h3>
<div class="outline-text-3" id="text-12-24">
<ul class="org-ul">
<li>Tensor reveals intuitive disease phenotypes:
<ul class="org-ul">
<li>Uncomplicated diabetes (17.6%)</li>
<li>Mild hypertension (31.1%)</li>
</ul></li>
<li>Also identifies subtypes (e.g., mild/moderate/severe hypertension).</li>
</ul>
</div>
</div>

<div id="outline-container-org24d9c04" class="outline-3">
<h3 id="org24d9c04"><span class="section-number-3">12.25.</span> Tensor vs NMF</h3>
<div class="outline-text-3" id="text-12-25">
<ul class="org-ul">
<li>Tensor factorization offers more concise, interpretable phenotypes.</li>
<li>NMF generates complex feature sets with many interactions.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc6c7d1e" class="outline-3">
<h3 id="orgc6c7d1e"><span class="section-number-3">12.26.</span> Summary  Phenotyping Via Tensor Factorization</h3>
<div class="outline-text-3" id="text-12-26">
<ul class="org-ul">
<li>Tensor factorization = unsupervised, interpretable phenotyping.</li>
<li>Phenotypes help in prediction (e.g., heart failure).</li>
<li>Benefits: no expert supervision needed, compact representations.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orge890cd6" class="outline-2">
<h2 id="orge890cd6"><span class="section-number-2">13.</span> Patient Similarity</h2>
<div class="outline-text-2" id="text-13">
</div>
<div id="outline-container-orgeb17e19" class="outline-3">
<h3 id="orgeb17e19"><span class="section-number-3">13.1.</span> Introduction to Patient Similarity</h3>
<div class="outline-text-3" id="text-13-1">
<ul class="org-ul">
<li>Patient similarity introduces a new paradigm in medical practice.</li>
<li>Enables identifying similar patients for specific clinical contexts.</li>
<li>Supports pragmatic trials and practice-based medicine.</li>
<li>A supervised distance metric learning algorithm is introduced.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaa1c659" class="outline-3">
<h3 id="orgaa1c659"><span class="section-number-3">13.2.</span> Motivation</h3>
<div class="outline-text-3" id="text-13-2">
<ul class="org-ul">
<li>Traditional medicine relies on randomized clinical trials (RCT) for evidence (evidence-based medicine).</li>
<li>Limitations: patient heterogeneity, outdated/inapplicable guidelines.</li>
<li>EHR data enables a new paradigm: precision medicine.</li>
<li>Personalized care requires ability to measure patient similarity.</li>
</ul>
</div>
</div>

<div id="outline-container-org4c6ed91" class="outline-3">
<h3 id="org4c6ed91"><span class="section-number-3">13.3.</span> Traditional Paradigm</h3>
<div class="outline-text-3" id="text-13-3">
<ul class="org-ul">
<li>Evidence-based medicine uses:
<ul class="org-ul">
<li>RCTs to test hypotheses.</li>
<li>Publications and drug approvals.</li>
<li>Expert-created clinical guidelines.</li>
<li>Guideline application in clinical practice.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgd3c0f30" class="outline-3">
<h3 id="orgd3c0f30"><span class="section-number-3">13.4.</span> New Paradigm</h3>
<div class="outline-text-3" id="text-13-4">
<ul class="org-ul">
<li>Precision medicine promotes personalized care through:
<ul class="org-ul">
<li>Pragmatic trials using EHR data.</li>
<li>Patient similarity searches to inform treatment.</li>
<li>Practice-based evidence derived from real-world data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org46196e3" class="outline-3">
<h3 id="org46196e3"><span class="section-number-3">13.5.</span> Randomized Clinical Trials</h3>
<div class="outline-text-3" id="text-13-5">
<ul class="org-ul">
<li>Involves randomly assigning patients to control or treatment groups.</li>
<li>Compares average outcomes to assess treatment effectiveness.</li>
<li>Considered successful if treatment group shows better outcomes.</li>
</ul>
</div>
</div>

<div id="outline-container-orgcfa6b9f" class="outline-3">
<h3 id="orgcfa6b9f"><span class="section-number-3">13.6.</span> RCT Quiz Question</h3>
<div class="outline-text-3" id="text-13-6">
<ul class="org-ul">
<li>Asks about drawbacks of RCTs.</li>
<li>Lists controlled environment, testing one thing at a time, costs, etc.</li>
</ul>
</div>
</div>

<div id="outline-container-orgfef50e9" class="outline-3">
<h3 id="orgfef50e9"><span class="section-number-3">13.7.</span> RCT Quiz Solution</h3>
<div class="outline-text-3" id="text-13-7">
<ul class="org-ul">
<li>RCTs:
<ul class="org-ul">
<li>Require controlled environments with selective populations.</li>
<li>Test only one intervention at a time.</li>
<li>Are expensive and time-consuming.</li>
<li>Discover causal relationships (not a drawback).</li>
<li>Do not typically deal with noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orga5d4bcd" class="outline-3">
<h3 id="orga5d4bcd"><span class="section-number-3">13.8.</span> Pragmatic Trials</h3>
<div class="outline-text-3" id="text-13-8">
<ul class="org-ul">
<li>Measure real-world treatment effectiveness using EHR data.</li>
<li>Use similarity searches to group patients by treatment and outcomes.</li>
<li>Aim to recommend best-performing treatments to current patients.</li>
<li>Reflect real-world patient variations and support individual care choices.</li>
</ul>
</div>
</div>

<div id="outline-container-org80ed9ea" class="outline-3">
<h3 id="org80ed9ea"><span class="section-number-3">13.9.</span> Pragmatic Trial Quiz Question</h3>
<div class="outline-text-3" id="text-13-9">
<ul class="org-ul">
<li>Questions benefits and limitations of pragmatic trials.</li>
</ul>
</div>
</div>

<div id="outline-container-org7465824" class="outline-3">
<h3 id="org7465824"><span class="section-number-3">13.10.</span> Pragmatic Trial Quiz Solution</h3>
<div class="outline-text-3" id="text-13-10">
<ul class="org-ul">
<li>Pragmatic trials:
<ul class="org-ul">
<li>Can't test new drugs.</li>
<li>Operate in real-world settings.</li>
<li>Can gather data continuously.</li>
<li>Are less costly and faster than RCTs.</li>
<li>Don't establish causality (no randomization).</li>
<li>Deal with noisy data.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org2d1d61b" class="outline-3">
<h3 id="org2d1d61b"><span class="section-number-3">13.11.</span> Utilize Patient Similarity</h3>
<div class="outline-text-3" id="text-13-11">
<ul class="org-ul">
<li>Practice-based medicine uses patient similarity to:
<ul class="org-ul">
<li>Generate hypotheses from retrospective evidence.</li>
<li>Confirm them through RCTs.</li>
<li>Update clinical guidelines.</li>
</ul></li>
<li>Decision process:
<ul class="org-ul">
<li>Apply guidelines if available.</li>
<li>Use similar patient data if guidelines aren't applicable.</li>
<li>Use professional judgment if similar data is lacking.</li>
</ul></li>
<li>Paradigms are complementary (evidence-based + practice-based).</li>
</ul>
</div>
</div>

<div id="outline-container-org4294595" class="outline-3">
<h3 id="org4294595"><span class="section-number-3">13.12.</span> Patient Similarity Approaches</h3>
<div class="outline-text-3" id="text-13-12">
<ul class="org-ul">
<li>Distance metric learning: supervised approach using feature vectors and labels.</li>
<li>Learn distance function to reflect similarity.</li>
<li>Graph-based methods use medical ontologies/disease networks to relate patients.</li>
</ul>
</div>
</div>

<div id="outline-container-org1e1fb5f" class="outline-3">
<h3 id="org1e1fb5f"><span class="section-number-3">13.13.</span> Patient Similarity Through LSML</h3>
<div class="outline-text-3" id="text-13-13">
<ul class="org-ul">
<li>LSML (Locally Supervised Metric Learning):
<ul class="org-ul">
<li>Learns a distance metric tailored to clinical context.</li>
<li>Uses baseline similarity to retrieve potential matches.</li>
<li>Refines metric to bring similar patients closer and dissimilar ones further.</li>
</ul></li>
<li>Objective: maximize margin between homogeneous and heterogeneous neighbors.</li>
<li>Mathematical formulation:
<ul class="org-ul">
<li>Learn a low-rank symmetric matrix W (via Mahalanobis distance).</li>
<li>Maximize trace(WᵀHW) where H captures margin information.</li>
<li>Solved via eigenvalue decomposition.</li>
</ul></li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgb9b1c82" class="outline-2">
<h2 id="orgb9b1c82"><span class="section-number-2">14.</span> Deep Neural Network</h2>
<div class="outline-text-2" id="text-14">
</div>
<div id="outline-container-orgcef6028" class="outline-3">
<h3 id="orgcef6028"><span class="section-number-3">14.1.</span> Backward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-1">
<ul class="org-ul">
<li>A neuron processes inputs \( x_1, x_2, \dots, x_n \) with corresponding weights and a bias term to output a value \( y \).</li>
<li>Computation involves:
<ul class="org-ul">
<li>Linear combination: \( z = \sum w_i x_i + b \)</li>
<li>Nonlinear activation: \( y = g(z) \)</li>
</ul></li>
<li>Output type:
<ul class="org-ul">
<li>Binary (for classification) or continuous (for regression)</li>
</ul></li>
<li>Learning involves:
<ul class="org-ul">
<li>Specifying the activation function \( g \)</li>
<li>Learning weights and bias from data</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgc486986" class="outline-3">
<h3 id="orgc486986"><span class="section-number-3">14.2.</span> Sigmoid Function</h3>
<div class="outline-text-3" id="text-14-2">
<ul class="org-ul">
<li>Nonlinear transformation specified by the modeler (not learned from data).</li>
<li>Popular activation functions: sigmoid, tanh, ReLU.</li>
<li>Sigmoid function: \( \sigma(x) = \frac{1}{1 + e^{-x}} \)</li>
<li>Output range: [0, 1]; interpretable as probability</li>
<li>Common for classification tasks</li>
<li>Problem: vanishing gradients when \( x \) is very small or very large</li>
</ul>
</div>
</div>

<div id="outline-container-org9cb7cf1" class="outline-3">
<h3 id="org9cb7cf1"><span class="section-number-3">14.3.</span> TANH Function</h3>
<div class="outline-text-3" id="text-14-3">
<ul class="org-ul">
<li>Defined as: \( \tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}} = 2\sigma(2x) - 1 \)</li>
<li>Output range: [–1, 1]; centered at 0</li>
<li>Larger gradients near zero than sigmoid</li>
<li>Still suffers from vanishing gradient when input is far from 0</li>
</ul>
</div>
</div>

<div id="outline-container-org8cea475" class="outline-3">
<h3 id="org8cea475"><span class="section-number-3">14.4.</span> Rectified Linear Function</h3>
<div class="outline-text-3" id="text-14-4">
<ul class="org-ul">
<li>ReLU: \( \max(0, x) \)</li>
<li>Output: [0, ∞); linear for \( x > 0 \), 0 otherwise</li>
<li>Does not suffer from vanishing gradient (gradient = 1 for \( x > 0 \))</li>
</ul>
</div>
</div>

<div id="outline-container-org45385c9" class="outline-3">
<h3 id="org45385c9"><span class="section-number-3">14.5.</span> Activation Functions_ Summary</h3>
<div class="outline-text-3" id="text-14-5">
<ul class="org-ul">
<li>Sigmoid and tanh are bounded; ReLU is unbounded above.</li>
<li>Choice of activation function is application dependent.</li>
</ul>
</div>
</div>

<div id="outline-container-org13c92b2" class="outline-3">
<h3 id="org13c92b2"><span class="section-number-3">14.6.</span> Train a Single Neuron</h3>
<div class="outline-text-3" id="text-14-6">
<ul class="org-ul">
<li>Goal: Learn weights and bias to minimize loss</li>
<li>Loss function measures difference between prediction \( y \) and target \( t \)</li>
<li>Example: squared Euclidean loss</li>
<li>Training = minimizing loss via optimization</li>
</ul>
</div>
</div>

<div id="outline-container-orga6e786a" class="outline-3">
<h3 id="orga6e786a"><span class="section-number-3">14.7.</span> Stochastic Gradient Descent</h3>
<div class="outline-text-3" id="text-14-7">
<ul class="org-ul">
<li>Efficient method to train neurons</li>
<li>Procedure:
<ol class="org-ol">
<li>Initialize weights and bias randomly</li>
<li>For each example:
<ul class="org-ul">
<li>Compute gradient of loss w.r.t. weights and bias</li>
<li>Update parameters using gradient and learning rate \( \eta \)</li>
</ul></li>
</ol></li>
</ul>
</div>
</div>

<div id="outline-container-org92b16f8" class="outline-3">
<h3 id="org92b16f8"><span class="section-number-3">14.8.</span> Forward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-8">
<ul class="org-ul">
<li>Two-pass process: forward to compute \( y \), backward for gradients</li>
<li>Derivatives:
<ul class="org-ul">
<li>\( \frac{\partial z}{\partial w_i} = x_i \)</li>
<li>\( \frac{\partial y}{\partial z} = y(1 - y) \) (for sigmoid)</li>
</ul></li>
<li>Software often automates this process</li>
</ul>
</div>
</div>

<div id="outline-container-org9cbd531" class="outline-3">
<h3 id="org9cbd531"><span class="section-number-3">14.9.</span> Backward Computation for a Neuron</h3>
<div class="outline-text-3" id="text-14-9">
<ul class="org-ul">
<li>Use chain rule to compute gradients:
<ul class="org-ul">
<li>\( \frac{\partial L}{\partial w_i} = (y - t) \cdot y(1 - y) \cdot x_i \)</li>
<li>\( \frac{\partial L}{\partial b} = (y - t) \cdot y(1 - y) \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7be7c1a" class="outline-3">
<h3 id="org7be7c1a"><span class="section-number-3">14.10.</span> Train a Multilayer Neural Network</h3>
<div class="outline-text-3" id="text-14-10">
<ul class="org-ul">
<li>Similar process as single neuron</li>
<li>SGD updates weights and biases for each layer</li>
<li>Challenge: computing gradients for all layers</li>
</ul>
</div>
</div>

<div id="outline-container-orge279baf" class="outline-3">
<h3 id="orge279baf"><span class="section-number-3">14.11.</span> Forward Computation Part 1</h3>
<div class="outline-text-3" id="text-14-11">
<ul class="org-ul">
<li>Step-by-step example with 3 hidden units (h1, h2, h3) and 1 output</li>
<li>Each unit computes:
<ul class="org-ul">
<li>Linear combination: \( z \)</li>
<li>Activation: \( a = g(z) \)</li>
</ul></li>
<li>Final output is computed from hidden layer outputs</li>
</ul>
</div>
</div>

<div id="outline-container-orgd6223c0" class="outline-3">
<h3 id="orgd6223c0"><span class="section-number-3">14.12.</span> Forward Computation Part 2</h3>
<div class="outline-text-3" id="text-14-12">
<ul class="org-ul">
<li>Summary of previous computation</li>
<li>Network is fully connected</li>
<li>Shows the symmetry and modularity of neural networks</li>
</ul>
</div>
</div>

<div id="outline-container-org53ef4c8" class="outline-3">
<h3 id="org53ef4c8"><span class="section-number-3">14.13.</span> Forward Computation_ Vector Form</h3>
<div class="outline-text-3" id="text-14-13">
<ul class="org-ul">
<li>Computation represented in vector/matrix form:
<ul class="org-ul">
<li>\( z = Wx + b \)</li>
<li>\( a = g(z) \)</li>
</ul></li>
<li>Activation applied element-wise to vectors</li>
</ul>
</div>
</div>

<div id="outline-container-org5900422" class="outline-3">
<h3 id="org5900422"><span class="section-number-3">14.14.</span> Forward Computation_ More General Form</h3>
<div class="outline-text-3" id="text-14-14">
<ul class="org-ul">
<li>General forward pass for layer \( L \rightarrow L+1 \):
<ul class="org-ul">
<li>\( z^{(L+1)} = W^{(L)}a^{(L)} + b^{(L)} \)</li>
<li>\( a^{(L+1)} = g(z^{(L+1)}) \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7b37361" class="outline-3">
<h3 id="org7b37361"><span class="section-number-3">14.15.</span> Forward Computation_ Summary</h3>
<div class="outline-text-3" id="text-14-15">
<ul class="org-ul">
<li>Computation can be expressed:
<ul class="org-ul">
<li>Unit-by-unit</li>
<li>Vectorized (preferred for efficiency)</li>
<li>Generalized for arbitrary depth</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org23e6b6a" class="outline-3">
<h3 id="org23e6b6a"><span class="section-number-3">14.16.</span> Gradient Descent for Neural Networks</h3>
<div class="outline-text-3" id="text-14-16">
<ul class="org-ul">
<li>Recap:
<ul class="org-ul">
<li>Initialize weights/biases</li>
<li>Compute gradients</li>
<li>Apply updates</li>
</ul></li>
<li>Efficient gradient computation is critical → leads to backpropagation</li>
</ul>
</div>
</div>

<div id="outline-container-org3395083" class="outline-3">
<h3 id="org3395083"><span class="section-number-3">14.17.</span> Backward Propagation Part 1</h3>
<div class="outline-text-3" id="text-14-17">
<ul class="org-ul">
<li>Gradient of loss w.r.t. \( w_{ji}^{(l)} \) is:
<ul class="org-ul">
<li>\( \delta_j^{(l)} \cdot a_i^{(l-1)} \)</li>
</ul></li>
<li>\( \delta_j^{(l)} \): error signal for unit j at layer l</li>
<li>Chain rule used to derive this</li>
</ul>
</div>
</div>

<div id="outline-container-org45521e8" class="outline-3">
<h3 id="org45521e8"><span class="section-number-3">14.18.</span> Backward Propagation Part 2</h3>
<div class="outline-text-3" id="text-14-18">
<ul class="org-ul">
<li>Gradient w.r.t. bias \( b_j^{(l)} = \delta_j^{(l)} \)</li>
<li>Bias derivatives are simpler (derivative of z w.r.t. bias = 1)</li>
</ul>
</div>
</div>

<div id="outline-container-orgc014ec4" class="outline-3">
<h3 id="orgc014ec4"><span class="section-number-3">14.19.</span> Backward Propagation Part 3</h3>
<div class="outline-text-3" id="text-14-19">
<ul class="org-ul">
<li>Need to compute all \( \delta_j^{(l)} \) efficiently</li>
<li>Done in backward pass from output to input layer</li>
</ul>
</div>
</div>

<div id="outline-container-org87747d5" class="outline-3">
<h3 id="org87747d5"><span class="section-number-3">14.20.</span> Backward Propagation Part 4</h3>
<div class="outline-text-3" id="text-14-20">
<ul class="org-ul">
<li>Compute final layer delta:
<ul class="org-ul">
<li>\( \delta_k^{(L)} = \frac{\partial L}{\partial z_k^{(L)}} = y - t \)</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org8f3fcee" class="outline-3">
<h3 id="org8f3fcee"><span class="section-number-3">14.21.</span> Backward Propagation Part 5</h3>
<div class="outline-text-3" id="text-14-21">
<ul class="org-ul">
<li>Compute delta for previous layer:
<ul class="org-ul">
<li>Chain rule involving gradient of activation and next layer delta</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org1434215" class="outline-3">
<h3 id="org1434215"><span class="section-number-3">14.22.</span> Backward Propagation Part 6</h3>
<div class="outline-text-3" id="text-14-22">
<ul class="org-ul">
<li>For hidden layers:
<ul class="org-ul">
<li>Delta includes:
<ul class="org-ul">
<li>Sum over deltas from next layer</li>
<li>Corresponding weights</li>
<li>Derivative of activation at current layer</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgbfc8198" class="outline-3">
<h3 id="orgbfc8198"><span class="section-number-3">14.23.</span> Backward Propagation Summary</h3>
<div class="outline-text-3" id="text-14-23">
<ul class="org-ul">
<li>General form of delta:
<ul class="org-ul">
<li>\( \delta_j^{(l)} = g'(z_j^{(l)}) \sum_k w_{kj}^{(l)} \delta_k^{(l+1)} \)</li>
</ul></li>
<li>Backpropagation allows computing all gradients in one pass</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-org328e135" class="outline-2">
<h2 id="org328e135"><span class="section-number-2">15.</span> Convolutional Neural Network</h2>
<div class="outline-text-2" id="text-15">
</div>
<div id="outline-container-orgf501402" class="outline-3">
<h3 id="orgf501402"><span class="section-number-3">15.1.</span> Convolution Local Networks</h3>
<div class="outline-text-3" id="text-15-1">
<ul class="org-ul">
<li>Feedforward networks connect all units between layers, causing high parameter counts for high-dimensional data.</li>
<li>Local connectivity can reduce parameters by limiting neuron connections to adjacent units.</li>
<li>Convolutional Neural Networks (CNNs) utilize this local connectivity across layers.</li>
<li><b><b>Weight sharing</b></b>:
<ul class="org-ul">
<li>Reuses the same parameters across locations, further reducing total parameters.</li>
<li>Inspired by convolution in signal processing.</li>
<li>Example: Sharing weights reduces 6 parameters to 3 for two hidden units.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org02be772" class="outline-3">
<h3 id="org02be772"><span class="section-number-3">15.2.</span> Pooling Handling Distortion</h3>
<div class="outline-text-3" id="text-15-2">
<ul class="org-ul">
<li>Pooling layers add <b><b>translational invariance</b></b>, crucial for images and time series.</li>
<li><b><b>Max pooling</b></b> extracts maximum values, aiding stability under input shifts.</li>
<li>Example: Two shifted inputs can yield the same output after convolution + pooling.</li>
<li>This robustness is desirable for real-world applications where input variance is common.</li>
</ul>
</div>
</div>

<div id="outline-container-orgaad049f" class="outline-3">
<h3 id="orgaad049f"><span class="section-number-3">15.3.</span> Convolutional Neural Networks</h3>
<div class="outline-text-3" id="text-15-3">
<ul class="org-ul">
<li>CNNs are effective for grid-like data (e.g., images).</li>
<li>Advantages:
<ul class="org-ul">
<li>Sparse interconnections.</li>
<li>Parameter sharing.</li>
<li>Supports translational invariance.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgdce997e" class="outline-3">
<h3 id="orgdce997e"><span class="section-number-3">15.4.</span> Key Stages of CNN</h3>
<div class="outline-text-3" id="text-15-4">
<ul class="org-ul">
<li>A typical CNN architecture consists of:
<ul class="org-ul">
<li>Convolution layers</li>
<li>Pooling layers</li>
<li>Fully connected layers</li>
</ul></li>
<li>Example architecture:
<ul class="org-ul">
<li>Conv → Max Pool → Conv → Max Pool → 3×Conv → Pool → 3×Fully Connected</li>
</ul></li>
<li>More detail on design choices to follow.</li>
</ul>
</div>
</div>

<div id="outline-container-org6c269c0" class="outline-3">
<h3 id="org6c269c0"><span class="section-number-3">15.5.</span> 1-D Convolution</h3>
<div class="outline-text-3" id="text-15-5">
<ul class="org-ul">
<li>Illustrates convolution with 1D input: `[1, 2, -1, 1, -3]` and kernel `[1, 0, -1]`.</li>
<li>Output computed via sliding window and inner product.</li>
<li>Padding used at edges (zero-padding).</li>
<li>Introduces concept of <b><b>stride</b></b> (spacing between kernel applications).</li>
<li>Computationally intensive but highly parallelizable on GPUs.</li>
</ul>
</div>
</div>

<div id="outline-container-orgc1adde8" class="outline-3">
<h3 id="orgc1adde8"><span class="section-number-3">15.6.</span> Neural Network for 1-D Convolution</h3>
<div class="outline-text-3" id="text-15-6">
<ul class="org-ul">
<li>Convolution operation visualized as a neural network:
<ul class="org-ul">
<li>Applies shared weights across sliding input windows.</li>
<li>Two outputs H1 and H2 for two positions of the kernel.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org467b4b1" class="outline-3">
<h3 id="org467b4b1"><span class="section-number-3">15.7.</span> Convolution Layers</h3>
<div class="outline-text-3" id="text-15-7">
<ul class="org-ul">
<li>After convolution, <b><b>nonlinear activations</b></b> (e.g., ReLU) are applied.</li>
<li>Multiple filters (kernels) are often used to capture different patterns.</li>
<li>Output becomes a 3D tensor: height × width × number of filters.</li>
<li>Filters may extract features like average values or edges.</li>
</ul>
</div>
</div>

<div id="outline-container-orga5ae2af" class="outline-3">
<h3 id="orga5ae2af"><span class="section-number-3">15.8.</span> Pooling Layers</h3>
<div class="outline-text-3" id="text-15-8">
<ul class="org-ul">
<li>Pooling reduces spatial size of feature maps.</li>
<li>Types: Max pooling (most common), average pooling, sum pooling.</li>
<li>Example: 2×2 max pooling reduces 4D input to 2D output by taking max in each region.</li>
</ul>
</div>
</div>

<div id="outline-container-org4533173" class="outline-3">
<h3 id="org4533173"><span class="section-number-3">15.9.</span> Dimension Calculation of Convolution Layers</h3>
<div class="outline-text-3" id="text-15-9">
<ul class="org-ul">
<li>Input: `W1 × H1 × D1`</li>
<li>Output: `W2 × H2 × D2`
<ul class="org-ul">
<li>`W2 = (W1 - F + 2P) / S + 1`</li>
<li>`H2 = (H1 - F + 2P) / S + 1`</li>
<li>`D2 = K` (number of filters)</li>
</ul></li>
<li>Illustrated with example: input `227×227×3`, filter size `11×11`, stride `4`, output `55×55×96`.</li>
</ul>
</div>
</div>

<div id="outline-container-org10e80c7" class="outline-3">
<h3 id="org10e80c7"><span class="section-number-3">15.10.</span> Dimension Calculation of Pooling Layer</h3>
<div class="outline-text-3" id="text-15-10">
<ul class="org-ul">
<li>Similar to convolution but without padding.</li>
<li>Output size:
<ul class="org-ul">
<li>`W2 = (W1 - F) / S + 1`</li>
<li>`H2 = (H1 - F) / S + 1`</li>
<li>`D2 = D1` (depth remains the same)</li>
</ul></li>
<li>Example: pooling `3×3` on input `55×55×96` with stride 2 → output `27×27×96`.</li>
</ul>
</div>
</div>

<div id="outline-container-org77dc5c8" class="outline-3">
<h3 id="org77dc5c8"><span class="section-number-3">15.11.</span> Parameter Count</h3>
<div class="outline-text-3" id="text-15-11">
<ul class="org-ul">
<li>Convolution layers often dominate the depth but not the parameter count.</li>
<li>Fully connected layers have the most parameters.</li>
<li>Example:
<ul class="org-ul">
<li>Conv layers: 3.7M parameters.</li>
<li>Final 3 FC layers: 58.6M parameters.</li>
</ul></li>
<li>Highlights computational efficiency of convolution layers.</li>
</ul>
</div>
</div>

<div id="outline-container-org8022846" class="outline-3">
<h3 id="org8022846"><span class="section-number-3">15.12.</span> Forward Calculation of Convolution</h3>
<div class="outline-text-3" id="text-15-12">
<ul class="org-ul">
<li>Number of computations = output size × filter size.</li>
<li>Formula: `W2 × H2 × D2 × F × F × D1`</li>
<li>Necessary to estimate computational cost.</li>
</ul>
</div>
</div>

<div id="outline-container-org0c5bde8" class="outline-3">
<h3 id="org0c5bde8"><span class="section-number-3">15.13.</span> Forward Calculation of Pooling Layers</h3>
<div class="outline-text-3" id="text-15-13">
<ul class="org-ul">
<li>Follows similar computation logic as convolution.</li>
<li>Number of operations = output size + filter size.</li>
</ul>
</div>
</div>

<div id="outline-container-orged4147e" class="outline-3">
<h3 id="orged4147e"><span class="section-number-3">15.14.</span> Forward Calculation of Fully-Connected Layers</h3>
<div class="outline-text-3" id="text-15-14">
<ul class="org-ul">
<li>Computation = input size × output size.</li>
<li>Converts 3D input tensor to a vector before FC layer.</li>
</ul>
</div>
</div>

<div id="outline-container-org3a5a288" class="outline-3">
<h3 id="org3a5a288"><span class="section-number-3">15.15.</span> CNN Architectures and Healthcare Applications</h3>
<div class="outline-text-3" id="text-15-15">
<ul class="org-ul">
<li>Discusses famous CNNs:
<ul class="org-ul">
<li><b><b>AlexNet</b></b>: Large improvement in image classification (2012).</li>
<li><b><b>Inception</b></b>: Deeper with parallel paths.</li>
<li><b><b>ResNet</b></b>: Introduced <b><b>skip connections</b></b>, mitigates vanishing gradients.</li>
</ul></li>
<li>Performance on ImageNet:
<ul class="org-ul">
<li>ResNet achieves &lt;4% error, near human-level.</li>
</ul></li>
<li>Important design traits:
<ul class="org-ul">
<li>Deeper networks.</li>
<li>Parallel paths and residual learning.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgcd7c08b" class="outline-3">
<h3 id="orgcd7c08b"><span class="section-number-3">15.16.</span> Diabetic Retinopathy Diagnosis</h3>
<div class="outline-text-3" id="text-15-16">
<ul class="org-ul">
<li>CNNs applied to eye images for diabetic retinopathy detection.</li>
<li>Published in JAMA.</li>
<li>Achieved performance on par with domain experts.</li>
<li>ROC curve shows CNNs matching or exceeding doctors.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd9c2bfd" class="outline-3">
<h3 id="orgd9c2bfd"><span class="section-number-3">15.17.</span> Dermatologist Classification of Skin Cancer with DNN</h3>
<div class="outline-text-3" id="text-15-17">
<ul class="org-ul">
<li>CNNs used to classify skin lesions (benign vs. malignant).</li>
<li>Published in <b>Nature</b>.</li>
<li>CNN performance comparable or superior to dermatologists.</li>
<li>Visualization: algorithm ROC curve often exceeds human expert dots.</li>
</ul>
</div>
</div>
</div>


<div id="outline-container-orgbbb9693" class="outline-2">
<h2 id="orgbbb9693"><span class="section-number-2">16.</span> Recurrent Neural Network</h2>
<div class="outline-text-2" id="text-16">
</div>
<div id="outline-container-org0c8a137" class="outline-3">
<h3 id="org0c8a137"><span class="section-number-3">16.1.</span> Basic Concepts of RNN</h3>
<div class="outline-text-3" id="text-16-1">
<ul class="org-ul">
<li>Feedforward neural networks (FFNN):
<ul class="org-ul">
<li>Map input vector X → hidden state H → output Ŷ.</li>
<li>X can be a multi-hot vector of patient diseases.</li>
<li>H represents patient embedding; Ŷ is prediction (e.g. heart failure).</li>
</ul></li>
<li>Recurrent Neural Networks (RNN):
<ul class="org-ul">
<li>Similar structure to FFNN, with key difference: recursion in hidden state.</li>
<li>Hidden state H = function of input X and previous hidden state H.</li>
<li>Components:
<ul class="org-ul">
<li>U: weights from input to hidden</li>
<li>W: weights from previous hidden to current hidden</li>
<li>V: weights from hidden to output</li>
<li>f, g: activation functions</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org67193ac" class="outline-3">
<h3 id="org67193ac"><span class="section-number-3">16.2.</span> Basic RNN Structure</h3>
<div class="outline-text-3" id="text-16-2">
<ul class="org-ul">
<li>Unfolding RNN over time steps (t-1, t, t+1).</li>
<li>Each time step:
<ul class="org-ul">
<li>Input: Xₜ</li>
<li>Hidden state: Hₜ = f(UXₜ + WHₜ₋₁ + b₁)</li>
<li>Output: Ŷₜ = g(VHₜ + b₂)</li>
</ul></li>
<li>Parameters (U, W, V) shared across time steps.</li>
<li>Common RNN architectures:
<ul class="org-ul">
<li>One-to-many: e.g., image captioning.</li>
<li>Many-to-one: e.g., disease classification from clinical notes.</li>
<li>Many-to-many:
<ul class="org-ul">
<li>Seq-to-seq: e.g., machine translation.</li>
<li>Time-aligned: e.g., sequential disease prediction.</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org415755a" class="outline-3">
<h3 id="org415755a"><span class="section-number-3">16.3.</span> Forward Computation</h3>
<div class="outline-text-3" id="text-16-3">
<ul class="org-ul">
<li>Forward pass in training:
<ol class="org-ol">
<li>Compute Zₜ = UXₜ + WHₜ₋₁ + b₁</li>
<li>Hₜ = f(Zₜ)</li>
<li>oₜ = VHₜ + b₂</li>
<li>Ŷₜ = g(oₜ)</li>
</ol></li>
<li>Loss function:
<ul class="org-ul">
<li>Negative log likelihood (NLL) for sequence output.</li>
<li>f: e.g., tanh; g: e.g., softmax.</li>
<li>Total loss = sum over time steps.</li>
<li>For binary classification, simplified loss form shown.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-orgafdfe3a" class="outline-3">
<h3 id="orgafdfe3a"><span class="section-number-3">16.4.</span> Backpropagation Through Time</h3>
<div class="outline-text-3" id="text-16-4">
<ul class="org-ul">
<li>Backward pass (BPTT):
<ul class="org-ul">
<li>Compute ∂L/∂oₜ, then ∂L/∂Hₜ.</li>
<li>Last time step: ∂L/∂Hₜ = Vᵀ ∂L/∂oₜ</li>
<li>Earlier steps: ∂L/∂Hₜ includes gradient through both oₜ and Hₜ₊₁.</li>
</ul></li>
<li>Parameter updates:
<ul class="org-ul">
<li>Gradients used to update U, W, V, b₁, b₂.</li>
</ul></li>
<li>Practical note:
<ul class="org-ul">
<li>Most modern libraries automate this.</li>
</ul></li>
<li>Vanishing gradient problem:
<ul class="org-ul">
<li>Gradients diminish through long sequences.</li>
<li>Makes learning long-term dependencies difficult.</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org0f2b320" class="outline-3">
<h3 id="org0f2b320"><span class="section-number-3">16.5.</span> Standard RNN</h3>
<div class="outline-text-3" id="text-16-5">
<ul class="org-ul">
<li>Standard RNN cell:
<ul class="org-ul">
<li>Hₜ = tanh(WXₜ + UHₜ₋₁ + b)</li>
</ul></li>
<li>Suffers from vanishing gradient problem on long sequences.</li>
<li>Motivation for improved architectures like LSTM and GRU.</li>
</ul>
</div>
</div>

<div id="outline-container-orgd5f7b5b" class="outline-3">
<h3 id="orgd5f7b5b"><span class="section-number-3">16.6.</span> Gated Recurred Unit</h3>
<div class="outline-text-3" id="text-16-6">
<ul class="org-ul">
<li>GRU: simpler alternative to LSTM with good performance.</li>
<li>Key components:
<ul class="org-ul">
<li>Update gate (Zₜ): controls how much to keep from old H.</li>
<li>Reset gate (Rₜ): controls influence of old H on candidate state.</li>
</ul></li>
<li>Hidden state update:
<ul class="org-ul">
<li>New Hₜ = (1 - Zₜ) ∘ Hₜ₋₁ + Zₜ ∘ H̃ₜ</li>
</ul></li>
<li>Uses fewer parameters than LSTM, performs comparably in practice.</li>
</ul>
</div>
</div>

<div id="outline-container-orga1ec563" class="outline-3">
<h3 id="orga1ec563"><span class="section-number-3">16.7.</span> Bidirectional RNN</h3>
<div class="outline-text-3" id="text-16-7">
<ul class="org-ul">
<li>BiRNN = two RNNs:
<ul class="org-ul">
<li>One reads sequence forward, other backward.</li>
<li>Outputs are concatenated for each time step.</li>
</ul></li>
<li>Benefits:
<ul class="org-ul">
<li>Provides full context from both past and future.</li>
<li>Especially useful in NLP (e.g., semantic understanding).</li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org44d6f02" class="outline-3">
<h3 id="org44d6f02"><span class="section-number-3">16.8.</span> Healthcare Applications</h3>
<div class="outline-text-3" id="text-16-8">
<ul class="org-ul">
<li>Application: Doctor AI (2016)
<ul class="org-ul">
<li>Predicts future patient events using RNNs.</li>
<li>Input: sequence of clinical visits (multi-hot encoded).</li>
<li>Output: next visit’s diagnosis codes.</li>
</ul></li>
<li>Dataset: 260,000 patients, 10 years of data, 38k unique codes.
<ul class="org-ul">
<li>Sparse input vectors; ~10–15 nonzero values per visit.</li>
<li>Output space: 1183 diagnosis codes.</li>
</ul></li>
<li>Evaluation:
<ul class="org-ul">
<li>Top-k recall: measures how many true labels are in top-k predictions.</li>
<li>RNN outperformed baselines (heuristics, logistic regression).</li>
</ul></li>
<li>Transfer learning:
<ul class="org-ul">
<li>Pretrained RNNs on large datasets can improve performance on smaller hospital datasets.
<ul class="org-ul">
<li>Warm-start initialization leads to better generalization.</li>
</ul></li>
</ul></li>
</ul>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: W</p>
<p class="date">Created: 2025-03-29 Sat 11:52</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
